{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import itertools as it\n",
    "import os\n",
    "\n",
    "import pysam\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from intervaltree import Interval, IntervalTree\n",
    "\n",
    "import functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "LINEAR_COLUMNS = ('read_id', \n",
    "\t\t\t\t  'chrom', \n",
    "\t\t\t\t#   'start', \n",
    "\t\t\t\t#   'end', \n",
    "\t\t\t\t  'align_is_reverse',\n",
    "\t\t\t\t  'blocks', \n",
    "\t\t\t\t  'cigar',\n",
    "\t\t\t\t  'read',\n",
    "\t\t\t\t  )\n",
    "CIGARTUPLE_CODES = {0: 'M',\n",
    "\t\t\t\t\t1: 'I',\n",
    "\t\t\t\t\t2: 'D',\n",
    "\t\t\t\t\t3: 'N',\n",
    "\t\t\t\t\t4: 'S',\n",
    "\t\t\t\t\t5: 'H',\n",
    "\t\t\t\t\t6: 'P',\n",
    "\t\t\t\t\t7: '=',\n",
    "\t\t\t\t\t8: 'X',\n",
    "\t\t\t\t\t9: 'B',\n",
    "\t\t\t\t\t}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def infer_segments(row):\n",
    "\tif len(row['blocks']) == 1:\n",
    "\t\treturn row['blocks']\n",
    "\t\n",
    "\tsegs = []\n",
    "\tcurrent_block = row['blocks'][0]\n",
    "\tfor blocks_index in range(1, len(row['blocks'])):\n",
    "\t\tnext_block = row['blocks'][blocks_index]\n",
    "\t\tcigar_index = 2*blocks_index - 1\n",
    "\t\tgap_type = row['cigar'][cigar_index][0]\n",
    "\n",
    "\t\tif gap_type == 'N':\n",
    "\t\t\tsegs.append(current_block)\n",
    "\t\t\tcurrent_block = next_block\n",
    "\t\telif gap_type in ('I', 'D'):\n",
    "\t\t\tcurrent_block = (current_block[0], next_block[1])\n",
    "\t\telse:\n",
    "\t\t\traise RuntimeError(f\"{row['blocks']}\\n{row['cigar']}\\n{blocks_index}\")\n",
    "\t\n",
    "\tsegs.append(current_block)\n",
    "\t\t\n",
    "\treturn tuple(segs)\n",
    "\n",
    "\n",
    "def infer_common_genes(df):\n",
    "\tgene_ids = []\n",
    "\tfor i, seg_row in df.iterrows():\n",
    "\t\tfeatures = seg_row['exons'].union(seg_row['introns'])\n",
    "\t\tseg_genes = set()\n",
    "\t\tfor feat in features:\n",
    "\t\t\tseg_genes.update(feat.data['gene_id'])\n",
    "\t\tgene_ids.append(seg_genes)\n",
    "\n",
    "\tcommon_genes = set(gid for gid in gene_ids[0] if all(gid in set_genes for set_genes in gene_ids))\n",
    "\treturn common_genes\n",
    "\n",
    "\n",
    "def tree_covers_interval(tree:IntervalTree, interval:Interval) -> bool:\n",
    "\ttotal_coverage = False\n",
    "\tmerged_tree = tree.copy()\n",
    "\tmerged_tree.merge_overlaps(strict=False)\n",
    "\tfor merged_interval in merged_tree:\n",
    "\t\tif merged_interval.contains_interval(interval):\n",
    "\t\t\ttotal_coverage = True\n",
    "\t\n",
    "\treturn total_coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Get args\n",
    "# ref_exons, ref_introns, output_base, log_level = sys.argv[1:]\n",
    "# # Get logger\n",
    "# log = functions.get_logger(log_level)\n",
    "# log.debug(f'Args recieved: {sys.argv[1:]}')\n",
    "\n",
    "ref_exons = '/Users/trumanmooney/Library/CloudStorage/OneDrive-BrownUniversity/Documents/Projects/Lariat_mapping/reference_data/hg38_ref/exons.tsv.gz'\n",
    "ref_introns = '/Users/trumanmooney/Library/CloudStorage/OneDrive-BrownUniversity/Documents/Projects/Lariat_mapping/reference_data/hg38_ref/introns.tsv.gz'\n",
    "output_base = '/Users/trumanmooney/Library/CloudStorage/OneDrive-BrownUniversity/Documents/Projects/Lariat_mapping/output/pipeline/C22-1_100k/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "introns = pd.read_csv(ref_introns, sep='\\t')\n",
    "introns['gene_id'] = introns.gene_id.transform(lambda gid: set(gid.split(',')))\n",
    "introns['interval'] = introns.apply(lambda row: Interval(row['start'], row['end'], {'gene_id': row['gene_id'], 'strand': row['strand']}), axis=1)\n",
    "introns = introns.groupby('chrom').interval.agg(IntervalTree).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "exons = pd.read_csv(ref_exons, sep='\\t')\n",
    "exons['gene_id'] = exons.gene_id.transform(lambda gid: set(gid.split(',')))\n",
    "exons['interval'] = exons.apply(lambda row: Interval(row['start'], row['end'], {'gene_id': row['gene_id'], 'strand': row['strand']}), axis=1)\n",
    "exons = exons.groupby('chrom').interval.agg(IntervalTree).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>align_is_reverse</th>\n",
       "      <th>blocks</th>\n",
       "      <th>cigar</th>\n",
       "      <th>read</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41499</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>True</td>\n",
       "      <td>[(43161160, 43161190), (43162802, 43162922)]</td>\n",
       "      <td>((M, 30), (N, 1612), (M, 120))</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41500</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>False</td>\n",
       "      <td>[(43161094, 43161190), (43162802, 43162856)]</td>\n",
       "      <td>((M, 96), (N, 1612), (M, 54))</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44307</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11992</td>\n",
       "      <td>chr9</td>\n",
       "      <td>False</td>\n",
       "      <td>[(9442096, 9442246)]</td>\n",
       "      <td>((M, 150),)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44308</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11992</td>\n",
       "      <td>chr9</td>\n",
       "      <td>True</td>\n",
       "      <td>[(9442107, 9442257)]</td>\n",
       "      <td>((M, 150),)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44642</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:12085</td>\n",
       "      <td>chr14</td>\n",
       "      <td>True</td>\n",
       "      <td>[(49862683, 49862833)]</td>\n",
       "      <td>((M, 150),)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105542</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:27352</td>\n",
       "      <td>chr14</td>\n",
       "      <td>True</td>\n",
       "      <td>[(32163959, 32164109)]</td>\n",
       "      <td>((M, 150),)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15935</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>False</td>\n",
       "      <td>[(48233242, 48233392)]</td>\n",
       "      <td>((M, 150),)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15934</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>True</td>\n",
       "      <td>[(48233331, 48233481)]</td>\n",
       "      <td>((M, 150),)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34263</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>True</td>\n",
       "      <td>[(49586704, 49586854)]</td>\n",
       "      <td>((M, 150),)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34264</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>False</td>\n",
       "      <td>[(49563422, 49563424), (49586600, 49586748)]</td>\n",
       "      <td>((M, 2), (N, 23176), (M, 148))</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112620 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               read_id  chrom  \\\n",
       "41499   NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "41500   NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "44307   NGSNJ-086:229:GW200110425th:1:1101:10004:11992   chr9   \n",
       "44308   NGSNJ-086:229:GW200110425th:1:1101:10004:11992   chr9   \n",
       "44642   NGSNJ-086:229:GW200110425th:1:1101:10004:12085  chr14   \n",
       "...                                                ...    ...   \n",
       "105542   NGSNJ-086:229:GW200110425th:1:1101:9995:27352  chr14   \n",
       "15935     NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13   \n",
       "15934     NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13   \n",
       "34263     NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14   \n",
       "34264     NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14   \n",
       "\n",
       "        align_is_reverse                                        blocks  \\\n",
       "41499               True  [(43161160, 43161190), (43162802, 43162922)]   \n",
       "41500              False  [(43161094, 43161190), (43162802, 43162856)]   \n",
       "44307              False                          [(9442096, 9442246)]   \n",
       "44308               True                          [(9442107, 9442257)]   \n",
       "44642               True                        [(49862683, 49862833)]   \n",
       "...                  ...                                           ...   \n",
       "105542              True                        [(32163959, 32164109)]   \n",
       "15935              False                        [(48233242, 48233392)]   \n",
       "15934               True                        [(48233331, 48233481)]   \n",
       "34263               True                        [(49586704, 49586854)]   \n",
       "34264              False  [(49563422, 49563424), (49586600, 49586748)]   \n",
       "\n",
       "                                 cigar read  \n",
       "41499   ((M, 30), (N, 1612), (M, 120))    1  \n",
       "41500    ((M, 96), (N, 1612), (M, 54))    2  \n",
       "44307                      ((M, 150),)    1  \n",
       "44308                      ((M, 150),)    2  \n",
       "44642                      ((M, 150),)    2  \n",
       "...                                ...  ...  \n",
       "105542                     ((M, 150),)    1  \n",
       "15935                      ((M, 150),)    2  \n",
       "15934                      ((M, 150),)    1  \n",
       "34263                      ((M, 150),)    1  \n",
       "34264   ((M, 2), (N, 23176), (M, 148))    2  \n",
       "\n",
       "[112620 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "linear_reads = []\n",
    "for align in pysam.AlignmentFile(f'{output_base}mapped_reads.bam', 'rb'):\n",
    "\tlinear_reads.append([\n",
    "\t\t\t\t\talign.query_name, \n",
    "\t\t\t\t\talign.reference_name, \n",
    "\t\t\t\t\t# align.reference_start,\n",
    "\t\t\t\t\t# align.reference_end,\n",
    "\t\t\t\t\talign.is_reverse,\n",
    "\t\t\t\t\talign.get_blocks(),\n",
    "\t\t\t\t\talign.cigartuples,\n",
    "\t\t\t\t\talign.is_read1,\n",
    "\t\t\t\t\t])\n",
    "\n",
    "linear_reads = pd.DataFrame(linear_reads, columns=LINEAR_COLUMNS)\n",
    "\n",
    "# Fix starting columns\n",
    "linear_reads.read_id = linear_reads.read_id.astype('string')\n",
    "linear_reads.chrom = linear_reads.chrom.astype('category')\n",
    "linear_reads.cigar = linear_reads.cigar.transform(lambda cigar: tuple((CIGARTUPLE_CODES[op],length) for op, length in cigar))\n",
    "linear_reads.read = linear_reads.read.map({True: '1', False: '2'}).astype('category')\n",
    "linear_reads = linear_reads.sort_values(['read_id'])\n",
    "\n",
    "linear_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spliced\n",
      "False    82975\n",
      "True     29645\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>align_is_reverse</th>\n",
       "      <th>read</th>\n",
       "      <th>spliced</th>\n",
       "      <th>seg</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(43161160, 43161190, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(43162802, 43162922, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>(43161094, 43161190, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>(43162802, 43162856, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11992</td>\n",
       "      <td>chr9</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(9442096, 9442246, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142693</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>(48233242, 48233392, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142694</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(48233331, 48233481, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142695</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(49586704, 49586854, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142696</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>(49563422, 49563424, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142697</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>(49586600, 49586748, None)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>142698 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               read_id  chrom  \\\n",
       "0       NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "1       NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "2       NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "3       NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "4       NGSNJ-086:229:GW200110425th:1:1101:10004:11992   chr9   \n",
       "...                                                ...    ...   \n",
       "142693    NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13   \n",
       "142694    NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13   \n",
       "142695    NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14   \n",
       "142696    NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14   \n",
       "142697    NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14   \n",
       "\n",
       "        align_is_reverse read  spliced                         seg  \n",
       "0                   True    1     True  (43161160, 43161190, None)  \n",
       "1                   True    1     True  (43162802, 43162922, None)  \n",
       "2                  False    2     True  (43161094, 43161190, None)  \n",
       "3                  False    2     True  (43162802, 43162856, None)  \n",
       "4                  False    1    False    (9442096, 9442246, None)  \n",
       "...                  ...  ...      ...                         ...  \n",
       "142693             False    2    False  (48233242, 48233392, None)  \n",
       "142694              True    1    False  (48233331, 48233481, None)  \n",
       "142695              True    1    False  (49586704, 49586854, None)  \n",
       "142696             False    2     True  (49563422, 49563424, None)  \n",
       "142697             False    2     True  (49586600, 49586748, None)  \n",
       "\n",
       "[142698 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "linear_reads['spliced'] = linear_reads.blocks.transform(lambda blocks: len(blocks)>1)\n",
    "print(linear_reads.spliced.value_counts())\n",
    "linear_reads['segs'] = linear_reads.apply(infer_segments, axis=1, result_type='reduce')\n",
    "linear_reads = linear_reads.explode('segs', ignore_index=True)\n",
    "linear_reads['seg'] = linear_reads.segs.transform(lambda segs: Interval(*segs))\n",
    "linear_reads = linear_reads.drop(columns=['blocks', 'segs', 'cigar'])\n",
    "linear_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No introns in chrM\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>align_is_reverse</th>\n",
       "      <th>read</th>\n",
       "      <th>spliced</th>\n",
       "      <th>seg</th>\n",
       "      <th>exons</th>\n",
       "      <th>introns</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(43161160, 43161190, None)</td>\n",
       "      <td>{(43161051, 43161190, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(43162802, 43162922, None)</td>\n",
       "      <td>{(43162802, 43163210, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>(43161094, 43161190, None)</td>\n",
       "      <td>{(43161051, 43161190, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>(43162802, 43162856, None)</td>\n",
       "      <td>{(43162802, 43163210, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11992</td>\n",
       "      <td>chr9</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(9442096, 9442246, None)</td>\n",
       "      <td>{(9442059, 9442380, {'gene_id': {'ENSG00000265...</td>\n",
       "      <td>{(9397482, 9574731, {'gene_id': {'ENSG00000153...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142693</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>(48233242, 48233392, None)</td>\n",
       "      <td>{(48233202, 48233477, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{(48232698, 48253806, {'gene_id': {'ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142694</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(48233331, 48233481, None)</td>\n",
       "      <td>{(48233202, 48233477, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{(48232698, 48253806, {'gene_id': {'ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142695</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(49586704, 49586854, None)</td>\n",
       "      <td>{(49586579, 49586878, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>{(49586049, 49598399, {'gene_id': {'ENSG000002...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142696</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>(49563422, 49563424, None)</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142697</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>(49586600, 49586748, None)</td>\n",
       "      <td>{(49586579, 49586878, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>{(49586049, 49598399, {'gene_id': {'ENSG000002...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>142698 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               read_id  chrom  \\\n",
       "0       NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "1       NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "2       NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "3       NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "4       NGSNJ-086:229:GW200110425th:1:1101:10004:11992   chr9   \n",
       "...                                                ...    ...   \n",
       "142693    NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13   \n",
       "142694    NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13   \n",
       "142695    NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14   \n",
       "142696    NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14   \n",
       "142697    NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14   \n",
       "\n",
       "        align_is_reverse read  spliced                         seg  \\\n",
       "0                   True    1     True  (43161160, 43161190, None)   \n",
       "1                   True    1     True  (43162802, 43162922, None)   \n",
       "2                  False    2     True  (43161094, 43161190, None)   \n",
       "3                  False    2     True  (43162802, 43162856, None)   \n",
       "4                  False    1    False    (9442096, 9442246, None)   \n",
       "...                  ...  ...      ...                         ...   \n",
       "142693             False    2    False  (48233242, 48233392, None)   \n",
       "142694              True    1    False  (48233331, 48233481, None)   \n",
       "142695              True    1    False  (49586704, 49586854, None)   \n",
       "142696             False    2     True  (49563422, 49563424, None)   \n",
       "142697             False    2     True  (49586600, 49586748, None)   \n",
       "\n",
       "                                                    exons  \\\n",
       "0       {(43161051, 43161190, {'gene_id': {'ENSG000001...   \n",
       "1       {(43162802, 43163210, {'gene_id': {'ENSG000001...   \n",
       "2       {(43161051, 43161190, {'gene_id': {'ENSG000001...   \n",
       "3       {(43162802, 43163210, {'gene_id': {'ENSG000001...   \n",
       "4       {(9442059, 9442380, {'gene_id': {'ENSG00000265...   \n",
       "...                                                   ...   \n",
       "142693  {(48233202, 48233477, {'gene_id': {'ENSG000001...   \n",
       "142694  {(48233202, 48233477, {'gene_id': {'ENSG000001...   \n",
       "142695  {(49586579, 49586878, {'gene_id': {'ENSG000002...   \n",
       "142696                                                 {}   \n",
       "142697  {(49586579, 49586878, {'gene_id': {'ENSG000002...   \n",
       "\n",
       "                                                  introns  \n",
       "0                                                      {}  \n",
       "1                                                      {}  \n",
       "2                                                      {}  \n",
       "3                                                      {}  \n",
       "4       {(9397482, 9574731, {'gene_id': {'ENSG00000153...  \n",
       "...                                                   ...  \n",
       "142693  {(48232698, 48253806, {'gene_id': {'ENSG000001...  \n",
       "142694  {(48232698, 48253806, {'gene_id': {'ENSG000001...  \n",
       "142695  {(49586049, 49598399, {'gene_id': {'ENSG000002...  \n",
       "142696                                                 {}  \n",
       "142697  {(49586049, 49598399, {'gene_id': {'ENSG000002...  \n",
       "\n",
       "[142698 rows x 8 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for chrom in linear_reads.chrom.unique():\n",
    "\tif chrom not in exons.keys():\n",
    "\t\tprint(f'No exons in {chrom}')\n",
    "\t\tcontinue\n",
    "\tchrom_exons = exons[chrom]\n",
    "\tlinear_reads.loc[linear_reads.chrom==chrom, 'exons'] = linear_reads.loc[linear_reads.chrom==chrom, 'seg'].transform(chrom_exons.overlap)\n",
    "\t\n",
    "\tif chrom not in introns.keys():\n",
    "\t\tprint(f'No introns in {chrom}')\n",
    "\t\tcontinue\n",
    "\tchrom_introns = introns[chrom]\n",
    "\tlinear_reads.loc[linear_reads.chrom==chrom, 'introns'] = linear_reads.loc[linear_reads.chrom==chrom, 'seg'].transform(chrom_introns.overlap)\n",
    "\n",
    "linear_reads.exons = linear_reads.exons.fillna('').transform(set)\n",
    "linear_reads.introns = linear_reads.introns.fillna('').transform(set)\n",
    "linear_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/zp/3k8jr14j3wgcym12kxf4tmfw0000gr/T/ipykernel_7089/4187516586.py:2: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  linear_reads['common_genes'] = linear_reads.read_id.map(linear_reads.groupby('read_id').apply(infer_common_genes))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "common_genes\n",
      "1    90038\n",
      "2    39312\n",
      "0    11667\n",
      "3     1578\n",
      "4       91\n",
      "5       12\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>align_is_reverse</th>\n",
       "      <th>read</th>\n",
       "      <th>spliced</th>\n",
       "      <th>seg</th>\n",
       "      <th>exons</th>\n",
       "      <th>introns</th>\n",
       "      <th>Intergenic</th>\n",
       "      <th>common_genes</th>\n",
       "      <th>exons_f</th>\n",
       "      <th>introns_f</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(43161160, 43161190, None)</td>\n",
       "      <td>{(43161051, 43161190, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "      <td>False</td>\n",
       "      <td>{ENSG00000100300.18}</td>\n",
       "      <td>[(43161051, 43161190, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(43162802, 43162922, None)</td>\n",
       "      <td>{(43162802, 43163210, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "      <td>False</td>\n",
       "      <td>{ENSG00000100300.18}</td>\n",
       "      <td>[(43162802, 43163210, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>(43161094, 43161190, None)</td>\n",
       "      <td>{(43161051, 43161190, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "      <td>False</td>\n",
       "      <td>{ENSG00000100300.18}</td>\n",
       "      <td>[(43161051, 43161190, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>(43162802, 43162856, None)</td>\n",
       "      <td>{(43162802, 43163210, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "      <td>False</td>\n",
       "      <td>{ENSG00000100300.18}</td>\n",
       "      <td>[(43162802, 43163210, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11992</td>\n",
       "      <td>chr9</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(9442096, 9442246, None)</td>\n",
       "      <td>{(9442059, 9442380, {'gene_id': {'ENSG00000265...</td>\n",
       "      <td>{(9397482, 9574731, {'gene_id': {'ENSG00000153...</td>\n",
       "      <td>False</td>\n",
       "      <td>{ENSG00000153707.19, ENSG00000265735.2}</td>\n",
       "      <td>[(9442059, 9442380, {'gene_id': {'ENSG00000265...</td>\n",
       "      <td>[(9397482, 9574731, {'gene_id': {'ENSG00000153...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142693</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>(48233242, 48233392, None)</td>\n",
       "      <td>{(48233202, 48233477, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{(48232698, 48253806, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>False</td>\n",
       "      <td>{ENSG00000136156.15}</td>\n",
       "      <td>[(48233202, 48233477, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>[(48232698, 48253806, {'gene_id': {'ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142694</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(48233331, 48233481, None)</td>\n",
       "      <td>{(48233202, 48233477, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{(48232698, 48253806, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>False</td>\n",
       "      <td>{ENSG00000136156.15}</td>\n",
       "      <td>[(48233202, 48233477, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>[(48232698, 48253806, {'gene_id': {'ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142695</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(49586704, 49586854, None)</td>\n",
       "      <td>{(49586579, 49586878, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>{(49586049, 49598399, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142696</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>(49563422, 49563424, None)</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142697</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>(49586600, 49586748, None)</td>\n",
       "      <td>{(49586579, 49586878, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>{(49586049, 49598399, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>142698 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               read_id  chrom  \\\n",
       "0       NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "1       NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "2       NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "3       NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "4       NGSNJ-086:229:GW200110425th:1:1101:10004:11992   chr9   \n",
       "...                                                ...    ...   \n",
       "142693    NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13   \n",
       "142694    NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13   \n",
       "142695    NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14   \n",
       "142696    NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14   \n",
       "142697    NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14   \n",
       "\n",
       "        align_is_reverse read  spliced                         seg  \\\n",
       "0                   True    1     True  (43161160, 43161190, None)   \n",
       "1                   True    1     True  (43162802, 43162922, None)   \n",
       "2                  False    2     True  (43161094, 43161190, None)   \n",
       "3                  False    2     True  (43162802, 43162856, None)   \n",
       "4                  False    1    False    (9442096, 9442246, None)   \n",
       "...                  ...  ...      ...                         ...   \n",
       "142693             False    2    False  (48233242, 48233392, None)   \n",
       "142694              True    1    False  (48233331, 48233481, None)   \n",
       "142695              True    1    False  (49586704, 49586854, None)   \n",
       "142696             False    2     True  (49563422, 49563424, None)   \n",
       "142697             False    2     True  (49586600, 49586748, None)   \n",
       "\n",
       "                                                    exons  \\\n",
       "0       {(43161051, 43161190, {'gene_id': {'ENSG000001...   \n",
       "1       {(43162802, 43163210, {'gene_id': {'ENSG000001...   \n",
       "2       {(43161051, 43161190, {'gene_id': {'ENSG000001...   \n",
       "3       {(43162802, 43163210, {'gene_id': {'ENSG000001...   \n",
       "4       {(9442059, 9442380, {'gene_id': {'ENSG00000265...   \n",
       "...                                                   ...   \n",
       "142693  {(48233202, 48233477, {'gene_id': {'ENSG000001...   \n",
       "142694  {(48233202, 48233477, {'gene_id': {'ENSG000001...   \n",
       "142695  {(49586579, 49586878, {'gene_id': {'ENSG000002...   \n",
       "142696                                                 {}   \n",
       "142697  {(49586579, 49586878, {'gene_id': {'ENSG000002...   \n",
       "\n",
       "                                                  introns  Intergenic  \\\n",
       "0                                                      {}       False   \n",
       "1                                                      {}       False   \n",
       "2                                                      {}       False   \n",
       "3                                                      {}       False   \n",
       "4       {(9397482, 9574731, {'gene_id': {'ENSG00000153...       False   \n",
       "...                                                   ...         ...   \n",
       "142693  {(48232698, 48253806, {'gene_id': {'ENSG000001...       False   \n",
       "142694  {(48232698, 48253806, {'gene_id': {'ENSG000001...       False   \n",
       "142695  {(49586049, 49598399, {'gene_id': {'ENSG000002...       False   \n",
       "142696                                                 {}        True   \n",
       "142697  {(49586049, 49598399, {'gene_id': {'ENSG000002...       False   \n",
       "\n",
       "                                   common_genes  \\\n",
       "0                          {ENSG00000100300.18}   \n",
       "1                          {ENSG00000100300.18}   \n",
       "2                          {ENSG00000100300.18}   \n",
       "3                          {ENSG00000100300.18}   \n",
       "4       {ENSG00000153707.19, ENSG00000265735.2}   \n",
       "...                                         ...   \n",
       "142693                     {ENSG00000136156.15}   \n",
       "142694                     {ENSG00000136156.15}   \n",
       "142695                                       {}   \n",
       "142696                                       {}   \n",
       "142697                                       {}   \n",
       "\n",
       "                                                  exons_f  \\\n",
       "0       [(43161051, 43161190, {'gene_id': {'ENSG000001...   \n",
       "1       [(43162802, 43163210, {'gene_id': {'ENSG000001...   \n",
       "2       [(43161051, 43161190, {'gene_id': {'ENSG000001...   \n",
       "3       [(43162802, 43163210, {'gene_id': {'ENSG000001...   \n",
       "4       [(9442059, 9442380, {'gene_id': {'ENSG00000265...   \n",
       "...                                                   ...   \n",
       "142693  [(48233202, 48233477, {'gene_id': {'ENSG000001...   \n",
       "142694  [(48233202, 48233477, {'gene_id': {'ENSG000001...   \n",
       "142695                                                 []   \n",
       "142696                                                 []   \n",
       "142697                                                 []   \n",
       "\n",
       "                                                introns_f  \n",
       "0                                                      []  \n",
       "1                                                      []  \n",
       "2                                                      []  \n",
       "3                                                      []  \n",
       "4       [(9397482, 9574731, {'gene_id': {'ENSG00000153...  \n",
       "...                                                   ...  \n",
       "142693  [(48232698, 48253806, {'gene_id': {'ENSG000001...  \n",
       "142694  [(48232698, 48253806, {'gene_id': {'ENSG000001...  \n",
       "142695                                                 []  \n",
       "142696                                                 []  \n",
       "142697                                                 []  \n",
       "\n",
       "[142698 rows x 12 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "linear_reads['Intergenic'] = (linear_reads.exons.transform(len)==0) & (linear_reads.introns.transform(len)==0)\n",
    "linear_reads['common_genes'] = linear_reads.read_id.map(linear_reads.groupby('read_id').apply(infer_common_genes))\n",
    "print(linear_reads.common_genes.transform(len).value_counts())\n",
    "linear_reads['exons_f'] = linear_reads.apply(lambda row: IntervalTree(exon for exon in row['exons'] if len(exon.data['gene_id'].intersection(row['common_genes']))>0), axis=1)\n",
    "linear_reads['introns_f'] = linear_reads.apply(lambda row: IntervalTree(intron for intron in row['introns'] if len(intron.data['gene_id'].intersection(row['common_genes']))>0), axis=1)\n",
    "linear_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "seg_class\n",
      "Exon junction     49030\n",
      "Ambiguous         40435\n",
      "Exonic            25832\n",
      "Intronic          22269\n",
      "Intergenic         5124\n",
      "Starts at 5'ss        8\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>align_is_reverse</th>\n",
       "      <th>read</th>\n",
       "      <th>spliced</th>\n",
       "      <th>seg</th>\n",
       "      <th>exons</th>\n",
       "      <th>introns</th>\n",
       "      <th>Intergenic</th>\n",
       "      <th>common_genes</th>\n",
       "      <th>exons_f</th>\n",
       "      <th>introns_f</th>\n",
       "      <th>seg_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(43161160, 43161190, None)</td>\n",
       "      <td>{(43161051, 43161190, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "      <td>False</td>\n",
       "      <td>{ENSG00000100300.18}</td>\n",
       "      <td>[(43161051, 43161190, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>[]</td>\n",
       "      <td>Exon junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(43162802, 43162922, None)</td>\n",
       "      <td>{(43162802, 43163210, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "      <td>False</td>\n",
       "      <td>{ENSG00000100300.18}</td>\n",
       "      <td>[(43162802, 43163210, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>[]</td>\n",
       "      <td>Exon junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>(43161094, 43161190, None)</td>\n",
       "      <td>{(43161051, 43161190, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "      <td>False</td>\n",
       "      <td>{ENSG00000100300.18}</td>\n",
       "      <td>[(43161051, 43161190, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>[]</td>\n",
       "      <td>Exon junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>(43162802, 43162856, None)</td>\n",
       "      <td>{(43162802, 43163210, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "      <td>False</td>\n",
       "      <td>{ENSG00000100300.18}</td>\n",
       "      <td>[(43162802, 43163210, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>[]</td>\n",
       "      <td>Exon junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11992</td>\n",
       "      <td>chr9</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(9442096, 9442246, None)</td>\n",
       "      <td>{(9442059, 9442380, {'gene_id': {'ENSG00000265...</td>\n",
       "      <td>{(9397482, 9574731, {'gene_id': {'ENSG00000153...</td>\n",
       "      <td>False</td>\n",
       "      <td>{ENSG00000153707.19, ENSG00000265735.2}</td>\n",
       "      <td>[(9442059, 9442380, {'gene_id': {'ENSG00000265...</td>\n",
       "      <td>[(9397482, 9574731, {'gene_id': {'ENSG00000153...</td>\n",
       "      <td>Ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142693</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>(48233242, 48233392, None)</td>\n",
       "      <td>{(48233202, 48233477, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{(48232698, 48253806, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>False</td>\n",
       "      <td>{ENSG00000136156.15}</td>\n",
       "      <td>[(48233202, 48233477, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>[(48232698, 48253806, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>Ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142694</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(48233331, 48233481, None)</td>\n",
       "      <td>{(48233202, 48233477, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{(48232698, 48253806, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>False</td>\n",
       "      <td>{ENSG00000136156.15}</td>\n",
       "      <td>[(48233202, 48233477, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>[(48232698, 48253806, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>Ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142695</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(49586704, 49586854, None)</td>\n",
       "      <td>{(49586579, 49586878, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>{(49586049, 49598399, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>Ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142696</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>(49563422, 49563424, None)</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>Intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142697</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>(49586600, 49586748, None)</td>\n",
       "      <td>{(49586579, 49586878, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>{(49586049, 49598399, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>Ambiguous</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>142698 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               read_id  chrom  \\\n",
       "0       NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "1       NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "2       NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "3       NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "4       NGSNJ-086:229:GW200110425th:1:1101:10004:11992   chr9   \n",
       "...                                                ...    ...   \n",
       "142693    NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13   \n",
       "142694    NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13   \n",
       "142695    NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14   \n",
       "142696    NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14   \n",
       "142697    NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14   \n",
       "\n",
       "        align_is_reverse read  spliced                         seg  \\\n",
       "0                   True    1     True  (43161160, 43161190, None)   \n",
       "1                   True    1     True  (43162802, 43162922, None)   \n",
       "2                  False    2     True  (43161094, 43161190, None)   \n",
       "3                  False    2     True  (43162802, 43162856, None)   \n",
       "4                  False    1    False    (9442096, 9442246, None)   \n",
       "...                  ...  ...      ...                         ...   \n",
       "142693             False    2    False  (48233242, 48233392, None)   \n",
       "142694              True    1    False  (48233331, 48233481, None)   \n",
       "142695              True    1    False  (49586704, 49586854, None)   \n",
       "142696             False    2     True  (49563422, 49563424, None)   \n",
       "142697             False    2     True  (49586600, 49586748, None)   \n",
       "\n",
       "                                                    exons  \\\n",
       "0       {(43161051, 43161190, {'gene_id': {'ENSG000001...   \n",
       "1       {(43162802, 43163210, {'gene_id': {'ENSG000001...   \n",
       "2       {(43161051, 43161190, {'gene_id': {'ENSG000001...   \n",
       "3       {(43162802, 43163210, {'gene_id': {'ENSG000001...   \n",
       "4       {(9442059, 9442380, {'gene_id': {'ENSG00000265...   \n",
       "...                                                   ...   \n",
       "142693  {(48233202, 48233477, {'gene_id': {'ENSG000001...   \n",
       "142694  {(48233202, 48233477, {'gene_id': {'ENSG000001...   \n",
       "142695  {(49586579, 49586878, {'gene_id': {'ENSG000002...   \n",
       "142696                                                 {}   \n",
       "142697  {(49586579, 49586878, {'gene_id': {'ENSG000002...   \n",
       "\n",
       "                                                  introns  Intergenic  \\\n",
       "0                                                      {}       False   \n",
       "1                                                      {}       False   \n",
       "2                                                      {}       False   \n",
       "3                                                      {}       False   \n",
       "4       {(9397482, 9574731, {'gene_id': {'ENSG00000153...       False   \n",
       "...                                                   ...         ...   \n",
       "142693  {(48232698, 48253806, {'gene_id': {'ENSG000001...       False   \n",
       "142694  {(48232698, 48253806, {'gene_id': {'ENSG000001...       False   \n",
       "142695  {(49586049, 49598399, {'gene_id': {'ENSG000002...       False   \n",
       "142696                                                 {}        True   \n",
       "142697  {(49586049, 49598399, {'gene_id': {'ENSG000002...       False   \n",
       "\n",
       "                                   common_genes  \\\n",
       "0                          {ENSG00000100300.18}   \n",
       "1                          {ENSG00000100300.18}   \n",
       "2                          {ENSG00000100300.18}   \n",
       "3                          {ENSG00000100300.18}   \n",
       "4       {ENSG00000153707.19, ENSG00000265735.2}   \n",
       "...                                         ...   \n",
       "142693                     {ENSG00000136156.15}   \n",
       "142694                     {ENSG00000136156.15}   \n",
       "142695                                       {}   \n",
       "142696                                       {}   \n",
       "142697                                       {}   \n",
       "\n",
       "                                                  exons_f  \\\n",
       "0       [(43161051, 43161190, {'gene_id': {'ENSG000001...   \n",
       "1       [(43162802, 43163210, {'gene_id': {'ENSG000001...   \n",
       "2       [(43161051, 43161190, {'gene_id': {'ENSG000001...   \n",
       "3       [(43162802, 43163210, {'gene_id': {'ENSG000001...   \n",
       "4       [(9442059, 9442380, {'gene_id': {'ENSG00000265...   \n",
       "...                                                   ...   \n",
       "142693  [(48233202, 48233477, {'gene_id': {'ENSG000001...   \n",
       "142694  [(48233202, 48233477, {'gene_id': {'ENSG000001...   \n",
       "142695                                                 []   \n",
       "142696                                                 []   \n",
       "142697                                                 []   \n",
       "\n",
       "                                                introns_f      seg_class  \n",
       "0                                                      []  Exon junction  \n",
       "1                                                      []  Exon junction  \n",
       "2                                                      []  Exon junction  \n",
       "3                                                      []  Exon junction  \n",
       "4       [(9397482, 9574731, {'gene_id': {'ENSG00000153...      Ambiguous  \n",
       "...                                                   ...            ...  \n",
       "142693  [(48232698, 48253806, {'gene_id': {'ENSG000001...      Ambiguous  \n",
       "142694  [(48232698, 48253806, {'gene_id': {'ENSG000001...      Ambiguous  \n",
       "142695                                                 []      Ambiguous  \n",
       "142696                                                 []     Intergenic  \n",
       "142697                                                 []      Ambiguous  \n",
       "\n",
       "[142698 rows x 13 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def classify_seg(row):\n",
    "\tif row['Intergenic'] is True:\n",
    "\t\treturn 'Intergenic'\n",
    "\t\n",
    "\tif len(row['exons_f'])==0 and len(row['introns_f'])==0:\n",
    "\t\treturn 'Ambiguous'\n",
    "\t\n",
    "\tif len(row['exons_f'])==0:\n",
    "\t\tif row['align_is_reverse'] is False and any(row['seg'].begin==intron.begin for intron in row['introns_f'] if intron.data['strand']=='+'):\n",
    "\t\t\treturn \"Starts at 5'ss\"\n",
    "\t\telif row['align_is_reverse'] is True and any(row['seg'].end==intron.end for intron in row['introns_f'] if intron.data['strand']=='-'):\n",
    "\t\t\treturn \"Starts at 5'ss\"\n",
    "\t\telse:\n",
    "\t\t\treturn 'Intronic'\n",
    "\t\n",
    "\t#TODO: Maybe move this inside next if and add after pre-mRNA check since segment could cross intron-exon junction AND get spliced at another exon junction\n",
    "\tif any([row['seg'].begin==exon.begin for exon in row['exons_f']]) or any([row['seg'].end==exon.end for exon in row['exons_f']]):\n",
    "\t\treturn 'Exon junction'\n",
    "\n",
    "\tif len(row['introns_f'])==0:\n",
    "\t\treturn 'Exonic'\n",
    "\t\n",
    "\t# At this point there's at least 1 filtered exon and 1 filtered intron\n",
    "\t# We need to see if it's pre-mRNA\n",
    "\tfor exon, intron in it.product(row['exons_f'], row['introns_f']):\n",
    "\t\tif len(exon.data['gene_id'].intersection(intron.data['gene_id']))==0:\n",
    "\t\t\tcontinue\n",
    "\t\t\n",
    "\t\t# Check if we have an exon-to-intron junction\n",
    "\t\tif exon.end-1 == intron.begin:\n",
    "\t\t\tjunc_spot = intron.begin\n",
    "\t\t\texon_5bp = junc_spot-5\n",
    "\t\t\tintron_5bp = junc_spot+4\n",
    "\n",
    "\t\t\t# # Disregard if segment doesn't have at least 5bp on each side\n",
    "\t\t\t# if exon_5bp < row['seg'].begin or intron_5bp >= row['seg'].end:\n",
    "\t\t\t# \tcontinue\n",
    "\t\t\t# Disregard if last 5bp of exon overlaps any introns\n",
    "\t\t\t# Disregard if first 5bp of intron overlaps any exons\n",
    "\t\t\t# This accounts for alternative splice sites\n",
    "\t\t\tif len(row['introns_f'].overlap(exon_5bp, junc_spot)) > 0:\n",
    "\t\t\t\tcontinue\n",
    "\t\t\tif len(row['exons_f'].overlap(junc_spot, intron_5bp+1)) > 0:\n",
    "\t\t\t\tcontinue\n",
    "\n",
    "\t\t\treturn 'pre-mRNA'\n",
    "\t\t\n",
    "\t\t# Check if we have an intron-to-exon junction\n",
    "\t\telif intron.end-1 == exon.begin:\n",
    "\t\t\tjunc_spot = exon.begin\n",
    "\t\t\tintron_5bp = junc_spot-5\n",
    "\t\t\texon_5bp = junc_spot+4\n",
    "\n",
    "\t\t\t# # Disregard if segment doesn't have at least 5bp on each side\n",
    "\t\t\t# if intron_5bp < row['seg'].begin or exon_5bp >= row['seg'].end:\n",
    "\t\t\t\t# continue\n",
    "\t\t\t# Disregard if last 5bp of exon overlaps any introns\n",
    "\t\t\t# Disregard if first 5bp of intron overlaps any exons\n",
    "\t\t\t# This accounts for alternative splice sites\n",
    "\t\t\tif len(row['exons_f'].overlap(intron_5bp, junc_spot)) > 0:\n",
    "\t\t\t\tcontinue\n",
    "\t\t\tif len(row['introns_f'].overlap(junc_spot, exon_5bp+1)) > 0:\n",
    "\t\t\t\tcontinue\n",
    "\n",
    "\t\t\treturn 'pre-mRNA'\n",
    "\t\t\n",
    "\treturn 'Ambiguous'\n",
    "\n",
    "\n",
    "\n",
    "linear_reads['seg_class'] = linear_reads.apply(classify_seg, axis=1)\n",
    "print(linear_reads[['read_id', 'read', 'seg_class']].seg_class.value_counts())\n",
    "linear_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "read_class\n",
      "Ambiguous                               19182\n",
      "mRNA                                    13291\n",
      "Exonic                                  11810\n",
      "Intronic                                10594\n",
      "Intergenic                               1639\n",
      "(Intronic, Ambiguous)                     413\n",
      "(Intronic, Exonic)                        302\n",
      "(Ambiguous, Exonic)                       211\n",
      "(Exonic, Ambiguous)                       202\n",
      "(Intronic, Exon junction)                  77\n",
      "Starts at 5'ss                              8\n",
      "(Intronic, Exon junction, Ambiguous)        7\n",
      "Name: count, dtype: int64\n",
      "57736\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/zp/3k8jr14j3wgcym12kxf4tmfw0000gr/T/ipykernel_7089/4264127190.py:49: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  linear_reads['read_class'] = linear_reads.read_id.map(linear_reads.groupby('read_id').apply(classify_read))\n"
     ]
    }
   ],
   "source": [
    "def classify_read(seg_rows:pd.DataFrame) -> str:\n",
    "\tclasses_set = set(seg_rows['seg_class'])\n",
    "\tif len(classes_set)==1:\n",
    "\t\tif classes_set == {'Exon junction'}:\n",
    "\t\t\tif seg_rows.spliced.any():\n",
    "\t\t\t\treturn 'mRNA'\n",
    "\t\t\telse:\n",
    "\t\t\t\treturn 'Exonic'\n",
    "\t\telse:\n",
    "\t\t\treturn classes_set.pop()\n",
    "\t\t\n",
    "\t# Now we know it's 2 classes at minimum\n",
    "\t\n",
    "\tif 'Intergenic' in classes_set:\n",
    "\t\treturn 'Ambiguous'\n",
    "\t\n",
    "\tif classes_set in (\n",
    "\t\t\t\t\t{'Exon junction', 'Exonic',},\n",
    "\t\t\t\t\t{'Exon junction', 'Exonic', 'Intronic'},\n",
    "\t\t\t\t\t{'Exon junction', 'Ambiguous'},\n",
    "\t\t\t\t\t{'Exon junction', 'Exonic', 'Ambiguous'},\n",
    "\t\t\t\t\t{'Exon junction', 'Exonic', 'Intronic', 'Ambiguous'},\n",
    "\t\t\t\t\t):\n",
    "\t\treturn 'mRNA'\n",
    "\t\n",
    "\tif classes_set in (\n",
    "\t\t\t\t\t{'pre-mRNA', 'Exonic'},\n",
    "\t\t\t\t\t{'pre-mRNA', 'Intronic'},\n",
    "\t\t\t\t\t{'pre-mRNA', 'Exonic', 'Intronic'},\n",
    "\t\t\t\t\t{'pre-mRNA', 'Ambiguous'},\n",
    "\t\t\t\t\t{'pre-mRNA', 'Exonic', 'Ambiguous'},\n",
    "\t\t\t\t\t{'pre-mRNA', 'Intronic', 'Ambiguous'},\n",
    "\t\t\t\t\t{'pre-mRNA', 'Exonic', 'Intronic', 'Ambiguous'},\n",
    "\t\t\t\t\t):\n",
    "\t\treturn 'pre-mRNA'\n",
    "\t\n",
    "\tif classes_set in (\n",
    "\t\t\t\t\t{\"Starts at 5'ss\", 'Intronic'},\n",
    "\t\t\t\t\t{\"Starts at 5'ss\", 'Ambiguous'},\n",
    "\t\t\t\t\t{\"Starts at 5'ss\", 'Intronic', 'Ambiguous'},\n",
    "\t\t\t\t\t):\n",
    "\t\treturn \"Starts at 5'ss\"\n",
    "\t\n",
    "\treturn tuple(classes_set)\n",
    "\t# return 'Ambiguous'\n",
    "\n",
    "\n",
    "# linear_reads.seg_class = linear_reads.seg_class.replace(['Ambiguous m', 'Ambiguous f'], 'Ambiguous')\n",
    "linear_reads['read_class'] = linear_reads.read_id.map(linear_reads.groupby('read_id').apply(classify_read))\n",
    "print(linear_reads[['read_id', 'read_class']].drop_duplicates().read_class.value_counts())\n",
    "print(linear_reads.read_id.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "linear_reads.read_class = linear_reads.read_class.transform(lambda rc: 'Ambiguous' if isinstance(rc, tuple) else rc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spliced\n",
      "False    38653\n",
      "True     19083\n",
      "Name: count, dtype: int64\n",
      "read_class\n",
      "Ambiguous         20394\n",
      "mRNA              13291\n",
      "Exonic            11810\n",
      "Intronic          10594\n",
      "Intergenic         1639\n",
      "Starts at 5'ss        8\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>read_class</th>\n",
       "      <th>stage_reached</th>\n",
       "      <th>spliced</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>linear_map</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11992</td>\n",
       "      <td>Ambiguous</td>\n",
       "      <td>linear_map</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:12085</td>\n",
       "      <td>Ambiguous</td>\n",
       "      <td>linear_map</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:13244</td>\n",
       "      <td>Ambiguous</td>\n",
       "      <td>linear_map</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:13589</td>\n",
       "      <td>Ambiguous</td>\n",
       "      <td>linear_map</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57731</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:25723</td>\n",
       "      <td>Intronic</td>\n",
       "      <td>linear_map</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57732</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:26318</td>\n",
       "      <td>Ambiguous</td>\n",
       "      <td>linear_map</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57733</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:27352</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>linear_map</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57734</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>Ambiguous</td>\n",
       "      <td>linear_map</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57735</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>Ambiguous</td>\n",
       "      <td>linear_map</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>57736 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              read_id  read_class  \\\n",
       "0      NGSNJ-086:229:GW200110425th:1:1101:10004:11303        mRNA   \n",
       "1      NGSNJ-086:229:GW200110425th:1:1101:10004:11992   Ambiguous   \n",
       "2      NGSNJ-086:229:GW200110425th:1:1101:10004:12085   Ambiguous   \n",
       "3      NGSNJ-086:229:GW200110425th:1:1101:10004:13244   Ambiguous   \n",
       "4      NGSNJ-086:229:GW200110425th:1:1101:10004:13589   Ambiguous   \n",
       "...                                               ...         ...   \n",
       "57731   NGSNJ-086:229:GW200110425th:1:1101:9995:25723    Intronic   \n",
       "57732   NGSNJ-086:229:GW200110425th:1:1101:9995:26318   Ambiguous   \n",
       "57733   NGSNJ-086:229:GW200110425th:1:1101:9995:27352  Intergenic   \n",
       "57734    NGSNJ-086:229:GW200110425th:1:1101:9995:4961   Ambiguous   \n",
       "57735    NGSNJ-086:229:GW200110425th:1:1101:9995:9502   Ambiguous   \n",
       "\n",
       "      stage_reached  spliced  \n",
       "0        linear_map     True  \n",
       "1        linear_map    False  \n",
       "2        linear_map    False  \n",
       "3        linear_map    False  \n",
       "4        linear_map    False  \n",
       "...             ...      ...  \n",
       "57731    linear_map    False  \n",
       "57732    linear_map    False  \n",
       "57733    linear_map    False  \n",
       "57734    linear_map    False  \n",
       "57735    linear_map     True  \n",
       "\n",
       "[57736 rows x 4 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "linear_reads_sum = linear_reads.groupby(['read_id', 'read_class'], as_index=False).agg({'spliced': any})\n",
    "linear_reads_sum['stage_reached'] = 'linear_map'\n",
    "linear_reads_sum = linear_reads_sum[['read_id', 'read_class', 'stage_reached', 'spliced']]\n",
    "assert linear_reads_sum.read_id.is_unique\n",
    "print(linear_reads_sum.spliced.value_counts())\n",
    "print(linear_reads_sum.read_class.value_counts())\n",
    "linear_reads_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[18], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39misfile(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00moutput_base\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124mread_classes.tsv\u001b[39m\u001b[38;5;124m'\u001b[39m):\n\u001b[0;32m----> 2\u001b[0m \t\u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m()\n\u001b[1;32m      3\u001b[0m linear_reads_sum\u001b[38;5;241m.\u001b[39mto_csv(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00moutput_base\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124mread_classes.tsv\u001b[39m\u001b[38;5;124m'\u001b[39m, sep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m, index\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "\u001b[0;31mRuntimeError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "if os.path.isfile(f'{output_base}read_classes.tsv'):\n",
    "\traise RuntimeError()\n",
    "linear_reads_sum.to_csv(f'{output_base}read_classes.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "truth",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
