{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dataclasses import dataclass, field\n",
    "import os, sys, subprocess\n",
    "import copy\n",
    "import statistics\n",
    "import math\n",
    "import time\n",
    "import datetime as dt\n",
    "import itertools as it\n",
    "import random\n",
    "\n",
    "import scalene\n",
    "import requests\n",
    "import gzip\n",
    "import pysam\n",
    "from defusedxml import ElementTree\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "from sklearn import cluster, decomposition, linear_model, metrics, model_selection, preprocessing, tree, utils \n",
    "from statsmodels.api import qqplot\n",
    "import plotnine as p9\n",
    "from mizani.formatters import percent_format, comma_format\n",
    "from Bio import SeqIO, Seq, SeqFeature, SeqRecord, AlignIO, Align, Entrez, Graphics\n",
    "from Bio.Blast import NCBIWWW\n",
    "from intervaltree import Interval, IntervalTree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### Constants #####\n",
    "DNA_BASES = ('A', 'T', 'G', 'C')\n",
    "HUMAN_CHROMOSOMES = ('chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chrX', 'chrY', 'chrM')\n",
    "\n",
    "KEYS_FILE = '/home/tmooney/Misc/keys.txt'\n",
    "GENCODE_v19_GENOME_DIR = '/datasets2/genomes/fasta_files/hg19.gencode.v19.primary_genome_by_chromosome'\n",
    "GENCODE_v19_ANNOTATIONS_FILE = '/datasets2/genomes/annotations/hg19.gencode.v19.annotation.gtf'\n",
    "INTROPOLIS_HG19_FILE = '/datasets2/genomes/intropolis/intropolis.v1.hg19.tsv'\n",
    "INTROPOLIS_LIFTOVER_FILE = '/datasets2/genomes/intropolis/intropolis.v1.hg19_with_liftover_to_hg38.tsv'\n",
    "BOWTIE2_HG19_INDEX = '/datasets2/genomes/indices/bowtie2/hg38.fa'\n",
    "SPLICEAI_HG19_DIR = '/datasets2/genomes/SpliceAI/hg19.gencode.v19.primary_genome'\n",
    "GENCODE_v44_GENOME_FILE = '/datasets2/genomes/fasta_files/hg38.gencode.v44.primary_assembly_standard_chromosomes.fa'\n",
    "GENCODE_v44_BASIC_ANNOTATIONS_FILE = '/datasets2/genomes/annotations/hg38.gencode.v44.basic.annotation.gtf'\n",
    "GENCODE_v44_COMP_ANNOTATIONS_FILE = '/datasets2/genomes/annotations/hg38.gencode.v44.comprehensive.annotation.gtf'\n",
    "HGNC_ANNOTATIONS_FILE = '/datasets2/genomes/annotations/hgnc_complete_set_2023-07-01.tsv'\n",
    "BOWTIE2_HG38_INDEX = '/datasets2/genomes/indices/bowtie2/hg19.fa'\n",
    "SPLICEAI_HG38_DIR = '/datasets2/genomes/SpliceAI/hg38.gencode.v41.primary_genome'\n",
    "SALMON_HG38_INDEX = '/datasets2/genomes/indices/salmon/hg38.gencode.v44'\n",
    "\n",
    "WORKING_DIR = '/home/tmooney/Lariat_mapping'\n",
    "COLLECTED_DATA_DIR = WORKING_DIR + '/collected_data'\n",
    "FINAL_RESULTS_DIR = WORKING_DIR + '/final_results'\n",
    "LOG_DIR = WORKING_DIR + '/logs'\n",
    "OUTPUT_DIR = WORKING_DIR + '/output'\n",
    "REFERENCE_DIR = WORKING_DIR + '/reference_data'\n",
    "SEQUENCES_DIR = REFERENCE_DIR + '/sequences'\n",
    "ANNOTATIONS_DIR = REFERENCE_DIR + '/annotations'\n",
    "SCRIPTS_DIR= WORKING_DIR + '/scripts'\n",
    "VISUALS_DIR = WORKING_DIR + '/visuals'\n",
    "PLOTS_DIR = VISUALS_DIR + '/plots'\n",
    "\n",
    "# INTRONS_BED = '/home/tmooney/Lariat_mapping/testing/references/hg38.gencode.v44.comprehensive.introns_mod.bed'\n",
    "INTRONS_BED = '/home/tmooney/Lariat_mapping/testing/new_references/introns.bed'\n",
    "GTF_FILE = '/home/tmooney/Lariat_mapping/testing/new_references/annotation.gtf'\n",
    "NCRNA_TYPES = ('lncRNA',\n",
    "\t\t\t'Mt_rRNA',\n",
    "\t\t\t'Mt_tRNA',\n",
    "\t\t\t'miRNA',\n",
    "\t\t\t'misc_RNA',\n",
    "\t\t\t'rRNA',\n",
    "\t\t\t'scRNA',\n",
    "\t\t\t'snRNA',\n",
    "\t\t\t'snoRNA',\n",
    "\t\t\t'ribozyme',\n",
    "\t\t\t'sRNA',\n",
    "\t\t\t'scaRNA')\n",
    "\n",
    "IG_TcR_TYPES = ('IG_C_gene',\n",
    "\t\t\t\t'IG_D_gene',\n",
    "\t\t\t\t'IG_J_gene',\n",
    "\t\t\t\t'IG_LV_gene',\n",
    "\t\t\t\t'IG_V_gene',\n",
    "\t\t\t\t'TR_C_gene',\n",
    "\t\t\t\t'TR_J_gene',\n",
    "\t\t\t\t'TR_V_gene',\n",
    "\t\t\t\t'TR_D_gene')\n",
    "\n",
    "READ_SPECIES =  ('Lariat', 'Repeat region', 'Circular intron', 'Template switching', \"Unmapped, has a 5'ss alignment\", \"Unmapped\", \"Ends at a 5'ss\", 'pre-mRNA', 'mRNA', 'Intronic', 'Exonic', 'Genic, ambiguous', 'Inter-genic')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### Functions #####\n",
    "def tree_covers_interval(tree:IntervalTree, interval:Interval) -> bool:\n",
    "\ttotal_coverage = False\n",
    "\tmerged_tree = tree.copy()\n",
    "\tmerged_tree.merge_overlaps(strict=False)\n",
    "\tfor merged_interval in merged_tree:\n",
    "\t\tif merged_interval.contains_interval(interval):\n",
    "\t\t\ttotal_coverage = True\n",
    "\t\n",
    "\treturn total_coverage\n",
    "\n",
    "\n",
    "def process_fivep_sites(sites:str):\n",
    "\tout = []\n",
    "\tfor site in sites.split(','):\n",
    "\t\tchrom, start, end, strand = site.split(';')\n",
    "\t\tstart = int(start)\n",
    "\t\tend = int(end)\n",
    "\t\tout.append((chrom, start, end, strand))\n",
    "\n",
    "\treturn tuple(out)\n",
    "\n",
    "\n",
    "def parse_attributes(attribute_string:str) -> dict:\n",
    "\tattributes = attribute_string.rstrip('\";').split('; ')\n",
    "\tattributes = [attr.split(' ') for attr in attributes]\n",
    "\ttags = [attr_val.strip('\"') for attr_name, attr_val in attributes if attr_name=='tag']\n",
    "\tattributes = {attr_name: attr_val.strip('\"') for attr_name, attr_val in attributes if attr_name!='tag'}\n",
    "\tattributes['tags'] = tags\n",
    "\n",
    "\treturn attributes\n",
    "\n",
    "def parse_transcript_info(ref_gtf:str):\n",
    "\t'''\n",
    "\t\n",
    "\t'''\n",
    "\t# Count header lines in GTF\n",
    "\theader_lines = 0\n",
    "\twith open(ref_gtf) as r:\n",
    "\t\tfor line in r:\n",
    "\t\t\tif line.startswith('##'):\n",
    "\t\t\t\theader_lines += 1\n",
    "\t\t\telse:\n",
    "\t\t\t\tbreak\n",
    "\n",
    "\t# Load GTF\n",
    "\ttranscripts = pd.read_csv(ref_gtf, names=['chrom', 'source', 'feature', 'start', 'end', 'score', 'strand', 'phase', 'attributes'], skiprows=header_lines, sep='\\t')\n",
    "\ttranscripts = transcripts.loc[transcripts.feature=='transcript'].reset_index(drop=True)\n",
    "\t\n",
    "\t# Pull out transcript x gene info\n",
    "\ttranscripts.attributes = transcripts.attributes.transform(parse_attributes)\n",
    "\ttranscripts['transcript_id'] = transcripts.attributes.transform(lambda attributes: attributes['transcript_id'])\n",
    "\ttranscripts['gene_id'] = transcripts.attributes.transform(lambda attributes: attributes['gene_id'])\n",
    "\ttranscripts['gene_name'] = transcripts.attributes.transform(lambda attributes: attributes['gene_name'])\n",
    "\ttranscripts['gene_type'] = transcripts.attributes.transform(lambda attributes: attributes['gene_type'])\n",
    "\n",
    "\treturn transcripts[['transcript_id', 'gene_id', 'gene_name', 'gene_type']]\n",
    "\n",
    "\n",
    "def parse_feature_name(name:str, transcripts):\n",
    "\tfeat, chrom, start, end, strand, annotations = name.split(';')\n",
    "\n",
    "\tstart = int(start)\n",
    "\tend = int(end)\n",
    "\tannotations = [transcript_and_num.split('-') for transcript_and_num in annotations.split('|')]\n",
    "\tannotations = tuple([ (transcript_id, int(num)) for transcript_id, num in annotations ])\n",
    "\tgene_types = tuple(set([transcripts[transcript_id][2] for transcript_id, num in annotations]))\n",
    "\t\n",
    "\treturn (feat, chrom, start, end, strand, annotations, gene_types)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C22-1 R1_100 C22-1|R1_100 /home/tmooney/Lariat_mapping/output/100k_truncs/C22-1_R1_100k_lariat_mapping\n"
     ]
    }
   ],
   "source": [
    "##### Settings #####\n",
    "pd.set_option('display.max_columns', 100)\n",
    "pd.set_option('display.max_rows', 100)\n",
    "\n",
    "p9.theme_set(p9.theme_classic)\n",
    "\n",
    "test_dir = '/home/tmooney/Lariat_mapping/output/100k_truncs'\n",
    "sources = (('C22-1', 'R1_100'), ('C22-1', 'R2_100'), ('C22-3', 'R1_100'), ('C22-3', 'R2_100'), ('C22-5', 'R1_100'), ('C22-5', 'R2_100'), \n",
    "\t\t('HEK293T-1', 'R1_100'), ('HEK293T-1', 'R2_100'), ('HEK293T-4', 'R1_100'), ('HEK293T-4', 'R2_100'), ('HEK293T-5', 'R1_100'), ('HEK293T-5', 'R2_100'))\n",
    "sources = [(sample, r_dir, f'{sample}|{r_dir}', f'{test_dir}/{sample}_{r_dir}k_lariat_mapping') for sample, r_dir in sources]\n",
    "\n",
    "samp, r_dir, source, out_dir = sources[0]\n",
    "input_fastq = f'/home/tmooney/Lariat_mapping/testing/input_data/{samp}_{r_dir}k.fastq.gz'\n",
    "print(samp, r_dir, source, out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>forward</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>blocks</th>\n",
       "      <th>cigar</th>\n",
       "      <th>tags</th>\n",
       "      <th>is_paired</th>\n",
       "      <th>mate_is_mapped</th>\n",
       "      <th>read</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:1868:1000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>False</td>\n",
       "      <td>76672340</td>\n",
       "      <td>76672490</td>\n",
       "      <td>[(76672340, 76672490)]</td>\n",
       "      <td>150M</td>\n",
       "      <td>[(AS, -10), (XN, 0), (XM, 2), (XO, 0), (XG, 0)...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:3857:1000</td>\n",
       "      <td>chr3</td>\n",
       "      <td>False</td>\n",
       "      <td>15780069</td>\n",
       "      <td>15780219</td>\n",
       "      <td>[(15780069, 15780219)]</td>\n",
       "      <td>150M</td>\n",
       "      <td>[(AS, -30), (XN, 0), (XM, 6), (XO, 0), (XG, 0)...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:5430:1000</td>\n",
       "      <td>chr3</td>\n",
       "      <td>True</td>\n",
       "      <td>15780035</td>\n",
       "      <td>15780185</td>\n",
       "      <td>[(15780035, 15780185)]</td>\n",
       "      <td>150M</td>\n",
       "      <td>[(AS, -15), (XN, 0), (XM, 3), (XO, 0), (XG, 0)...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:11632:1000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>True</td>\n",
       "      <td>109642834</td>\n",
       "      <td>109642984</td>\n",
       "      <td>[(109642834, 109642984)]</td>\n",
       "      <td>150M</td>\n",
       "      <td>[(AS, -6), (XN, 0), (XM, 2), (XO, 0), (XG, 0),...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:11668:1000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>False</td>\n",
       "      <td>183083780</td>\n",
       "      <td>183084730</td>\n",
       "      <td>[(183083780, 183083854), (183084654, 183084730)]</td>\n",
       "      <td>74M800N76M</td>\n",
       "      <td>[(AS, 0), (XN, 0), (XM, 0), (XO, 0), (XG, 0), ...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38497</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:14570:29136</td>\n",
       "      <td>chr1</td>\n",
       "      <td>True</td>\n",
       "      <td>156287269</td>\n",
       "      <td>156288739</td>\n",
       "      <td>[(156287269, 156287338), (156288658, 156288739)]</td>\n",
       "      <td>69M1320N81M</td>\n",
       "      <td>[(AS, 0), (XN, 0), (XM, 0), (XO, 0), (XG, 0), ...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38498</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:14913:29136</td>\n",
       "      <td>chr3</td>\n",
       "      <td>False</td>\n",
       "      <td>127771281</td>\n",
       "      <td>127771706</td>\n",
       "      <td>[(127771281, 127771402), (127771677, 127771706)]</td>\n",
       "      <td>121M275N29M</td>\n",
       "      <td>[(AS, -30), (XN, 0), (XM, 6), (XO, 0), (XG, 0)...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38499</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:14950:29136</td>\n",
       "      <td>chr1</td>\n",
       "      <td>True</td>\n",
       "      <td>200833115</td>\n",
       "      <td>200833265</td>\n",
       "      <td>[(200833115, 200833265)]</td>\n",
       "      <td>150M</td>\n",
       "      <td>[(AS, -30), (XN, 0), (XM, 6), (XO, 0), (XG, 0)...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38500</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:16559:29136</td>\n",
       "      <td>chr3</td>\n",
       "      <td>False</td>\n",
       "      <td>15780034</td>\n",
       "      <td>15780179</td>\n",
       "      <td>[(15780034, 15780042), (15780042, 15780179)]</td>\n",
       "      <td>8M5I137M</td>\n",
       "      <td>[(AS, -35), (XN, 0), (XM, 3), (XO, 1), (XG, 5)...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38501</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:17716:29136</td>\n",
       "      <td>chr3</td>\n",
       "      <td>True</td>\n",
       "      <td>15742172</td>\n",
       "      <td>15780209</td>\n",
       "      <td>[(15742172, 15742178), (15780065, 15780209)]</td>\n",
       "      <td>6M37887N144M</td>\n",
       "      <td>[(AS, -15), (ZS, -15), (XN, 0), (XM, 3), (XO, ...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38502 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              read_id chrom  forward  \\\n",
       "0        NGSNJ-086:229:GW200110425th:1:1101:1868:1000  chr2    False   \n",
       "1        NGSNJ-086:229:GW200110425th:1:1101:3857:1000  chr3    False   \n",
       "2        NGSNJ-086:229:GW200110425th:1:1101:5430:1000  chr3     True   \n",
       "3       NGSNJ-086:229:GW200110425th:1:1101:11632:1000  chr1     True   \n",
       "4       NGSNJ-086:229:GW200110425th:1:1101:11668:1000  chr1    False   \n",
       "...                                               ...   ...      ...   \n",
       "38497  NGSNJ-086:229:GW200110425th:1:1101:14570:29136  chr1     True   \n",
       "38498  NGSNJ-086:229:GW200110425th:1:1101:14913:29136  chr3    False   \n",
       "38499  NGSNJ-086:229:GW200110425th:1:1101:14950:29136  chr1     True   \n",
       "38500  NGSNJ-086:229:GW200110425th:1:1101:16559:29136  chr3    False   \n",
       "38501  NGSNJ-086:229:GW200110425th:1:1101:17716:29136  chr3     True   \n",
       "\n",
       "           start        end                                            blocks  \\\n",
       "0       76672340   76672490                            [(76672340, 76672490)]   \n",
       "1       15780069   15780219                            [(15780069, 15780219)]   \n",
       "2       15780035   15780185                            [(15780035, 15780185)]   \n",
       "3      109642834  109642984                          [(109642834, 109642984)]   \n",
       "4      183083780  183084730  [(183083780, 183083854), (183084654, 183084730)]   \n",
       "...          ...        ...                                               ...   \n",
       "38497  156287269  156288739  [(156287269, 156287338), (156288658, 156288739)]   \n",
       "38498  127771281  127771706  [(127771281, 127771402), (127771677, 127771706)]   \n",
       "38499  200833115  200833265                          [(200833115, 200833265)]   \n",
       "38500   15780034   15780179      [(15780034, 15780042), (15780042, 15780179)]   \n",
       "38501   15742172   15780209      [(15742172, 15742178), (15780065, 15780209)]   \n",
       "\n",
       "              cigar                                               tags  \\\n",
       "0              150M  [(AS, -10), (XN, 0), (XM, 2), (XO, 0), (XG, 0)...   \n",
       "1              150M  [(AS, -30), (XN, 0), (XM, 6), (XO, 0), (XG, 0)...   \n",
       "2              150M  [(AS, -15), (XN, 0), (XM, 3), (XO, 0), (XG, 0)...   \n",
       "3              150M  [(AS, -6), (XN, 0), (XM, 2), (XO, 0), (XG, 0),...   \n",
       "4        74M800N76M  [(AS, 0), (XN, 0), (XM, 0), (XO, 0), (XG, 0), ...   \n",
       "...             ...                                                ...   \n",
       "38497   69M1320N81M  [(AS, 0), (XN, 0), (XM, 0), (XO, 0), (XG, 0), ...   \n",
       "38498   121M275N29M  [(AS, -30), (XN, 0), (XM, 6), (XO, 0), (XG, 0)...   \n",
       "38499          150M  [(AS, -30), (XN, 0), (XM, 6), (XO, 0), (XG, 0)...   \n",
       "38500      8M5I137M  [(AS, -35), (XN, 0), (XM, 3), (XO, 1), (XG, 5)...   \n",
       "38501  6M37887N144M  [(AS, -15), (ZS, -15), (XN, 0), (XM, 3), (XO, ...   \n",
       "\n",
       "       is_paired  mate_is_mapped   read  \n",
       "0           True           False  False  \n",
       "1           True           False  False  \n",
       "2           True           False  False  \n",
       "3           True           False  False  \n",
       "4           True           False  False  \n",
       "...          ...             ...    ...  \n",
       "38497       True            True  False  \n",
       "38498       True           False  False  \n",
       "38499       True           False   True  \n",
       "38500       True           False  False  \n",
       "38501       True           False   True  \n",
       "\n",
       "[38502 rows x 11 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LINEAR_COLUMNS = ('read_id', 'chrom', 'forward', 'start', 'end', 'blocks', 'cigar', 'tags', 'is_paired', 'mate_is_mapped', 'read',)\n",
    "linear_reads = []\n",
    "for align in pysam.AlignmentFile('/Users/trumanmooney/Library/CloudStorage/OneDrive-BrownUniversity/Documents/Lariat_mapping/output/100k_truncs/mapped_reads.bam', 'rb'):\n",
    "\tif align.is_unmapped:\n",
    "\t\tcontinue\n",
    "\tlinear_reads.append([align.query_name, \n",
    "\t\t\t\t\talign.reference_name, \n",
    "\t\t\t\t\talign.is_forward,\n",
    "\t\t\t\t\talign.reference_start, \n",
    "\t\t\t\t\talign.reference_end, \n",
    "\t\t\t\t\talign.get_blocks(),\n",
    "\t\t\t\t\talign.cigarstring,\n",
    "\t\t\t\t\talign.get_tags(),\n",
    "\t\t\t\t\talign.is_paired,\n",
    "\t\t\t\t\talign.mate_is_mapped,\n",
    "\t\t\t\t\talign.is_read1,\n",
    "\t\t\t\t\t])\n",
    "\n",
    "linear_reads = pd.DataFrame(linear_reads, columns=LINEAR_COLUMNS)\n",
    "linear_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>forward</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>blocks</th>\n",
       "      <th>cigar</th>\n",
       "      <th>tags</th>\n",
       "      <th>is_paired</th>\n",
       "      <th>mate_is_mapped</th>\n",
       "      <th>read</th>\n",
       "      <th>pa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15367</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11992</td>\n",
       "      <td>chr3</td>\n",
       "      <td>True</td>\n",
       "      <td>15773247</td>\n",
       "      <td>15780208</td>\n",
       "      <td>[(15773247, 15773254), (15780065, 15780208)]</td>\n",
       "      <td>7M6811N143M</td>\n",
       "      <td>{'AS': -15, 'ZS': -15, 'XN': 0, 'XM': 3, 'XO':...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15368</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11992</td>\n",
       "      <td>chr3</td>\n",
       "      <td>False</td>\n",
       "      <td>15773258</td>\n",
       "      <td>15780219</td>\n",
       "      <td>[(15773258, 15773260), (15780071, 15780219)]</td>\n",
       "      <td>2M6811N148M</td>\n",
       "      <td>{'AS': -15, 'ZS': -15, 'XN': 0, 'XM': 3, 'XO':...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>CP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15498</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:12085</td>\n",
       "      <td>chr3</td>\n",
       "      <td>True</td>\n",
       "      <td>15780037</td>\n",
       "      <td>15780187</td>\n",
       "      <td>[(15780037, 15780187)]</td>\n",
       "      <td>150M</td>\n",
       "      <td>{'AS': 0, 'XN': 0, 'XM': 0, 'XO': 0, 'XG': 0, ...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>UP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17515</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:13589</td>\n",
       "      <td>chr3</td>\n",
       "      <td>False</td>\n",
       "      <td>141595425</td>\n",
       "      <td>141595575</td>\n",
       "      <td>[(141595425, 141595575)]</td>\n",
       "      <td>150M</td>\n",
       "      <td>{'AS': -4, 'XN': 0, 'XM': 1, 'XO': 0, 'XG': 0,...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>DP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17514</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:13589</td>\n",
       "      <td>chr3</td>\n",
       "      <td>True</td>\n",
       "      <td>141595427</td>\n",
       "      <td>141595577</td>\n",
       "      <td>[(141595427, 141595577)]</td>\n",
       "      <td>150M</td>\n",
       "      <td>{'AS': -8, 'XN': 0, 'XM': 2, 'XO': 0, 'XG': 0,...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>DP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31072</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:23719</td>\n",
       "      <td>chr3</td>\n",
       "      <td>True</td>\n",
       "      <td>15780037</td>\n",
       "      <td>15780187</td>\n",
       "      <td>[(15780037, 15780187)]</td>\n",
       "      <td>150M</td>\n",
       "      <td>{'AS': 0, 'XN': 0, 'XM': 0, 'XO': 0, 'XG': 0, ...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>CP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31071</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:23719</td>\n",
       "      <td>chr3</td>\n",
       "      <td>False</td>\n",
       "      <td>15780073</td>\n",
       "      <td>15780223</td>\n",
       "      <td>[(15780073, 15780223)]</td>\n",
       "      <td>150M</td>\n",
       "      <td>{'AS': 0, 'XN': 0, 'XM': 0, 'XO': 0, 'XG': 0, ...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34666</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:26318</td>\n",
       "      <td>chr1</td>\n",
       "      <td>True</td>\n",
       "      <td>167905021</td>\n",
       "      <td>167905171</td>\n",
       "      <td>[(167905021, 167905171)]</td>\n",
       "      <td>150M</td>\n",
       "      <td>{'AS': 0, 'XN': 0, 'XM': 0, 'XO': 0, 'XG': 0, ...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34667</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:26318</td>\n",
       "      <td>chr1</td>\n",
       "      <td>False</td>\n",
       "      <td>167905021</td>\n",
       "      <td>167905171</td>\n",
       "      <td>[(167905021, 167905171)]</td>\n",
       "      <td>150M</td>\n",
       "      <td>{'AS': 0, 'XN': 0, 'XM': 0, 'XO': 0, 'XG': 0, ...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>CP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11853</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr3</td>\n",
       "      <td>True</td>\n",
       "      <td>15780040</td>\n",
       "      <td>15780190</td>\n",
       "      <td>[(15780040, 15780190)]</td>\n",
       "      <td>150M</td>\n",
       "      <td>{'AS': -14, 'ZS': -19, 'XN': 0, 'XM': 3, 'XO':...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>UP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38502 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              read_id chrom  forward  \\\n",
       "15367  NGSNJ-086:229:GW200110425th:1:1101:10004:11992  chr3     True   \n",
       "15368  NGSNJ-086:229:GW200110425th:1:1101:10004:11992  chr3    False   \n",
       "15498  NGSNJ-086:229:GW200110425th:1:1101:10004:12085  chr3     True   \n",
       "17515  NGSNJ-086:229:GW200110425th:1:1101:10004:13589  chr3    False   \n",
       "17514  NGSNJ-086:229:GW200110425th:1:1101:10004:13589  chr3     True   \n",
       "...                                               ...   ...      ...   \n",
       "31072   NGSNJ-086:229:GW200110425th:1:1101:9995:23719  chr3     True   \n",
       "31071   NGSNJ-086:229:GW200110425th:1:1101:9995:23719  chr3    False   \n",
       "34666   NGSNJ-086:229:GW200110425th:1:1101:9995:26318  chr1     True   \n",
       "34667   NGSNJ-086:229:GW200110425th:1:1101:9995:26318  chr1    False   \n",
       "11853    NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr3     True   \n",
       "\n",
       "           start        end                                        blocks  \\\n",
       "15367   15773247   15780208  [(15773247, 15773254), (15780065, 15780208)]   \n",
       "15368   15773258   15780219  [(15773258, 15773260), (15780071, 15780219)]   \n",
       "15498   15780037   15780187                        [(15780037, 15780187)]   \n",
       "17515  141595425  141595575                      [(141595425, 141595575)]   \n",
       "17514  141595427  141595577                      [(141595427, 141595577)]   \n",
       "...          ...        ...                                           ...   \n",
       "31072   15780037   15780187                        [(15780037, 15780187)]   \n",
       "31071   15780073   15780223                        [(15780073, 15780223)]   \n",
       "34666  167905021  167905171                      [(167905021, 167905171)]   \n",
       "34667  167905021  167905171                      [(167905021, 167905171)]   \n",
       "11853   15780040   15780190                        [(15780040, 15780190)]   \n",
       "\n",
       "             cigar                                               tags  \\\n",
       "15367  7M6811N143M  {'AS': -15, 'ZS': -15, 'XN': 0, 'XM': 3, 'XO':...   \n",
       "15368  2M6811N148M  {'AS': -15, 'ZS': -15, 'XN': 0, 'XM': 3, 'XO':...   \n",
       "15498         150M  {'AS': 0, 'XN': 0, 'XM': 0, 'XO': 0, 'XG': 0, ...   \n",
       "17515         150M  {'AS': -4, 'XN': 0, 'XM': 1, 'XO': 0, 'XG': 0,...   \n",
       "17514         150M  {'AS': -8, 'XN': 0, 'XM': 2, 'XO': 0, 'XG': 0,...   \n",
       "...            ...                                                ...   \n",
       "31072         150M  {'AS': 0, 'XN': 0, 'XM': 0, 'XO': 0, 'XG': 0, ...   \n",
       "31071         150M  {'AS': 0, 'XN': 0, 'XM': 0, 'XO': 0, 'XG': 0, ...   \n",
       "34666         150M  {'AS': 0, 'XN': 0, 'XM': 0, 'XO': 0, 'XG': 0, ...   \n",
       "34667         150M  {'AS': 0, 'XN': 0, 'XM': 0, 'XO': 0, 'XG': 0, ...   \n",
       "11853         150M  {'AS': -14, 'ZS': -19, 'XN': 0, 'XM': 3, 'XO':...   \n",
       "\n",
       "       is_paired  mate_is_mapped   read  pa  \n",
       "15367       True            True   True  CP  \n",
       "15368       True            True  False  CP  \n",
       "15498       True           False  False  UP  \n",
       "17515       True            True  False  DP  \n",
       "17514       True            True   True  DP  \n",
       "...          ...             ...    ...  ..  \n",
       "31072       True            True  False  CP  \n",
       "31071       True            True   True  CP  \n",
       "34666       True            True   True  CP  \n",
       "34667       True            True  False  CP  \n",
       "11853       True           False  False  UP  \n",
       "\n",
       "[38502 rows x 12 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "linear_reads.sort_values('read_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "read\n",
      "True     19379\n",
      "False    19123\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(linear_reads.read.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "linear_reads.tags = linear_reads.tags.apply(lambda tags: {tag: val for tag, val in tags})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pa\n",
       "CP    26432\n",
       "UP     9044\n",
       "DP     3026\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "linear_reads['pa'] = linear_reads.tags.transform(lambda tags: tags['YT'])\n",
    "# print(linear_reads.read_id.nunique())\n",
    "# print(linear_reads.read_id_r.nunique())\n",
    "linear_reads.pa.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mate_is_mapped  pa\n",
       "False           UP     8066\n",
       "True            CP    26432\n",
       "                DP     3026\n",
       "                UP      978\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "linear_reads[['mate_is_mapped', 'pa']].value_counts().sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23284\n",
      "15218\n",
      "is_paired\n",
      "True    38502\n",
      "Name: count, dtype: int64\n",
      "mate_is_mapped  mate_is_forward\n",
      "False           True                8066\n",
      "True            False              15212\n",
      "                True               15224\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(linear_reads.read_id.nunique())\n",
    "print(linear_reads.loc[linear_reads.mate_is_mapped].read_id.nunique())\n",
    "print(linear_reads.is_paired.value_counts())\n",
    "print(linear_reads[['mate_is_mapped', 'mate_is_forward']].value_counts().sort_index())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENST00000456328.2': ('ENSG00000290825.1', 'DDX11L2', 'lncRNA'),\n",
       " 'ENST00000450305.2': ('ENSG00000223972.6',\n",
       "  'DDX11L1',\n",
       "  'transcribed_unprocessed_pseudogene'),\n",
       " 'ENST00000488147.1': ('ENSG00000227232.5',\n",
       "  'WASH7P',\n",
       "  'unprocessed_pseudogene'),\n",
       " 'ENST00000619216.1': ('ENSG00000278267.1', 'MIR6859-1', 'miRNA'),\n",
       " 'ENST00000473358.1': ('ENSG00000243485.5', 'MIR1302-2HG', 'lncRNA'),\n",
       " 'ENST00000469289.1': ('ENSG00000243485.5', 'MIR1302-2HG', 'lncRNA'),\n",
       " 'ENST00000607096.1': ('ENSG00000284332.1', 'MIR1302-2', 'miRNA'),\n",
       " 'ENST00000417324.1': ('ENSG00000237613.2', 'FAM138A', 'lncRNA'),\n",
       " 'ENST00000461467.1': ('ENSG00000237613.2', 'FAM138A', 'lncRNA'),\n",
       " 'ENST00000606857.1': ('ENSG00000268020.3',\n",
       "  'OR4G4P',\n",
       "  'unprocessed_pseudogene'),\n",
       " 'ENST00000642116.1': ('ENSG00000290826.1', 'ENSG00000290826', 'lncRNA'),\n",
       " 'ENST00000492842.2': ('ENSG00000240361.3',\n",
       "  'OR4G11P',\n",
       "  'transcribed_unprocessed_pseudogene'),\n",
       " 'ENST00000641515.2': ('ENSG00000186092.7', 'OR4F5', 'protein_coding'),\n",
       " 'ENST00000466430.5': ('ENSG00000238009.6', 'ENSG00000238009', 'lncRNA'),\n",
       " 'ENST00000477740.5': ('ENSG00000238009.6', 'ENSG00000238009', 'lncRNA'),\n",
       " 'ENST00000471248.1': ('ENSG00000238009.6', 'ENSG00000238009', 'lncRNA'),\n",
       " 'ENST00000610542.1': ('ENSG00000238009.6', 'ENSG00000238009', 'lncRNA'),\n",
       " 'ENST00000453576.2': ('ENSG00000238009.6', 'ENSG00000238009', 'lncRNA'),\n",
       " 'ENST00000495576.1': ('ENSG00000239945.1', 'ENSG00000239945', 'lncRNA'),\n",
       " 'ENST00000442987.3': ('ENSG00000233750.3', 'CICP27', 'processed_pseudogene'),\n",
       " 'ENST00000494149.2': ('ENSG00000268903.1',\n",
       "  'ENSG00000268903',\n",
       "  'processed_pseudogene'),\n",
       " 'ENST00000595919.1': ('ENSG00000269981.1',\n",
       "  'ENSG00000269981',\n",
       "  'processed_pseudogene'),\n",
       " 'ENST00000493797.1': ('ENSG00000239906.1', 'ENSG00000239906', 'lncRNA'),\n",
       " 'ENST00000484859.1': ('ENSG00000241860.7', 'ENSG00000241860', 'lncRNA'),\n",
       " 'ENST00000490997.5': ('ENSG00000241860.7', 'ENSG00000241860', 'lncRNA'),\n",
       " 'ENST00000466557.6': ('ENSG00000241860.7', 'ENSG00000241860', 'lncRNA'),\n",
       " 'ENST00000662089.1': ('ENSG00000241860.7', 'ENSG00000241860', 'lncRNA'),\n",
       " 'ENST00000491962.1': ('ENSG00000241860.7', 'ENSG00000241860', 'lncRNA'),\n",
       " 'ENST00000655252.1': ('ENSG00000241860.7', 'ENSG00000241860', 'lncRNA'),\n",
       " 'ENST00000410691.1': ('ENSG00000222623.1', 'RNU6-1100P', 'snRNA'),\n",
       " 'ENST00000496488.1': ('ENSG00000241599.1', 'ENSG00000241599', 'lncRNA'),\n",
       " 'ENST00000624431.2': ('ENSG00000279928.2',\n",
       "  'DDX11L17',\n",
       "  'unprocessed_pseudogene'),\n",
       " 'ENST00000623083.4': ('ENSG00000279457.4',\n",
       "  'WASH9P',\n",
       "  'unprocessed_pseudogene'),\n",
       " 'ENST00000612080.1': ('ENSG00000273874.1', 'MIR6859-2', 'miRNA'),\n",
       " 'ENST00000442116.1': ('ENSG00000228463.10',\n",
       "  'ENSG00000228463',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000448958.2': ('ENSG00000228463.10',\n",
       "  'ENSG00000228463',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000441866.2': ('ENSG00000228463.10',\n",
       "  'ENSG00000228463',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000634344.2': ('ENSG00000228463.10',\n",
       "  'ENSG00000228463',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000450734.1': ('ENSG00000228463.10',\n",
       "  'ENSG00000228463',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000424587.7': ('ENSG00000228463.10',\n",
       "  'ENSG00000228463',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000669836.1': ('ENSG00000286448.1', 'ENSG00000286448', 'lncRNA'),\n",
       " 'ENST00000458203.2': ('ENSG00000236679.2',\n",
       "  'RPL23AP24',\n",
       "  'processed_pseudogene'),\n",
       " 'ENST00000450983.1': ('ENSG00000236601.2', 'ENSG00000236601', 'lncRNA'),\n",
       " 'ENST00000412666.1': ('ENSG00000236601.2', 'ENSG00000236601', 'lncRNA'),\n",
       " 'ENST00000635159.1': ('ENSG00000236601.2', 'ENSG00000236601', 'lncRNA'),\n",
       " 'ENST00000431321.2': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000453935.1': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000440163.1': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000455207.5': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000455464.7': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000601814.5': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000431812.1': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000419160.4': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000440038.7': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000642074.1': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000423728.6': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000616311.5': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000599771.6': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000641845.1': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000641916.1': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000601486.5': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000641579.1': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000616947.2': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000432964.1': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000608420.1': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000641303.1': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000641063.1': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000641049.1': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000642124.1': ('ENSG00000290385.1', 'ENSG00000290385', 'lncRNA'),\n",
       " 'ENST00000437905.2': ('ENSG00000269732.1', 'WBP1LP7', 'processed_pseudogene'),\n",
       " 'ENST00000426406.4': ('ENSG00000284733.2', 'OR4F29', 'protein_coding'),\n",
       " 'ENST00000445840.2': ('ENSG00000237094.13',\n",
       "  'ENSG00000237094',\n",
       "  'transcribed_unprocessed_pseudogene'),\n",
       " 'ENST00000432723.3': ('ENSG00000233653.3', 'CICP7', 'processed_pseudogene'),\n",
       " 'ENST00000514436.1': ('ENSG00000250575.1',\n",
       "  'ENSG00000250575',\n",
       "  'unprocessed_pseudogene'),\n",
       " 'ENST00000614007.1': ('ENSG00000278757.1', 'U6', 'snRNA'),\n",
       " 'ENST00000634833.2': ('ENSG00000230021.10',\n",
       "  'ENSG00000230021',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000417636.2': ('ENSG00000230021.10',\n",
       "  'ENSG00000230021',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000419394.2': ('ENSG00000230021.10',\n",
       "  'ENSG00000230021',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000440196.3': ('ENSG00000230021.10',\n",
       "  'ENSG00000230021',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000641296.1': ('ENSG00000230021.10',\n",
       "  'ENSG00000230021',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000357876.6': ('ENSG00000230021.10',\n",
       "  'ENSG00000230021',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000635509.2': ('ENSG00000230021.10',\n",
       "  'ENSG00000230021',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000634337.2': ('ENSG00000230021.10',\n",
       "  'ENSG00000230021',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000648019.1': ('ENSG00000230021.10',\n",
       "  'ENSG00000230021',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000440200.5': ('ENSG00000230021.10',\n",
       "  'ENSG00000230021',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000452176.2': ('ENSG00000230021.10',\n",
       "  'ENSG00000230021',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000441245.5': ('ENSG00000230021.10',\n",
       "  'ENSG00000230021',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000414688.6': ('ENSG00000230021.10',\n",
       "  'ENSG00000230021',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000636676.1': ('ENSG00000230021.10',\n",
       "  'ENSG00000230021',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000447954.2': ('ENSG00000230021.10',\n",
       "  'ENSG00000230021',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000423796.1': ('ENSG00000235146.2', 'ENSG00000235146', 'lncRNA'),\n",
       " 'ENST00000450696.1': ('ENSG00000235146.2', 'ENSG00000235146', 'lncRNA'),\n",
       " 'ENST00000416931.1': ('ENSG00000225972.1',\n",
       "  'MTND1P23',\n",
       "  'unprocessed_pseudogene'),\n",
       " 'ENST00000457540.1': ('ENSG00000225630.1',\n",
       "  'MTND2P28',\n",
       "  'unprocessed_pseudogene'),\n",
       " 'ENST00000414273.1': ('ENSG00000237973.1',\n",
       "  'MTCO1P12',\n",
       "  'unprocessed_pseudogene'),\n",
       " 'ENST00000621981.1': ('ENSG00000278791.1', 'ENSG00000278791', 'miRNA'),\n",
       " 'ENST00000427426.1': ('ENSG00000229344.1',\n",
       "  'MTCO2P12',\n",
       "  'unprocessed_pseudogene'),\n",
       " 'ENST00000467115.1': ('ENSG00000240409.1',\n",
       "  'MTATP8P1',\n",
       "  'unprocessed_pseudogene'),\n",
       " 'ENST00000514057.1': ('ENSG00000248527.1',\n",
       "  'MTATP6P1',\n",
       "  'unprocessed_pseudogene'),\n",
       " 'ENST00000416718.2': ('ENSG00000198744.5',\n",
       "  'MTCO3P12',\n",
       "  'unprocessed_pseudogene'),\n",
       " 'ENST00000438434.2': ('ENSG00000268663.1', 'WBP1LP6', 'processed_pseudogene'),\n",
       " 'ENST00000332831.5': ('ENSG00000284662.2', 'OR4F16', 'protein_coding'),\n",
       " 'ENST00000440782.3': ('ENSG00000229376.3', 'CICP3', 'processed_pseudogene'),\n",
       " 'ENST00000506640.3': ('ENSG00000291215.1', 'ENSG00000291215', 'lncRNA'),\n",
       " 'ENST00000692781.2': ('ENSG00000291215.1', 'ENSG00000291215', 'lncRNA'),\n",
       " 'ENST00000411249.1': ('ENSG00000223181.1', 'RNU6-1199P', 'snRNA'),\n",
       " 'ENST00000422528.1': ('ENSG00000229905.1', 'ENSG00000229905', 'lncRNA'),\n",
       " 'ENST00000428504.2': ('ENSG00000228327.6',\n",
       "  'ENSG00000228327',\n",
       "  'transcribed_unprocessed_pseudogene'),\n",
       " 'ENST00000655765.1': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000412115.2': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000670700.1': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000429505.6': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000657896.1': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000434264.6': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000457084.1': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000658648.1': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000656571.1': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000665867.1': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000669749.1': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000666217.1': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000665719.1': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000651411.1': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000589899.5': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000609830.1': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000443772.2': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000585826.1': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000592547.1': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000586928.5': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000585745.5': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000585768.5': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000590848.5': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000591440.5': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000587530.5': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000593022.5': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000589531.5': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000588951.5': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000586288.1': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000591702.1': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000655384.1': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000667728.1': ('ENSG00000237491.10', 'LINC01409', 'lncRNA'),\n",
       " 'ENST00000644482.1': ('ENSG00000285268.1',\n",
       "  'ENSG00000285268',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000447500.4': ('ENSG00000290784.1', 'ENSG00000290784', 'lncRNA'),\n",
       " 'ENST00000590817.3': ('ENSG00000230092.8',\n",
       "  'ENSG00000230092',\n",
       "  'transcribed_unprocessed_pseudogene'),\n",
       " 'ENST00000326734.2': ('ENSG00000177757.2', 'FAM87B', 'lncRNA'),\n",
       " 'ENST00000473798.2': ('ENSG00000225880.6', 'LINC00115', 'lncRNA'),\n",
       " 'ENST00000624927.3': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000623808.4': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000685334.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000445118.7': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000669922.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000670780.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000610067.7': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000688309.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000659124.2': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000667414.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000658846.2': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000666741.2': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000688008.2': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000685764.2': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000685879.2': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000623070.5': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000608189.6': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000448975.6': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000449005.7': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000686238.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000661237.3': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000685566.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000441765.6': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000416570.6': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000701598.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000609139.7': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000609009.6': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000701768.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000657837.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000691316.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000692194.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000702273.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000702098.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000702557.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000668541.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000702847.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000691234.2': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000657175.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000660797.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000622921.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000670824.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000415295.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000688420.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000425657.1': ('ENSG00000228794.12', 'LINC01128', 'lncRNA'),\n",
       " 'ENST00000671208.1': ('ENSG00000288531.1', 'ENSG00000288531', 'lncRNA'),\n",
       " 'ENST00000661463.1': ('ENSG00000288531.1', 'ENSG00000288531', 'lncRNA'),\n",
       " 'ENST00000446136.1': ('ENSG00000230368.2', 'FAM41C', 'lncRNA'),\n",
       " 'ENST00000432963.1': ('ENSG00000230368.2', 'FAM41C', 'lncRNA'),\n",
       " 'ENST00000427857.1': ('ENSG00000230368.2', 'FAM41C', 'lncRNA'),\n",
       " 'ENST00000415481.1': ('ENSG00000234711.1',\n",
       "  'TUBB8P11',\n",
       "  'unprocessed_pseudogene'),\n",
       " 'ENST00000635557.1': ('ENSG00000283040.1',\n",
       "  'ENSG00000283040',\n",
       "  'transcribed_unprocessed_pseudogene'),\n",
       " 'ENST00000607769.1': ('ENSG00000272438.1', 'ENSG00000272438', 'lncRNA'),\n",
       " 'ENST00000448179.1': ('ENSG00000230699.2', 'ENSG00000230699', 'lncRNA'),\n",
       " 'ENST00000398216.2': ('ENSG00000241180.1', 'ENSG00000241180', 'lncRNA'),\n",
       " 'ENST00000609207.1': ('ENSG00000223764.2', 'LINC02593', 'lncRNA'),\n",
       " 'ENST00000417705.1': ('ENSG00000223764.2', 'LINC02593', 'lncRNA'),\n",
       " 'ENST00000432961.1': ('ENSG00000223764.2', 'LINC02593', 'lncRNA'),\n",
       " 'ENST00000616016.5': ('ENSG00000187634.13', 'SAMD11', 'protein_coding'),\n",
       " 'ENST00000618323.5': ('ENSG00000187634.13', 'SAMD11', 'protein_coding'),\n",
       " 'ENST00000437963.5': ('ENSG00000187634.13', 'SAMD11', 'protein_coding'),\n",
       " 'ENST00000342066.8': ('ENSG00000187634.13', 'SAMD11', 'protein_coding'),\n",
       " 'ENST00000616125.5': ('ENSG00000187634.13', 'SAMD11', 'protein_coding'),\n",
       " 'ENST00000618779.5': ('ENSG00000187634.13', 'SAMD11', 'protein_coding'),\n",
       " 'ENST00000622503.5': ('ENSG00000187634.13', 'SAMD11', 'protein_coding'),\n",
       " 'ENST00000618181.5': ('ENSG00000187634.13', 'SAMD11', 'protein_coding'),\n",
       " 'ENST00000617307.5': ('ENSG00000187634.13', 'SAMD11', 'protein_coding'),\n",
       " 'ENST00000341065.8': ('ENSG00000187634.13', 'SAMD11', 'protein_coding'),\n",
       " 'ENST00000455979.1': ('ENSG00000187634.13', 'SAMD11', 'protein_coding'),\n",
       " 'ENST00000478729.1': ('ENSG00000187634.13', 'SAMD11', 'protein_coding'),\n",
       " 'ENST00000474461.1': ('ENSG00000187634.13', 'SAMD11', 'protein_coding'),\n",
       " 'ENST00000466827.1': ('ENSG00000187634.13', 'SAMD11', 'protein_coding'),\n",
       " 'ENST00000464948.1': ('ENSG00000187634.13', 'SAMD11', 'protein_coding'),\n",
       " 'ENST00000327044.7': ('ENSG00000188976.11', 'NOC2L', 'protein_coding'),\n",
       " 'ENST00000483767.5': ('ENSG00000188976.11', 'NOC2L', 'protein_coding'),\n",
       " 'ENST00000477976.5': ('ENSG00000188976.11', 'NOC2L', 'protein_coding'),\n",
       " 'ENST00000496938.1': ('ENSG00000188976.11', 'NOC2L', 'protein_coding'),\n",
       " 'ENST00000487214.1': ('ENSG00000188976.11', 'NOC2L', 'protein_coding'),\n",
       " 'ENST00000469563.1': ('ENSG00000188976.11', 'NOC2L', 'protein_coding'),\n",
       " 'ENST00000338591.8': ('ENSG00000187961.15', 'KLHL17', 'protein_coding'),\n",
       " 'ENST00000463212.1': ('ENSG00000187961.15', 'KLHL17', 'protein_coding'),\n",
       " 'ENST00000466300.1': ('ENSG00000187961.15', 'KLHL17', 'protein_coding'),\n",
       " 'ENST00000481067.1': ('ENSG00000187961.15', 'KLHL17', 'protein_coding'),\n",
       " 'ENST00000379410.8': ('ENSG00000187583.11', 'PLEKHN1', 'protein_coding'),\n",
       " 'ENST00000379407.7': ('ENSG00000187583.11', 'PLEKHN1', 'protein_coding'),\n",
       " 'ENST00000379409.6': ('ENSG00000187583.11', 'PLEKHN1', 'protein_coding'),\n",
       " 'ENST00000480267.1': ('ENSG00000187583.11', 'PLEKHN1', 'protein_coding'),\n",
       " 'ENST00000491024.1': ('ENSG00000187583.11', 'PLEKHN1', 'protein_coding'),\n",
       " 'ENST00000433179.4': ('ENSG00000187642.10', 'PERM1', 'protein_coding'),\n",
       " 'ENST00000341290.6': ('ENSG00000187642.10', 'PERM1', 'protein_coding'),\n",
       " 'ENST00000479361.1': ('ENSG00000187642.10', 'PERM1', 'protein_coding'),\n",
       " 'ENST00000694917.1': ('ENSG00000187642.10', 'PERM1', 'protein_coding'),\n",
       " 'ENST00000606034.1': ('ENSG00000272512.1', 'ENSG00000272512', 'lncRNA'),\n",
       " 'ENST00000428771.6': ('ENSG00000188290.11', 'HES4', 'protein_coding'),\n",
       " 'ENST00000304952.11': ('ENSG00000188290.11', 'HES4', 'protein_coding'),\n",
       " 'ENST00000481869.1': ('ENSG00000188290.11', 'HES4', 'protein_coding'),\n",
       " 'ENST00000484667.2': ('ENSG00000188290.11', 'HES4', 'protein_coding'),\n",
       " 'ENST00000624697.4': ('ENSG00000187608.10', 'ISG15', 'protein_coding'),\n",
       " 'ENST00000624652.1': ('ENSG00000187608.10', 'ISG15', 'protein_coding'),\n",
       " 'ENST00000649529.1': ('ENSG00000187608.10', 'ISG15', 'protein_coding'),\n",
       " 'ENST00000423619.2': ('ENSG00000231702.2',\n",
       "  'ENSG00000231702',\n",
       "  'processed_pseudogene'),\n",
       " 'ENST00000458555.1': ('ENSG00000224969.1', 'ENSG00000224969', 'lncRNA'),\n",
       " 'ENST00000379370.7': ('ENSG00000188157.15', 'AGRN', 'protein_coding'),\n",
       " 'ENST00000620552.4': ('ENSG00000188157.15', 'AGRN', 'protein_coding'),\n",
       " 'ENST00000651234.1': ('ENSG00000188157.15', 'AGRN', 'protein_coding'),\n",
       " 'ENST00000652369.1': ('ENSG00000188157.15', 'AGRN', 'protein_coding'),\n",
       " 'ENST00000469403.1': ('ENSG00000188157.15', 'AGRN', 'protein_coding'),\n",
       " 'ENST00000479707.1': ('ENSG00000188157.15', 'AGRN', 'protein_coding'),\n",
       " 'ENST00000466223.1': ('ENSG00000188157.15', 'AGRN', 'protein_coding'),\n",
       " 'ENST00000478677.1': ('ENSG00000188157.15', 'AGRN', 'protein_coding'),\n",
       " 'ENST00000492947.1': ('ENSG00000188157.15', 'AGRN', 'protein_coding'),\n",
       " 'ENST00000461111.1': ('ENSG00000188157.15', 'AGRN', 'protein_coding'),\n",
       " 'ENST00000418300.1': ('ENSG00000242590.1', 'ENSG00000242590', 'lncRNA'),\n",
       " 'ENST00000688131.1': ('ENSG00000291156.1', 'ENSG00000291156', 'lncRNA'),\n",
       " 'ENST00000456409.6': ('ENSG00000291156.1', 'ENSG00000291156', 'lncRNA'),\n",
       " 'ENST00000702922.1': ('ENSG00000291156.1', 'ENSG00000291156', 'lncRNA'),\n",
       " 'ENST00000692821.2': ('ENSG00000291156.1', 'ENSG00000291156', 'lncRNA'),\n",
       " 'ENST00000394517.7': ('ENSG00000291156.1', 'ENSG00000291156', 'lncRNA'),\n",
       " 'ENST00000427998.6': ('ENSG00000291156.1', 'ENSG00000291156', 'lncRNA'),\n",
       " 'ENST00000685011.1': ('ENSG00000291156.1', 'ENSG00000291156', 'lncRNA'),\n",
       " 'ENST00000692266.1': ('ENSG00000291156.1', 'ENSG00000291156', 'lncRNA'),\n",
       " 'ENST00000433695.1': ('ENSG00000291156.1', 'ENSG00000291156', 'lncRNA'),\n",
       " 'ENST00000451054.1': ('ENSG00000291156.1', 'ENSG00000291156', 'lncRNA'),\n",
       " 'ENST00000442292.2': ('ENSG00000273443.1', 'ENSG00000273443', 'lncRNA'),\n",
       " 'ENST00000412397.2': ('ENSG00000217801.13',\n",
       "  'ENSG00000217801',\n",
       "  'transcribed_unprocessed_pseudogene'),\n",
       " 'ENST00000453464.3': ('ENSG00000237330.3', 'RNF223', 'protein_coding'),\n",
       " 'ENST00000379339.5': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000465822.5': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000421241.7': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000379325.7': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000487177.5': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000467751.5': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000477196.5': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000464905.1': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000379320.5': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000379319.5': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000434641.5': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000482816.5': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000462097.5': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000473600.1': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000475119.5': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000442117.5': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000427787.5': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000472741.5': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000480643.1': ('ENSG00000131591.18', 'C1orf159', 'protein_coding'),\n",
       " 'ENST00000648183.1': ('ENSG00000285812.1', 'ENSG00000285812', 'lncRNA'),\n",
       " 'ENST00000416774.1': ('ENSG00000223823.1', 'LINC01342', 'lncRNA'),\n",
       " 'ENST00000384997.3': ('ENSG00000207730.3', 'MIR200B', 'miRNA'),\n",
       " 'ENST00000384875.3': ('ENSG00000207607.3', 'MIR200A', 'miRNA'),\n",
       " 'ENST00000362106.1': ('ENSG00000198976.1', 'MIR429', 'miRNA'),\n",
       " 'ENST00000606993.1': ('ENSG00000272141.1', 'ENSG00000272141', 'lncRNA'),\n",
       " 'ENST00000379317.1': ('ENSG00000205231.1', 'TTLL10-AS1', 'lncRNA'),\n",
       " 'ENST00000379289.6': ('ENSG00000162571.14', 'TTLL10', 'protein_coding'),\n",
       " 'ENST00000379290.6': ('ENSG00000162571.14', 'TTLL10', 'protein_coding'),\n",
       " 'ENST00000506177.1': ('ENSG00000162571.14', 'TTLL10', 'protein_coding'),\n",
       " 'ENST00000673999.1': ('ENSG00000162571.14', 'TTLL10', 'protein_coding'),\n",
       " 'ENST00000379288.3': ('ENSG00000162571.14', 'TTLL10', 'protein_coding'),\n",
       " 'ENST00000460998.1': ('ENSG00000162571.14', 'TTLL10', 'protein_coding'),\n",
       " 'ENST00000514695.1': ('ENSG00000162571.14', 'TTLL10', 'protein_coding'),\n",
       " 'ENST00000486379.1': ('ENSG00000162571.14', 'TTLL10', 'protein_coding'),\n",
       " 'ENST00000328596.10': ('ENSG00000186891.14', 'TNFRSF18', 'protein_coding'),\n",
       " 'ENST00000379268.7': ('ENSG00000186891.14', 'TNFRSF18', 'protein_coding'),\n",
       " 'ENST00000486728.5': ('ENSG00000186891.14', 'TNFRSF18', 'protein_coding'),\n",
       " 'ENST00000379265.5': ('ENSG00000186891.14', 'TNFRSF18', 'protein_coding'),\n",
       " 'ENST00000699969.1': ('ENSG00000186827.12', 'TNFRSF4', 'protein_coding'),\n",
       " 'ENST00000699970.1': ('ENSG00000186827.12', 'TNFRSF4', 'protein_coding'),\n",
       " 'ENST00000497869.5': ('ENSG00000186827.12', 'TNFRSF4', 'protein_coding'),\n",
       " 'ENST00000379236.4': ('ENSG00000186827.12', 'TNFRSF4', 'protein_coding'),\n",
       " 'ENST00000699971.1': ('ENSG00000186827.12', 'TNFRSF4', 'protein_coding'),\n",
       " 'ENST00000699972.1': ('ENSG00000186827.12', 'TNFRSF4', 'protein_coding'),\n",
       " 'ENST00000699973.1': ('ENSG00000186827.12', 'TNFRSF4', 'protein_coding'),\n",
       " 'ENST00000699974.1': ('ENSG00000186827.12', 'TNFRSF4', 'protein_coding'),\n",
       " 'ENST00000699975.1': ('ENSG00000186827.12', 'TNFRSF4', 'protein_coding'),\n",
       " 'ENST00000699976.1': ('ENSG00000186827.12', 'TNFRSF4', 'protein_coding'),\n",
       " 'ENST00000699977.1': ('ENSG00000186827.12', 'TNFRSF4', 'protein_coding'),\n",
       " 'ENST00000699978.1': ('ENSG00000186827.12', 'TNFRSF4', 'protein_coding'),\n",
       " 'ENST00000699979.1': ('ENSG00000186827.12', 'TNFRSF4', 'protein_coding'),\n",
       " 'ENST00000453580.1': ('ENSG00000186827.12', 'TNFRSF4', 'protein_coding'),\n",
       " 'ENST00000360001.12': ('ENSG00000078808.20', 'SDF4', 'protein_coding'),\n",
       " 'ENST00000263741.12': ('ENSG00000078808.20', 'SDF4', 'protein_coding'),\n",
       " 'ENST00000465727.5': ('ENSG00000078808.20', 'SDF4', 'protein_coding'),\n",
       " 'ENST00000494748.1': ('ENSG00000078808.20', 'SDF4', 'protein_coding'),\n",
       " 'ENST00000478938.1': ('ENSG00000078808.20', 'SDF4', 'protein_coding'),\n",
       " 'ENST00000403997.2': ('ENSG00000078808.20', 'SDF4', 'protein_coding'),\n",
       " 'ENST00000459994.2': ('ENSG00000078808.20', 'SDF4', 'protein_coding'),\n",
       " 'ENST00000379198.5': ('ENSG00000176022.7', 'B3GALT6', 'protein_coding'),\n",
       " 'ENST00000462849.1': ('ENSG00000184163.3', 'C1QTNF12', 'protein_coding'),\n",
       " 'ENST00000478606.1': ('ENSG00000184163.3', 'C1QTNF12', 'protein_coding'),\n",
       " 'ENST00000330388.2': ('ENSG00000184163.3', 'C1QTNF12', 'protein_coding'),\n",
       " 'ENST00000468365.1': ('ENSG00000184163.3', 'C1QTNF12', 'protein_coding'),\n",
       " 'ENST00000486627.1': ('ENSG00000184163.3', 'C1QTNF12', 'protein_coding'),\n",
       " 'ENST00000565563.1': ('ENSG00000260179.1', 'ENSG00000260179', 'lncRNA'),\n",
       " 'ENST00000450390.6': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000349431.11': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000347370.6': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000400929.6': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000464036.5': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000360466.6': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000400930.8': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000467339.1': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000473215.5': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000435198.5': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000509720.5': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000491779.5': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000466752.5': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000503294.5': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000477894.5': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000471154.1': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000422076.5': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000502382.1': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000461142.3': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000488418.2': ('ENSG00000160087.21', 'UBE2J2', 'protein_coding'),\n",
       " 'ENST00000434139.1': ('ENSG00000230415.1', 'LINC01786', 'lncRNA'),\n",
       " 'ENST00000453732.1': ('ENSG00000230415.1', 'LINC01786', 'lncRNA'),\n",
       " 'ENST00000379116.10': ('ENSG00000162572.21', 'SCNN1D', 'protein_coding'),\n",
       " 'ENST00000379101.8': ('ENSG00000162572.21', 'SCNN1D', 'protein_coding'),\n",
       " 'ENST00000338555.6': ('ENSG00000162572.21', 'SCNN1D', 'protein_coding'),\n",
       " 'ENST00000400928.7': ('ENSG00000162572.21', 'SCNN1D', 'protein_coding'),\n",
       " 'ENST00000470022.1': ('ENSG00000162572.21', 'SCNN1D', 'protein_coding'),\n",
       " 'ENST00000325425.12': ('ENSG00000162572.21', 'SCNN1D', 'protein_coding'),\n",
       " 'ENST00000467651.2': ('ENSG00000162572.21', 'SCNN1D', 'protein_coding'),\n",
       " 'ENST00000379099.3': ('ENSG00000162572.21', 'SCNN1D', 'protein_coding'),\n",
       " 'ENST00000492936.5': ('ENSG00000131584.19', 'ACAP3', 'protein_coding'),\n",
       " 'ENST00000354700.10': ('ENSG00000131584.19', 'ACAP3', 'protein_coding'),\n",
       " 'ENST00000467278.5': ('ENSG00000131584.19', 'ACAP3', 'protein_coding'),\n",
       " 'ENST00000353662.4': ('ENSG00000131584.19', 'ACAP3', 'protein_coding'),\n",
       " 'ENST00000470659.1': ('ENSG00000131584.19', 'ACAP3', 'protein_coding'),\n",
       " 'ENST00000379037.6': ('ENSG00000131584.19', 'ACAP3', 'protein_coding'),\n",
       " 'ENST00000476572.1': ('ENSG00000131584.19', 'ACAP3', 'protein_coding'),\n",
       " 'ENST00000493992.1': ('ENSG00000131584.19', 'ACAP3', 'protein_coding'),\n",
       " 'ENST00000472541.5': ('ENSG00000131584.19', 'ACAP3', 'protein_coding'),\n",
       " 'ENST00000354980.7': ('ENSG00000131584.19', 'ACAP3', 'protein_coding'),\n",
       " 'ENST00000478065.5': ('ENSG00000131584.19', 'ACAP3', 'protein_coding'),\n",
       " 'ENST00000479108.5': ('ENSG00000131584.19', 'ACAP3', 'protein_coding'),\n",
       " 'ENST00000438966.1': ('ENSG00000131584.19', 'ACAP3', 'protein_coding'),\n",
       " 'ENST00000613751.1': ('ENSG00000278073.1', 'MIR6726', 'miRNA'),\n",
       " 'ENST00000379031.10': ('ENSG00000169972.13', 'PUSL1', 'protein_coding'),\n",
       " 'ENST00000470520.5': ('ENSG00000169972.13', 'PUSL1', 'protein_coding'),\n",
       " 'ENST00000463758.1': ('ENSG00000169972.13', 'PUSL1', 'protein_coding'),\n",
       " 'ENST00000467712.1': ('ENSG00000169972.13', 'PUSL1', 'protein_coding'),\n",
       " 'ENST00000706601.1': ('ENSG00000169972.13', 'PUSL1', 'protein_coding'),\n",
       " 'ENST00000493657.1': ('ENSG00000169972.13', 'PUSL1', 'protein_coding'),\n",
       " 'ENST00000706602.1': ('ENSG00000169972.13', 'PUSL1', 'protein_coding'),\n",
       " 'ENST00000470030.6': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000540437.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000545578.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000620829.4': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000497304.6': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000323275.10': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000435064.6': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000411962.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000462432.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000478641.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000421495.6': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000458452.7': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000419704.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000532772.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000450926.6': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000527098.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000485710.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000461514.6': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000525164.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000528879.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000525769.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000531377.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000527383.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000531292.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000467408.6': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000434694.6': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000531020.1': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000533916.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000488042.6': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000526797.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000526332.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000430786.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000527719.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000530031.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000526113.1': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000526904.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000525603.1': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000429572.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000498173.2': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000496353.1': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000493534.6': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000534345.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000532952.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000490853.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000531019.5': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000498476.6': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000470679.3': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000525285.1': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000530233.1': ('ENSG00000127054.22', 'INTS11', 'protein_coding'),\n",
       " 'ENST00000620702.1': ('ENSG00000283712.1', 'MIR6727', 'miRNA'),\n",
       " 'ENST00000444968.1': ('ENSG00000240731.1', 'ENSG00000240731', 'lncRNA'),\n",
       " 'ENST00000464957.1': ('ENSG00000224051.7', 'CPTP', 'protein_coding'),\n",
       " 'ENST00000343938.9': ('ENSG00000224051.7', 'CPTP', 'protein_coding'),\n",
       " 'ENST00000488011.1': ('ENSG00000224051.7', 'CPTP', 'protein_coding'),\n",
       " 'ENST00000339381.6': ('ENSG00000169962.5', 'TAS1R3', 'protein_coding'),\n",
       " 'ENST00000378891.9': ('ENSG00000107404.21', 'DVL1', 'protein_coding'),\n",
       " 'ENST00000378888.10': ('ENSG00000107404.21', 'DVL1', 'protein_coding'),\n",
       " 'ENST00000632445.1': ('ENSG00000107404.21', 'DVL1', 'protein_coding'),\n",
       " 'ENST00000631679.1': ('ENSG00000107404.21', 'DVL1', 'protein_coding'),\n",
       " 'ENST00000634054.1': ('ENSG00000107404.21', 'DVL1', 'protein_coding'),\n",
       " 'ENST00000633096.1': ('ENSG00000107404.21', 'DVL1', 'protein_coding'),\n",
       " 'ENST00000472445.1': ('ENSG00000107404.21', 'DVL1', 'protein_coding'),\n",
       " 'ENST00000616525.1': ('ENSG00000284372.1', 'MIR6808', 'miRNA'),\n",
       " 'ENST00000474033.5': ('ENSG00000162576.17', 'MXRA8', 'protein_coding'),\n",
       " 'ENST00000342753.8': ('ENSG00000162576.17', 'MXRA8', 'protein_coding'),\n",
       " 'ENST00000309212.11': ('ENSG00000162576.17', 'MXRA8', 'protein_coding'),\n",
       " 'ENST00000445648.5': ('ENSG00000162576.17', 'MXRA8', 'protein_coding'),\n",
       " 'ENST00000477278.3': ('ENSG00000162576.17', 'MXRA8', 'protein_coding'),\n",
       " 'ENST00000473097.5': ('ENSG00000162576.17', 'MXRA8', 'protein_coding'),\n",
       " 'ENST00000478517.5': ('ENSG00000162576.17', 'MXRA8', 'protein_coding'),\n",
       " 'ENST00000476718.1': ('ENSG00000162576.17', 'MXRA8', 'protein_coding'),\n",
       " 'ENST00000464351.1': ('ENSG00000162576.17', 'MXRA8', 'protein_coding'),\n",
       " 'ENST00000460473.1': ('ENSG00000162576.17', 'MXRA8', 'protein_coding'),\n",
       " 'ENST00000338370.7': ('ENSG00000175756.14', 'AURKAIP1', 'protein_coding'),\n",
       " 'ENST00000321751.9': ('ENSG00000175756.14', 'AURKAIP1', 'protein_coding'),\n",
       " 'ENST00000338338.10': ('ENSG00000175756.14', 'AURKAIP1', 'protein_coding'),\n",
       " 'ENST00000378853.3': ('ENSG00000175756.14', 'AURKAIP1', 'protein_coding'),\n",
       " 'ENST00000489799.1': ('ENSG00000175756.14', 'AURKAIP1', 'protein_coding'),\n",
       " 'ENST00000496905.1': ('ENSG00000175756.14', 'AURKAIP1', 'protein_coding'),\n",
       " 'ENST00000435351.2': ('ENSG00000223663.2',\n",
       "  'NDUFB4P8',\n",
       "  'processed_pseudogene'),\n",
       " 'ENST00000480479.5': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000400809.8': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000482621.5': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000463260.5': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000496007.5': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000482365.5': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000481223.6': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000418865.6': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000505849.5': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000488340.5': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000492998.5': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000469113.5': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000480646.1': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000471930.5': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000408918.8': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000473872.5': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000425598.2': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000463895.1': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000497013.1': ('ENSG00000221978.13', 'CCNL2', 'protein_coding'),\n",
       " 'ENST00000444362.3': ('ENSG00000224870.9', 'MRPL20-AS1', 'lncRNA'),\n",
       " 'ENST00000448629.8': ('ENSG00000224870.9', 'MRPL20-AS1', 'lncRNA'),\n",
       " 'ENST00000572242.2': ('ENSG00000224870.9', 'MRPL20-AS1', 'lncRNA'),\n",
       " 'ENST00000659048.2': ('ENSG00000224870.9', 'MRPL20-AS1', 'lncRNA'),\n",
       " 'ENST00000685968.1': ('ENSG00000224870.9', 'MRPL20-AS1', 'lncRNA'),\n",
       " 'ENST00000418833.2': ('ENSG00000224870.9', 'MRPL20-AS1', 'lncRNA'),\n",
       " 'ENST00000689419.2': ('ENSG00000224870.9', 'MRPL20-AS1', 'lncRNA'),\n",
       " 'ENST00000690954.1': ('ENSG00000224870.9', 'MRPL20-AS1', 'lncRNA'),\n",
       " 'ENST00000576232.2': ('ENSG00000224870.9', 'MRPL20-AS1', 'lncRNA'),\n",
       " 'ENST00000570344.1': ('ENSG00000224870.9', 'MRPL20-AS1', 'lncRNA'),\n",
       " 'ENST00000702056.1': ('ENSG00000224870.9', 'MRPL20-AS1', 'lncRNA'),\n",
       " 'ENST00000447725.2': ('ENSG00000224870.9', 'MRPL20-AS1', 'lncRNA'),\n",
       " 'ENST00000702620.1': ('ENSG00000224870.9', 'MRPL20-AS1', 'lncRNA'),\n",
       " 'ENST00000453521.1': ('ENSG00000224870.9', 'MRPL20-AS1', 'lncRNA'),\n",
       " 'ENST00000344843.12': ('ENSG00000242485.6', 'MRPL20', 'protein_coding'),\n",
       " 'ENST00000493287.5': ('ENSG00000242485.6', 'MRPL20', 'protein_coding'),\n",
       " 'ENST00000487659.1': ('ENSG00000242485.6', 'MRPL20', 'protein_coding'),\n",
       " 'ENST00000492508.1': ('ENSG00000242485.6', 'MRPL20', 'protein_coding'),\n",
       " 'ENST00000482352.1': ('ENSG00000242485.6', 'MRPL20', 'protein_coding'),\n",
       " 'ENST00000477686.1': ('ENSG00000242485.6', 'MRPL20', 'protein_coding'),\n",
       " 'ENST00000582431.2': ('ENSG00000264293.2', 'RN7SL657P', 'misc_RNA'),\n",
       " 'ENST00000607307.1': ('ENSG00000272455.1', 'MRPL20-DT', 'lncRNA'),\n",
       " 'ENST00000537107.6': ('ENSG00000235098.9', 'ANKRD65', 'protein_coding'),\n",
       " 'ENST00000520296.5': ('ENSG00000235098.9', 'ANKRD65', 'protein_coding'),\n",
       " 'ENST00000427211.3': ('ENSG00000235098.9', 'ANKRD65', 'protein_coding'),\n",
       " 'ENST00000454272.2': ('ENSG00000235098.9', 'ANKRD65', 'protein_coding'),\n",
       " 'ENST00000442470.1': ('ENSG00000235098.9', 'ANKRD65', 'protein_coding'),\n",
       " 'ENST00000428932.1': ('ENSG00000225905.1', 'ANKRD65-AS1', 'lncRNA'),\n",
       " 'ENST00000378821.4': ('ENSG00000205116.4', 'TMEM88B', 'protein_coding'),\n",
       " 'ENST00000430109.1': ('ENSG00000225285.1', 'LINC01770', 'lncRNA'),\n",
       " 'ENST00000434150.1': ('ENSG00000225285.1', 'LINC01770', 'lncRNA'),\n",
       " 'ENST00000454562.1': ('ENSG00000225285.1', 'LINC01770', 'lncRNA'),\n",
       " 'ENST00000417917.1': ('ENSG00000225285.1', 'LINC01770', 'lncRNA'),\n",
       " 'ENST00000495558.1': ('ENSG00000179403.12', 'VWA1', 'protein_coding'),\n",
       " 'ENST00000338660.5': ('ENSG00000179403.12', 'VWA1', 'protein_coding'),\n",
       " 'ENST00000476993.2': ('ENSG00000179403.12', 'VWA1', 'protein_coding'),\n",
       " 'ENST00000471398.1': ('ENSG00000179403.12', 'VWA1', 'protein_coding'),\n",
       " 'ENST00000378785.7': ('ENSG00000215915.10', 'ATAD3C', 'protein_coding'),\n",
       " 'ENST00000475091.2': ('ENSG00000215915.10', 'ATAD3C', 'protein_coding'),\n",
       " 'ENST00000484537.2': ('ENSG00000215915.10', 'ATAD3C', 'protein_coding'),\n",
       " 'ENST00000673477.1': ('ENSG00000160072.20', 'ATAD3B', 'protein_coding'),\n",
       " 'ENST00000308647.8': ('ENSG00000160072.20', 'ATAD3B', 'protein_coding'),\n",
       " 'ENST00000472194.6': ('ENSG00000160072.20', 'ATAD3B', 'protein_coding'),\n",
       " 'ENST00000378736.3': ('ENSG00000160072.20', 'ATAD3B', 'protein_coding'),\n",
       " 'ENST00000485748.5': ('ENSG00000160072.20', 'ATAD3B', 'protein_coding'),\n",
       " 'ENST00000474481.1': ('ENSG00000160072.20', 'ATAD3B', 'protein_coding'),\n",
       " 'ENST00000641974.1': ('ENSG00000290916.1', 'ENSG00000290916', 'lncRNA'),\n",
       " 'ENST00000692936.1': ('ENSG00000290916.1', 'ENSG00000290916', 'lncRNA'),\n",
       " 'ENST00000641679.1': ('ENSG00000284740.3',\n",
       "  'ENSG00000284740',\n",
       "  'transcribed_unprocessed_pseudogene'),\n",
       " 'ENST00000378756.8': ('ENSG00000197785.14', 'ATAD3A', 'protein_coding'),\n",
       " 'ENST00000672388.1': ('ENSG00000197785.14', 'ATAD3A', 'protein_coding'),\n",
       " 'ENST00000378755.9': ('ENSG00000197785.14', 'ATAD3A', 'protein_coding'),\n",
       " 'ENST00000339113.9': ('ENSG00000197785.14', 'ATAD3A', 'protein_coding'),\n",
       " 'ENST00000536055.6': ('ENSG00000197785.14', 'ATAD3A', 'protein_coding'),\n",
       " 'ENST00000439513.1': ('ENSG00000197785.14', 'ATAD3A', 'protein_coding'),\n",
       " 'ENST00000429957.1': ('ENSG00000197785.14', 'ATAD3A', 'protein_coding'),\n",
       " 'ENST00000400830.4': ('ENSG00000197785.14', 'ATAD3A', 'protein_coding'),\n",
       " 'ENST00000378733.9': ('ENSG00000205090.10', 'TMEM240', 'protein_coding'),\n",
       " 'ENST00000624426.1': ('ENSG00000205090.10', 'TMEM240', 'protein_coding'),\n",
       " 'ENST00000291386.4': ('ENSG00000160075.12', 'SSU72', 'protein_coding'),\n",
       " 'ENST00000378726.1': ('ENSG00000160075.12', 'SSU72', 'protein_coding'),\n",
       " 'ENST00000378725.3': ('ENSG00000160075.12', 'SSU72', 'protein_coding'),\n",
       " 'ENST00000359060.5': ('ENSG00000160075.12', 'SSU72', 'protein_coding'),\n",
       " 'ENST00000647566.1': ('ENSG00000215014.5', 'ENSG00000215014', 'lncRNA'),\n",
       " 'ENST00000366221.3': ('ENSG00000215014.5', 'ENSG00000215014', 'lncRNA'),\n",
       " 'ENST00000621860.1': ('ENSG00000274481.1',\n",
       "  'ENSG00000274481',\n",
       "  'processed_pseudogene'),\n",
       " 'ENST00000422725.4': ('ENSG00000228594.4', 'FNDC10', 'protein_coding'),\n",
       " 'ENST00000658591.1': ('ENSG00000286989.1', 'ENSG00000286989', 'lncRNA'),\n",
       " 'ENST00000607222.1': ('ENSG00000272106.1', 'ENSG00000272106', 'lncRNA'),\n",
       " 'ENST00000479659.5': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000489635.5': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000355826.10': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000505820.7': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000518681.6': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000520777.6': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000511502.5': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000477990.1': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000514363.1': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000487053.5': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000502470.5': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000378712.5': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000507082.1': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000463492.1': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000510793.5': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000504599.6': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000507229.5': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000503789.5': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000378708.5': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000506488.5': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000512004.1': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000508455.1': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000514234.5': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000467597.1': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000508148.1': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000486072.1': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000505370.1': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000473511.5': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000464570.1': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000483015.1': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000511910.1': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000470373.1': ('ENSG00000197530.13', 'MIB2', 'protein_coding'),\n",
       " 'ENST00000472264.1': ('ENSG00000189409.14', 'MMP23B', 'protein_coding'),\n",
       " 'ENST00000356026.10': ('ENSG00000189409.14', 'MMP23B', 'protein_coding'),\n",
       " 'ENST00000512731.5': ('ENSG00000189409.14', 'MMP23B', 'protein_coding'),\n",
       " 'ENST00000378675.7': ('ENSG00000189409.14', 'MMP23B', 'protein_coding'),\n",
       " 'ENST00000479814.5': ('ENSG00000189409.14', 'MMP23B', 'protein_coding'),\n",
       " 'ENST00000435358.5': ('ENSG00000189409.14', 'MMP23B', 'protein_coding'),\n",
       " 'ENST00000503792.1': ('ENSG00000189409.14', 'MMP23B', 'protein_coding'),\n",
       " 'ENST00000489782.1': ('ENSG00000189409.14', 'MMP23B', 'protein_coding'),\n",
       " 'ENST00000486400.1': ('ENSG00000189409.14', 'MMP23B', 'protein_coding'),\n",
       " 'ENST00000490017.1': ('ENSG00000189409.14', 'MMP23B', 'protein_coding'),\n",
       " 'ENST00000341832.11': ('ENSG00000248333.9', 'CDK11B', 'protein_coding'),\n",
       " 'ENST00000407249.7': ('ENSG00000248333.9', 'CDK11B', 'protein_coding'),\n",
       " 'ENST00000340677.9': ('ENSG00000248333.9', 'CDK11B', 'protein_coding'),\n",
       " 'ENST00000611150.3': ('ENSG00000248333.9', 'CDK11B', 'protein_coding'),\n",
       " 'ENST00000615951.4': ('ENSG00000248333.9', 'CDK11B', 'protein_coding'),\n",
       " 'ENST00000626918.2': ('ENSG00000248333.9', 'CDK11B', 'protein_coding'),\n",
       " 'ENST00000629289.2': ('ENSG00000248333.9', 'CDK11B', 'protein_coding'),\n",
       " 'ENST00000629312.2': ('ENSG00000248333.9', 'CDK11B', 'protein_coding'),\n",
       " 'ENST00000341028.8': ('ENSG00000248333.9', 'CDK11B', 'protein_coding'),\n",
       " 'ENST00000607013.1': ('ENSG00000272004.1', 'ENSG00000272004', 'lncRNA'),\n",
       " 'ENST00000614300.4': ('ENSG00000189339.12', 'SLC35E2B', 'protein_coding'),\n",
       " 'ENST00000617444.5': ('ENSG00000189339.12', 'SLC35E2B', 'protein_coding'),\n",
       " 'ENST00000611123.1': ('ENSG00000189339.12', 'SLC35E2B', 'protein_coding'),\n",
       " 'ENST00000480991.1': ('ENSG00000189339.12', 'SLC35E2B', 'protein_coding'),\n",
       " 'ENST00000481276.1': ('ENSG00000189339.12', 'SLC35E2B', 'protein_coding'),\n",
       " 'ENST00000596308.1': ('ENSG00000269737.2',\n",
       "  'ENSG00000269737',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000597891.1': ('ENSG00000269737.2',\n",
       "  'ENSG00000269737',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000234610.5': ('ENSG00000215914.4',\n",
       "  'MMP23A',\n",
       "  'unprocessed_pseudogene'),\n",
       " 'ENST00000404249.8': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000356200.7': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000463652.5': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000356937.7': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000378638.6': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000509982.5': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000357760.6': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000358779.9': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000378633.5': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000460465.5': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000478901.1': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000468397.1': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000491311.1': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000495016.5': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000468800.5': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000464748.1': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000489300.1': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000492390.5': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000474916.1': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000479362.1': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000498810.1': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000487462.1': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000401096.2': ('ENSG00000008128.23', 'CDK11A', 'protein_coding'),\n",
       " 'ENST00000598846.1': ('ENSG00000268575.1', 'ENSG00000268575', 'lncRNA'),\n",
       " 'ENST00000577672.1': ('ENSG00000227775.3',\n",
       "  'ENSG00000227775',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000417099.1': ('ENSG00000227775.3',\n",
       "  'ENSG00000227775',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000424604.1': ('ENSG00000227775.3',\n",
       "  'ENSG00000227775',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000418950.2': ('ENSG00000227775.3',\n",
       "  'ENSG00000227775',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000647043.1': ('ENSG00000290854.1', 'ENSG00000290854', 'lncRNA'),\n",
       " 'ENST00000475229.2': ('ENSG00000290854.1', 'ENSG00000290854', 'lncRNA'),\n",
       " 'ENST00000246421.4': ('ENSG00000290854.1', 'ENSG00000290854', 'lncRNA'),\n",
       " 'ENST00000643943.1': ('ENSG00000215790.9',\n",
       "  'SLC35E2A',\n",
       "  'transcribed_unprocessed_pseudogene'),\n",
       " 'ENST00000341426.9': ('ENSG00000008130.15', 'NADK', 'protein_coding'),\n",
       " 'ENST00000341991.7': ('ENSG00000008130.15', 'NADK', 'protein_coding'),\n",
       " 'ENST00000378625.5': ('ENSG00000008130.15', 'NADK', 'protein_coding'),\n",
       " 'ENST00000342348.9': ('ENSG00000008130.15', 'NADK', 'protein_coding'),\n",
       " 'ENST00000498806.1': ('ENSG00000008130.15', 'NADK', 'protein_coding'),\n",
       " 'ENST00000469045.5': ('ENSG00000008130.15', 'NADK', 'protein_coding'),\n",
       " 'ENST00000480499.5': ('ENSG00000008130.15', 'NADK', 'protein_coding'),\n",
       " 'ENST00000497747.5': ('ENSG00000008130.15', 'NADK', 'protein_coding'),\n",
       " 'ENST00000477235.5': ('ENSG00000008130.15', 'NADK', 'protein_coding'),\n",
       " 'ENST00000489538.1': ('ENSG00000008130.15', 'NADK', 'protein_coding'),\n",
       " 'ENST00000497615.1': ('ENSG00000008130.15', 'NADK', 'protein_coding'),\n",
       " 'ENST00000497186.5': ('ENSG00000008130.15', 'NADK', 'protein_coding'),\n",
       " 'ENST00000492845.6': ('ENSG00000008130.15', 'NADK', 'protein_coding'),\n",
       " 'ENST00000492768.5': ('ENSG00000008130.15', 'NADK', 'protein_coding'),\n",
       " 'ENST00000460602.5': ('ENSG00000008130.15', 'NADK', 'protein_coding'),\n",
       " 'ENST00000464373.1': ('ENSG00000008130.15', 'NADK', 'protein_coding'),\n",
       " 'ENST00000480542.1': ('ENSG00000008130.15', 'NADK', 'protein_coding'),\n",
       " 'ENST00000615252.5': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000610897.4': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000378609.9': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000439272.7': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703692.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703693.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703694.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703695.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703696.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703697.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703698.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703699.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703700.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703701.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703702.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703703.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703704.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703705.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703706.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703707.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703708.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703709.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703710.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703711.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000461893.5': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000471354.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000434686.6': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000437146.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000607589.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000472614.2': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000703712.1': ('ENSG00000078369.19', 'GNB1', 'protein_coding'),\n",
       " 'ENST00000689170.1': ('ENSG00000231050.3', 'GNB1-DT', 'lncRNA'),\n",
       " 'ENST00000686550.1': ('ENSG00000231050.3', 'GNB1-DT', 'lncRNA'),\n",
       " 'ENST00000685593.1': ('ENSG00000231050.3', 'GNB1-DT', 'lncRNA'),\n",
       " 'ENST00000702668.1': ('ENSG00000231050.3', 'GNB1-DT', 'lncRNA'),\n",
       " 'ENST00000412228.1': ('ENSG00000231050.3', 'GNB1-DT', 'lncRNA'),\n",
       " 'ENST00000378604.3': ('ENSG00000169885.10', 'CALML6', 'protein_coding'),\n",
       " 'ENST00000307786.8': ('ENSG00000169885.10', 'CALML6', 'protein_coding'),\n",
       " 'ENST00000482402.1': ('ENSG00000169885.10', 'CALML6', 'protein_coding'),\n",
       " 'ENST00000462293.1': ('ENSG00000169885.10', 'CALML6', 'protein_coding'),\n",
       " 'ENST00000378602.3': ('ENSG00000178821.13', 'TMEM52', 'protein_coding'),\n",
       " 'ENST00000310991.8': ('ENSG00000178821.13', 'TMEM52', 'protein_coding'),\n",
       " 'ENST00000378598.4': ('ENSG00000178821.13', 'TMEM52', 'protein_coding'),\n",
       " 'ENST00000470931.2': ('ENSG00000178821.13', 'TMEM52', 'protein_coding'),\n",
       " 'ENST00000602604.1': ('ENSG00000178821.13', 'TMEM52', 'protein_coding'),\n",
       " 'ENST00000416272.1': ('ENSG00000178821.13', 'TMEM52', 'protein_coding'),\n",
       " 'ENST00000493964.5': ('ENSG00000142609.20', 'CFAP74', 'protein_coding'),\n",
       " 'ENST00000464311.5': ('ENSG00000142609.20', 'CFAP74', 'protein_coding'),\n",
       " 'ENST00000682832.2': ('ENSG00000142609.20', 'CFAP74', 'protein_coding'),\n",
       " 'ENST00000412120.2': ('ENSG00000142609.20', 'CFAP74', 'protein_coding'),\n",
       " 'ENST00000493316.5': ('ENSG00000142609.20', 'CFAP74', 'protein_coding'),\n",
       " 'ENST00000270720.11': ('ENSG00000142609.20', 'CFAP74', 'protein_coding'),\n",
       " 'ENST00000468610.5': ('ENSG00000142609.20', 'CFAP74', 'protein_coding'),\n",
       " 'ENST00000461752.3': ('ENSG00000142609.20', 'CFAP74', 'protein_coding'),\n",
       " 'ENST00000378592.1': ('ENSG00000142609.20', 'CFAP74', 'protein_coding'),\n",
       " 'ENST00000378590.4': ('ENSG00000142609.20', 'CFAP74', 'protein_coding'),\n",
       " 'ENST00000443930.1': ('ENSG00000233542.2', 'ENSG00000233542', 'lncRNA'),\n",
       " 'ENST00000659674.1': ('ENSG00000233542.2', 'ENSG00000233542', 'lncRNA'),\n",
       " 'ENST00000638411.1': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000378585.7': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000639045.1': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000640067.1': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000638604.1': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000638771.1': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000639935.1': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000640949.1': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000640030.1': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000640423.1': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000639625.1': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000639777.1': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000640981.1': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000638763.1': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000639070.1': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000640892.1': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000640317.1': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000638804.1': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000640688.1': ('ENSG00000187730.9', 'GABRD', 'protein_coding'),\n",
       " 'ENST00000445600.1': ('ENSG00000226969.1', 'PRKCZ-DT', 'lncRNA'),\n",
       " 'ENST00000449154.1': ('ENSG00000226969.1', 'PRKCZ-DT', 'lncRNA'),\n",
       " 'ENST00000378567.8': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000468310.5': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000484419.1': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000503297.1': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000479566.5': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000466651.5': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000400921.6': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000495347.5': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000461106.6': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000461465.5': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000470596.5': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000496325.5': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000481140.5': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000482686.5': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000486681.5': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000470986.5': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000470511.5': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000471018.6': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000466352.5': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000497183.5': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000419838.6': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000479263.5': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000503672.1': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000478770.6': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000505322.5': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000472017.1': ('ENSG00000067606.17', 'PRKCZ', 'protein_coding'),\n",
       " 'ENST00000606533.1': ('ENSG00000271806.1', 'ENSG00000271806', 'lncRNA'),\n",
       " 'ENST00000333854.2': ('ENSG00000182873.5', 'PRKCZ-AS1', 'lncRNA'),\n",
       " 'ENST00000444529.1': ('ENSG00000182873.5', 'PRKCZ-AS1', 'lncRNA'),\n",
       " 'ENST00000400919.7': ('ENSG00000162585.17', 'FAAP20', 'protein_coding'),\n",
       " 'ENST00000469733.5': ('ENSG00000162585.17', 'FAAP20', 'protein_coding'),\n",
       " 'ENST00000428120.5': ('ENSG00000162585.17', 'FAAP20', 'protein_coding'),\n",
       " 'ENST00000414253.5': ('ENSG00000162585.17', 'FAAP20', 'protein_coding'),\n",
       " 'ENST00000400918.7': ('ENSG00000162585.17', 'FAAP20', 'protein_coding'),\n",
       " 'ENST00000497675.2': ('ENSG00000162585.17', 'FAAP20', 'protein_coding'),\n",
       " 'ENST00000378543.2': ('ENSG00000162585.17', 'FAAP20', 'protein_coding'),\n",
       " 'ENST00000378546.9': ('ENSG00000162585.17', 'FAAP20', 'protein_coding'),\n",
       " 'ENST00000401813.7': ('ENSG00000162585.17', 'FAAP20', 'protein_coding'),\n",
       " 'ENST00000487186.5': ('ENSG00000162585.17', 'FAAP20', 'protein_coding'),\n",
       " 'ENST00000476803.1': ('ENSG00000162585.17', 'FAAP20', 'protein_coding'),\n",
       " 'ENST00000440825.6': ('ENSG00000162585.17', 'FAAP20', 'protein_coding'),\n",
       " 'ENST00000420515.1': ('ENSG00000162585.17', 'FAAP20', 'protein_coding'),\n",
       " 'ENST00000514625.1': ('ENSG00000162585.17', 'FAAP20', 'protein_coding'),\n",
       " 'ENST00000420964.1': ('ENSG00000162585.17', 'FAAP20', 'protein_coding'),\n",
       " 'ENST00000442483.2': ('ENSG00000234396.3', 'ENSG00000234396', 'lncRNA'),\n",
       " 'ENST00000704337.1': ('ENSG00000157933.11', 'SKI', 'protein_coding'),\n",
       " 'ENST00000378536.5': ('ENSG00000157933.11', 'SKI', 'protein_coding'),\n",
       " 'ENST00000508416.1': ('ENSG00000157933.11', 'SKI', 'protein_coding'),\n",
       " 'ENST00000478223.2': ('ENSG00000157933.11', 'SKI', 'protein_coding'),\n",
       " 'ENST00000507179.1': ('ENSG00000157933.11', 'SKI', 'protein_coding'),\n",
       " 'ENST00000659232.1': ('ENSG00000287356.1', 'ENSG00000287356', 'lncRNA'),\n",
       " 'ENST00000661319.1': ('ENSG00000287356.1', 'ENSG00000287356', 'lncRNA'),\n",
       " 'ENST00000378531.8': ('ENSG00000116151.14', 'MORN1', 'protein_coding'),\n",
       " 'ENST00000606372.5': ('ENSG00000116151.14', 'MORN1', 'protein_coding'),\n",
       " 'ENST00000378529.7': ('ENSG00000116151.14', 'MORN1', 'protein_coding'),\n",
       " 'ENST00000607342.5': ('ENSG00000116151.14', 'MORN1', 'protein_coding'),\n",
       " 'ENST00000469374.5': ('ENSG00000116151.14', 'MORN1', 'protein_coding'),\n",
       " 'ENST00000419785.1': ('ENSG00000116151.14', 'MORN1', 'protein_coding'),\n",
       " 'ENST00000449373.2': ('ENSG00000116151.14', 'MORN1', 'protein_coding'),\n",
       " 'ENST00000378525.2': ('ENSG00000116151.14', 'MORN1', 'protein_coding'),\n",
       " 'ENST00000607031.1': ('ENSG00000116151.14', 'MORN1', 'protein_coding'),\n",
       " 'ENST00000475812.1': ('ENSG00000116151.14', 'MORN1', 'protein_coding'),\n",
       " 'ENST00000494279.2': ('ENSG00000116151.14', 'MORN1', 'protein_coding'),\n",
       " 'ENST00000607720.1': ('ENSG00000272161.1', 'ENSG00000272161', 'lncRNA'),\n",
       " 'ENST00000602865.1': ('ENSG00000269896.2',\n",
       "  'ENSG00000269896',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000417061.2': ('ENSG00000269896.2',\n",
       "  'ENSG00000269896',\n",
       "  'transcribed_processed_pseudogene'),\n",
       " 'ENST00000607858.1': ('ENSG00000272420.1', 'ENSG00000272420', 'lncRNA'),\n",
       " 'ENST00000378513.7': ('ENSG00000157916.20', 'RER1', 'protein_coding'),\n",
       " 'ENST00000378518.5': ('ENSG00000157916.20', 'RER1', 'protein_coding'),\n",
       " 'ENST00000605895.6': ('ENSG00000157916.20', 'RER1', 'protein_coding'),\n",
       " 'ENST00000306256.13': ('ENSG00000157916.20', 'RER1', 'protein_coding'),\n",
       " 'ENST00000493207.1': ('ENSG00000157916.20', 'RER1', 'protein_coding'),\n",
       " 'ENST00000378512.5': ('ENSG00000157916.20', 'RER1', 'protein_coding'),\n",
       " 'ENST00000443438.5': ('ENSG00000157916.20', 'RER1', 'protein_coding'),\n",
       " 'ENST00000488353.2': ('ENSG00000157916.20', 'RER1', 'protein_coding'),\n",
       " 'ENST00000462129.1': ('ENSG00000157916.20', 'RER1', 'protein_coding'),\n",
       " 'ENST00000288774.8': ('ENSG00000157911.11', 'PEX10', 'protein_coding'),\n",
       " 'ENST00000447513.7': ('ENSG00000157911.11', 'PEX10', 'protein_coding'),\n",
       " 'ENST00000650293.1': ('ENSG00000157911.11', 'PEX10', 'protein_coding'),\n",
       " 'ENST00000507596.5': ('ENSG00000157911.11', 'PEX10', 'protein_coding'),\n",
       " 'ENST00000510434.1': ('ENSG00000157911.11', 'PEX10', 'protein_coding'),\n",
       " 'ENST00000508384.5': ('ENSG00000157911.11', 'PEX10', 'protein_coding'),\n",
       " 'ENST00000515760.1': ('ENSG00000157911.11', 'PEX10', 'protein_coding'),\n",
       " 'ENST00000502666.1': ('ENSG00000157911.11', 'PEX10', 'protein_coding'),\n",
       " 'ENST00000514502.1': ('ENSG00000157911.11', 'PEX10', 'protein_coding'),\n",
       " 'ENST00000609981.5': ('ENSG00000149527.18', 'PLCH2', 'protein_coding'),\n",
       " 'ENST00000449969.5': ('ENSG00000149527.18', 'PLCH2', 'protein_coding'),\n",
       " 'ENST00000378486.8': ('ENSG00000149527.18', 'PLCH2', 'protein_coding'),\n",
       " 'ENST00000419816.6': ('ENSG00000149527.18', 'PLCH2', 'protein_coding'),\n",
       " 'ENST00000343889.8': ('ENSG00000149527.18', 'PLCH2', 'protein_coding'),\n",
       " 'ENST00000278878.6': ('ENSG00000149527.18', 'PLCH2', 'protein_coding'),\n",
       " 'ENST00000473964.5': ('ENSG00000149527.18', 'PLCH2', 'protein_coding'),\n",
       " 'ENST00000462379.1': ('ENSG00000149527.18', 'PLCH2', 'protein_coding'),\n",
       " 'ENST00000424657.1': ('ENSG00000224387.1', 'ENSG00000224387', 'lncRNA'),\n",
       " 'ENST00000442305.1': ('ENSG00000229393.1', 'ENSG00000229393', 'lncRNA'),\n",
       " 'ENST00000378466.9': ('ENSG00000157881.16', 'PANK4', 'protein_coding'),\n",
       " 'ENST00000435556.8': ('ENSG00000157881.16', 'PANK4', 'protein_coding'),\n",
       " 'ENST00000505228.5': ('ENSG00000157881.16', 'PANK4', 'protein_coding'),\n",
       " 'ENST00000502512.1': ('ENSG00000157881.16', 'PANK4', 'protein_coding'),\n",
       " 'ENST00000471361.1': ('ENSG00000157881.16', 'PANK4', 'protein_coding'),\n",
       " 'ENST00000487804.1': ('ENSG00000157881.16', 'PANK4', 'protein_coding'),\n",
       " 'ENST00000515423.1': ('ENSG00000157881.16', 'PANK4', 'protein_coding'),\n",
       " 'ENST00000468002.2': ('ENSG00000157881.16', 'PANK4', 'protein_coding'),\n",
       " 'ENST00000486396.1': ('ENSG00000157881.16', 'PANK4', 'protein_coding'),\n",
       " 'ENST00000502770.2': ('ENSG00000157881.16', 'PANK4', 'protein_coding'),\n",
       " 'ENST00000491212.1': ('ENSG00000157881.16', 'PANK4', 'protein_coding'),\n",
       " 'ENST00000514922.1': ('ENSG00000157881.16', 'PANK4', 'protein_coding'),\n",
       " 'ENST00000378453.4': ('ENSG00000197921.6', 'HES5', 'protein_coding'),\n",
       " 'ENST00000644246.1': ('ENSG00000272449.3', 'ENSG00000272449', 'lncRNA'),\n",
       " 'ENST00000606645.3': ('ENSG00000272449.3', 'ENSG00000272449', 'lncRNA'),\n",
       " 'ENST00000449660.5': ('ENSG00000238164.6', 'TNFRSF14-AS1', 'lncRNA'),\n",
       " 'ENST00000443892.2': ('ENSG00000238164.6', 'TNFRSF14-AS1', 'lncRNA'),\n",
       " 'ENST00000416860.3': ('ENSG00000238164.6', 'TNFRSF14-AS1', 'lncRNA'),\n",
       " 'ENST00000448624.2': ('ENSG00000238164.6', 'TNFRSF14-AS1', 'lncRNA'),\n",
       " 'ENST00000432521.2': ('ENSG00000238164.6', 'TNFRSF14-AS1', 'lncRNA'),\n",
       " 'ENST00000452793.1': ('ENSG00000238164.6', 'TNFRSF14-AS1', 'lncRNA'),\n",
       " 'ENST00000426449.5': ('ENSG00000157873.18', 'TNFRSF14', 'protein_coding'),\n",
       " 'ENST00000434817.5': ('ENSG00000157873.18', 'TNFRSF14', 'protein_coding'),\n",
       " 'ENST00000435221.6': ('ENSG00000157873.18', 'TNFRSF14', 'protein_coding'),\n",
       " 'ENST00000451778.5': ('ENSG00000157873.18', 'TNFRSF14', 'protein_coding'),\n",
       " 'ENST00000409119.5': ('ENSG00000157873.18', 'TNFRSF14', 'protein_coding'),\n",
       " 'ENST00000355716.5': ('ENSG00000157873.18', 'TNFRSF14', 'protein_coding'),\n",
       " 'ENST00000475523.5': ('ENSG00000157873.18', 'TNFRSF14', 'protein_coding'),\n",
       " 'ENST00000442392.6': ('ENSG00000157873.18', 'TNFRSF14', 'protein_coding'),\n",
       " 'ENST00000482074.6': ('ENSG00000157873.18', 'TNFRSF14', 'protein_coding'),\n",
       " 'ENST00000496064.5': ('ENSG00000157873.18', 'TNFRSF14', 'protein_coding'),\n",
       " 'ENST00000463471.6': ('ENSG00000157873.18', 'TNFRSF14', 'protein_coding'),\n",
       " 'ENST00000466750.5': ('ENSG00000157873.18', 'TNFRSF14', 'protein_coding'),\n",
       " 'ENST00000471768.1': ('ENSG00000157873.18', 'TNFRSF14', 'protein_coding'),\n",
       " 'ENST00000463835.5': ('ENSG00000157873.18', 'TNFRSF14', 'protein_coding'),\n",
       " 'ENST00000482602.5': ('ENSG00000157873.18', 'TNFRSF14', 'protein_coding'),\n",
       " 'ENST00000480305.1': ('ENSG00000157873.18', 'TNFRSF14', 'protein_coding'),\n",
       " 'ENST00000456687.3': ('ENSG00000225931.3', 'ENSG00000225931', 'TEC'),\n",
       " 'ENST00000424215.1': ('ENSG00000228037.1', 'ENSG00000228037', 'lncRNA'),\n",
       " 'ENST00000686752.2': ('ENSG00000289610.2', 'ENSG00000289610', 'lncRNA'),\n",
       " 'ENST00000493183.5': ('ENSG00000157870.17', 'PRXL2B', 'protein_coding'),\n",
       " 'ENST00000474659.5': ('ENSG00000157870.17', 'PRXL2B', 'protein_coding'),\n",
       " 'ENST00000444521.6': ('ENSG00000157870.17', 'PRXL2B', 'protein_coding'),\n",
       " 'ENST00000378425.9': ('ENSG00000157870.17', 'PRXL2B', 'protein_coding'),\n",
       " 'ENST00000419916.8': ('ENSG00000157870.17', 'PRXL2B', 'protein_coding'),\n",
       " 'ENST00000378424.9': ('ENSG00000157870.17', 'PRXL2B', 'protein_coding'),\n",
       " 'ENST00000537325.6': ('ENSG00000157870.17', 'PRXL2B', 'protein_coding'),\n",
       " 'ENST00000378427.6': ('ENSG00000157870.17', 'PRXL2B', 'protein_coding'),\n",
       " 'ENST00000464043.5': ('ENSG00000157870.17', 'PRXL2B', 'protein_coding'),\n",
       " 'ENST00000465233.6': ('ENSG00000157870.17', 'PRXL2B', 'protein_coding'),\n",
       " 'ENST00000498083.2': ('ENSG00000157870.17', 'PRXL2B', 'protein_coding'),\n",
       " 'ENST00000481683.5': ('ENSG00000157870.17', 'PRXL2B', 'protein_coding'),\n",
       " 'ENST00000484099.5': ('ENSG00000157870.17', 'PRXL2B', 'protein_coding'),\n",
       " 'ENST00000477045.1': ('ENSG00000157870.17', 'PRXL2B', 'protein_coding'),\n",
       " 'ENST00000476686.1': ('ENSG00000157870.17', 'PRXL2B', 'protein_coding'),\n",
       " 'ENST00000471840.5': ('ENSG00000142606.16', 'MMEL1', 'protein_coding'),\n",
       " 'ENST00000378412.8': ('ENSG00000142606.16', 'MMEL1', 'protein_coding'),\n",
       " 'ENST00000502556.5': ('ENSG00000142606.16', 'MMEL1', 'protein_coding'),\n",
       " 'ENST00000504800.5': ('ENSG00000142606.16', 'MMEL1', 'protein_coding'),\n",
       " 'ENST00000491941.5': ('ENSG00000142606.16', 'MMEL1', 'protein_coding'),\n",
       " 'ENST00000464195.1': ('ENSG00000142606.16', 'MMEL1', 'protein_coding'),\n",
       " 'ENST00000469962.1': ('ENSG00000142606.16', 'MMEL1', 'protein_coding'),\n",
       " 'ENST00000509374.1': ('ENSG00000142606.16', 'MMEL1', 'protein_coding'),\n",
       " 'ENST00000511099.1': ('ENSG00000142606.16', 'MMEL1', 'protein_coding'),\n",
       " 'ENST00000427302.1': ('ENSG00000237058.1', 'MMEL1-AS1', 'lncRNA'),\n",
       " 'ENST00000401095.9': ('ENSG00000215912.13', 'TTC34', 'protein_coding'),\n",
       " 'ENST00000637179.1': ('ENSG00000215912.13', 'TTC34', 'protein_coding'),\n",
       " 'ENST00000648684.1': ('ENSG00000285945.1', 'ENSG00000285945', 'lncRNA'),\n",
       " 'ENST00000638002.1': ('ENSG00000283259.1', 'ENSG00000283259', 'lncRNA'),\n",
       " 'ENST00000457440.1': ('ENSG00000233234.1', 'ENSG00000233234', 'lncRNA'),\n",
       " 'ENST00000664694.1': ('ENSG00000287396.1', 'ENSG00000287396', 'lncRNA'),\n",
       " 'ENST00000429389.1': ('ENSG00000231630.1', 'ENSG00000231630', 'lncRNA'),\n",
       " 'ENST00000641229.1': ('ENSG00000284745.1', 'ENSG00000284745', 'lncRNA'),\n",
       " 'ENST00000642073.1': ('ENSG00000284745.1', 'ENSG00000284745', 'lncRNA'),\n",
       " 'ENST00000378404.4': ('ENSG00000169717.7', 'ACTRT2', 'protein_coding'),\n",
       " 'ENST00000687743.1': ('ENSG00000177133.14', 'PRDM16-DT', 'lncRNA'),\n",
       " 'ENST00000606861.1': ('ENSG00000177133.14', 'PRDM16-DT', 'lncRNA'),\n",
       " 'ENST00000321399.7': ('ENSG00000177133.14', 'PRDM16-DT', 'lncRNA'),\n",
       " 'ENST00000413472.5': ('ENSG00000177133.14', 'PRDM16-DT', 'lncRNA'),\n",
       " 'ENST00000453118.5': ('ENSG00000177133.14', 'PRDM16-DT', 'lncRNA'),\n",
       " 'ENST00000445317.1': ('ENSG00000177133.14', 'PRDM16-DT', 'lncRNA'),\n",
       " 'ENST00000667159.1': ('ENSG00000177133.14', 'PRDM16-DT', 'lncRNA'),\n",
       " 'ENST00000670730.1': ('ENSG00000177133.14', 'PRDM16-DT', 'lncRNA'),\n",
       " 'ENST00000415573.1': ('ENSG00000177133.14', 'PRDM16-DT', 'lncRNA'),\n",
       " 'ENST00000321336.1': ('ENSG00000177133.14', 'PRDM16-DT', 'lncRNA'),\n",
       " 'ENST00000702683.1': ('ENSG00000177133.14', 'PRDM16-DT', 'lncRNA'),\n",
       " 'ENST00000511072.5': ('ENSG00000142611.17', 'PRDM16', 'protein_coding'),\n",
       " 'ENST00000607632.1': ('ENSG00000142611.17', 'PRDM16', 'protein_coding'),\n",
       " 'ENST00000378391.6': ('ENSG00000142611.17', 'PRDM16', 'protein_coding'),\n",
       " 'ENST00000270722.10': ('ENSG00000142611.17', 'PRDM16', 'protein_coding'),\n",
       " 'ENST00000514189.5': ('ENSG00000142611.17', 'PRDM16', 'protein_coding'),\n",
       " 'ENST00000512462.5': ('ENSG00000142611.17', 'PRDM16', 'protein_coding'),\n",
       " 'ENST00000463591.1': ('ENSG00000142611.17', 'PRDM16', 'protein_coding'),\n",
       " 'ENST00000509860.1': ('ENSG00000142611.17', 'PRDM16', 'protein_coding'),\n",
       " 'ENST00000378389.5': ('ENSG00000142611.17', 'PRDM16', 'protein_coding'),\n",
       " 'ENST00000606170.1': ('ENSG00000142611.17', 'PRDM16', 'protein_coding'),\n",
       " 'ENST00000635819.1': ('ENSG00000283572.1', 'MIR4251', 'miRNA'),\n",
       " 'ENST00000440516.1': ('ENSG00000226286.1', 'ENSG00000226286', 'lncRNA'),\n",
       " 'ENST00000624175.1': ('ENSG00000279839.1', 'ENSG00000279839', 'TEC'),\n",
       " 'ENST00000607061.1': ('ENSG00000272235.1', 'ENSG00000272235', 'lncRNA'),\n",
       " 'ENST00000669140.1': ('ENSG00000287828.1', 'ENSG00000287828', 'lncRNA'),\n",
       " 'ENST00000665088.1': ('ENSG00000286518.1', 'ENSG00000286518', 'lncRNA'),\n",
       " 'ENST00000378378.9': ('ENSG00000130762.15', 'ARHGEF16', 'protein_coding'),\n",
       " 'ENST00000378373.5': ('ENSG00000130762.15', 'ARHGEF16', 'protein_coding'),\n",
       " 'ENST00000378371.6': ('ENSG00000130762.15', 'ARHGEF16', 'protein_coding'),\n",
       " 'ENST00000445297.5': ('ENSG00000130762.15', 'ARHGEF16', 'protein_coding'),\n",
       " 'ENST00000464620.1': ('ENSG00000130762.15', 'ARHGEF16', 'protein_coding'),\n",
       " 'ENST00000418137.1': ('ENSG00000130762.15', 'ARHGEF16', 'protein_coding'),\n",
       " 'ENST00000485984.1': ('ENSG00000130762.15', 'ARHGEF16', 'protein_coding'),\n",
       " 'ENST00000606489.1': ('ENSG00000272088.1', 'ENSG00000272088', 'lncRNA'),\n",
       " 'ENST00000356575.9': ('ENSG00000162591.17', 'MEGF6', 'protein_coding'),\n",
       " 'ENST00000697102.1': ('ENSG00000162591.17', 'MEGF6', 'protein_coding'),\n",
       " 'ENST00000294599.8': ('ENSG00000162591.17', 'MEGF6', 'protein_coding'),\n",
       " 'ENST00000494257.1': ('ENSG00000162591.17', 'MEGF6', 'protein_coding'),\n",
       " 'ENST00000485002.6': ('ENSG00000162591.17', 'MEGF6', 'protein_coding'),\n",
       " 'ENST00000491842.1': ('ENSG00000162591.17', 'MEGF6', 'protein_coding'),\n",
       " 'ENST00000475790.5': ('ENSG00000162591.17', 'MEGF6', 'protein_coding'),\n",
       " 'ENST00000461795.1': ('ENSG00000162591.17', 'MEGF6', 'protein_coding'),\n",
       " 'ENST00000385042.1': ('ENSG00000207776.1', 'MIR551A', 'miRNA'),\n",
       " 'ENST00000688510.1': ('ENSG00000238260.3', 'ENSG00000238260', 'lncRNA'),\n",
       " 'ENST00000687047.2': ('ENSG00000238260.3', 'ENSG00000238260', 'lncRNA'),\n",
       " 'ENST00000702520.1': ('ENSG00000238260.3', 'ENSG00000238260', 'lncRNA'),\n",
       " 'ENST00000435049.2': ('ENSG00000238260.3', 'ENSG00000238260', 'lncRNA'),\n",
       " 'ENST00000687651.1': ('ENSG00000238260.3', 'ENSG00000238260', 'lncRNA'),\n",
       " 'ENST00000702864.1': ('ENSG00000238260.3', 'ENSG00000238260', 'lncRNA'),\n",
       " 'ENST00000691868.1': ('ENSG00000238260.3', 'ENSG00000238260', 'lncRNA'),\n",
       " 'ENST00000378344.7': ('ENSG00000158109.15', 'TPRG1L', 'protein_coding'),\n",
       " 'ENST00000344579.5': ('ENSG00000158109.15', 'TPRG1L', 'protein_coding'),\n",
       " 'ENST00000378322.7': ('ENSG00000116213.16', 'WRAP73', 'protein_coding'),\n",
       " 'ENST00000270708.12': ('ENSG00000116213.16', 'WRAP73', 'protein_coding'),\n",
       " 'ENST00000471223.1': ('ENSG00000116213.16', 'WRAP73', 'protein_coding'),\n",
       " 'ENST00000424367.5': ('ENSG00000116213.16', 'WRAP73', 'protein_coding'),\n",
       " 'ENST00000497940.5': ('ENSG00000116213.16', 'WRAP73', 'protein_coding'),\n",
       " 'ENST00000469643.1': ('ENSG00000116213.16', 'WRAP73', 'protein_coding'),\n",
       " 'ENST00000494884.1': ('ENSG00000116213.16', 'WRAP73', 'protein_coding'),\n",
       " 'ENST00000465916.5': ('ENSG00000116213.16', 'WRAP73', 'protein_coding'),\n",
       " 'ENST00000419924.2': ('ENSG00000116213.16', 'WRAP73', 'protein_coding'),\n",
       " 'ENST00000479331.1': ('ENSG00000116213.16', 'WRAP73', 'protein_coding'),\n",
       " 'ENST00000378295.9': ('ENSG00000078900.15', 'TP73', 'protein_coding'),\n",
       " 'ENST00000604074.5': ('ENSG00000078900.15', 'TP73', 'protein_coding'),\n",
       " 'ENST00000354437.8': ('ENSG00000078900.15', 'TP73', 'protein_coding'),\n",
       " 'ENST00000346387.8': ('ENSG00000078900.15', 'TP73', 'protein_coding'),\n",
       " 'ENST00000357733.7': ('ENSG00000078900.15', 'TP73', 'protein_coding'),\n",
       " 'ENST00000603362.5': ('ENSG00000078900.15', 'TP73', 'protein_coding'),\n",
       " 'ENST00000604479.5': ('ENSG00000078900.15', 'TP73', 'protein_coding'),\n",
       " 'ENST00000378280.5': ('ENSG00000078900.15', 'TP73', 'protein_coding'),\n",
       " 'ENST00000378285.5': ('ENSG00000078900.15', 'TP73', 'protein_coding'),\n",
       " 'ENST00000378288.8': ('ENSG00000078900.15', 'TP73', 'protein_coding'),\n",
       " 'ENST00000603364.5': ('ENSG00000078900.15', 'TP73', 'protein_coding'),\n",
       " 'ENST00000604194.1': ('ENSG00000078900.15', 'TP73', 'protein_coding'),\n",
       " 'ENST00000378290.4': ('ENSG00000078900.15', 'TP73', 'protein_coding'),\n",
       " 'ENST00000604566.1': ('ENSG00000078900.15', 'TP73', 'protein_coding'),\n",
       " 'ENST00000443034.1': ('ENSG00000227589.1', 'TP73-AS3', 'lncRNA'),\n",
       " 'ENST00000416554.1': ('ENSG00000235131.1', 'TP73-AS2', 'lncRNA'),\n",
       " 'ENST00000648526.1': ('ENSG00000227372.12',\n",
       "  'TP73-AS1',\n",
       "  'transcribed_unitary_pseudogene'),\n",
       " 'ENST00000624167.1': ('ENSG00000227372.12',\n",
       "  'TP73-AS1',\n",
       "  'transcribed_unitary_pseudogene'),\n",
       " 'ENST00000418088.6': ('ENSG00000227372.12',\n",
       "  'TP73-AS1',\n",
       "  'transcribed_unitary_pseudogene'),\n",
       " 'ENST00000648432.1': ('ENSG00000227372.12',\n",
       "  'TP73-AS1',\n",
       "  'transcribed_unitary_pseudogene'),\n",
       " 'ENST00000648479.1': ('ENSG00000227372.12',\n",
       "  'TP73-AS1',\n",
       "  'transcribed_unitary_pseudogene'),\n",
       " 'ENST00000452079.5': ('ENSG00000227372.12',\n",
       "  'TP73-AS1',\n",
       "  'transcribed_unitary_pseudogene'),\n",
       " 'ENST00000650545.1': ('ENSG00000227372.12',\n",
       "  'TP73-AS1',\n",
       "  'transcribed_unitary_pseudogene'),\n",
       " 'ENST00000419973.1': ('ENSG00000227372.12',\n",
       "  'TP73-AS1',\n",
       "  'transcribed_unitary_pseudogene'),\n",
       " ...}"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcripts = parse_transcript_info(GTF_FILE)\n",
    "transcripts['gene_info'] = transcripts.apply(lambda row: (row['gene_id'], row['gene_name'], row['gene_type']), axis=1)\n",
    "transcripts = transcripts.set_index('transcript_id')\n",
    "transcripts = transcripts['gene_info'].to_dict()\n",
    "transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 389738 entries, 0 to 389737\n",
      "Data columns (total 9 columns):\n",
      " #   Column       Non-Null Count   Dtype \n",
      "---  ------       --------------   ----- \n",
      " 0   chrom        389738 non-null  object\n",
      " 1   start        389738 non-null  int64 \n",
      " 2   end          389738 non-null  int64 \n",
      " 3   strand       389738 non-null  object\n",
      " 4   name         389738 non-null  object\n",
      " 5   coords       389738 non-null  object\n",
      " 6   fivep_pos    389738 non-null  int64 \n",
      " 7   annotations  389738 non-null  object\n",
      " 8   adj_exons    389738 non-null  object\n",
      "dtypes: int64(3), object(6)\n",
      "memory usage: 26.8+ MB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>name</th>\n",
       "      <th>coords</th>\n",
       "      <th>fivep_pos</th>\n",
       "      <th>annotations</th>\n",
       "      <th>adj_exons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12721</td>\n",
       "      <td>13220</td>\n",
       "      <td>+</td>\n",
       "      <td>intron;chr1;12721;13220;+;ENST00000456328.2-2</td>\n",
       "      <td>(chr1, 12721, 13220, +)</td>\n",
       "      <td>12721</td>\n",
       "      <td>((ENST00000456328.2, 2),)</td>\n",
       "      <td>(((ENST00000456328.2, 2), (ENST00000456328.2, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30039</td>\n",
       "      <td>30563</td>\n",
       "      <td>+</td>\n",
       "      <td>intron;chr1;30039;30563;+;ENST00000473358.1-1</td>\n",
       "      <td>(chr1, 30039, 30563, +)</td>\n",
       "      <td>30039</td>\n",
       "      <td>((ENST00000473358.1, 1),)</td>\n",
       "      <td>(((ENST00000473358.1, 1), (ENST00000473358.1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30667</td>\n",
       "      <td>30975</td>\n",
       "      <td>+</td>\n",
       "      <td>intron;chr1;30667;30975;+;ENST00000469289.1-1|...</td>\n",
       "      <td>(chr1, 30667, 30975, +)</td>\n",
       "      <td>30667</td>\n",
       "      <td>((ENST00000469289.1, 1), (ENST00000473358.1, 2))</td>\n",
       "      <td>(((ENST00000469289.1, 1), (ENST00000469289.1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35174</td>\n",
       "      <td>35276</td>\n",
       "      <td>-</td>\n",
       "      <td>intron;chr1;35174;35276;-;ENST00000417324.1-2</td>\n",
       "      <td>(chr1, 35174, 35276, -)</td>\n",
       "      <td>35275</td>\n",
       "      <td>((ENST00000417324.1, 2),)</td>\n",
       "      <td>(((ENST00000417324.1, 2), (ENST00000417324.1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35481</td>\n",
       "      <td>35720</td>\n",
       "      <td>-</td>\n",
       "      <td>intron;chr1;35481;35720;-;ENST00000417324.1-1|...</td>\n",
       "      <td>(chr1, 35481, 35720, -)</td>\n",
       "      <td>35719</td>\n",
       "      <td>((ENST00000417324.1, 1), (ENST00000461467.1, 1))</td>\n",
       "      <td>(((ENST00000417324.1, 1), (ENST00000417324.1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389733</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209733</td>\n",
       "      <td>57209821</td>\n",
       "      <td>+</td>\n",
       "      <td>intron;chrY;57209733;57209821;+;ENST0000071127...</td>\n",
       "      <td>(chrY, 57209733, 57209821, +)</td>\n",
       "      <td>57209733</td>\n",
       "      <td>((ENST00000711272.1, 6), (ENST00000711275.1, 5...</td>\n",
       "      <td>(((ENST00000711272.1, 6), (ENST00000711272.1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389734</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209980</td>\n",
       "      <td>57210639</td>\n",
       "      <td>+</td>\n",
       "      <td>intron;chrY;57209980;57210639;+;ENST0000071127...</td>\n",
       "      <td>(chrY, 57209980, 57210639, +)</td>\n",
       "      <td>57209980</td>\n",
       "      <td>((ENST00000711272.1, 7), (ENST00000711276.1, 5...</td>\n",
       "      <td>(((ENST00000711272.1, 7), (ENST00000711272.1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389735</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57210792</td>\n",
       "      <td>57211551</td>\n",
       "      <td>+</td>\n",
       "      <td>intron;chrY;57210792;57211551;+;ENST0000071127...</td>\n",
       "      <td>(chrY, 57210792, 57211551, +)</td>\n",
       "      <td>57210792</td>\n",
       "      <td>((ENST00000711272.1, 8), (ENST00000711276.1, 6...</td>\n",
       "      <td>(((ENST00000711272.1, 8), (ENST00000711272.1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389736</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57211569</td>\n",
       "      <td>57211760</td>\n",
       "      <td>+</td>\n",
       "      <td>intron;chrY;57211569;57211760;+;ENST0000071128...</td>\n",
       "      <td>(chrY, 57211569, 57211760, +)</td>\n",
       "      <td>57211569</td>\n",
       "      <td>((ENST00000711285.1, 2),)</td>\n",
       "      <td>(((ENST00000711285.1, 2), (ENST00000711285.1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389737</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57211620</td>\n",
       "      <td>57211760</td>\n",
       "      <td>+</td>\n",
       "      <td>intron;chrY;57211620;57211760;+;ENST0000071127...</td>\n",
       "      <td>(chrY, 57211620, 57211760, +)</td>\n",
       "      <td>57211620</td>\n",
       "      <td>((ENST00000711272.1, 9), (ENST00000711276.1, 7...</td>\n",
       "      <td>(((ENST00000711272.1, 9), (ENST00000711272.1, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>389738 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom     start       end strand  \\\n",
       "0       chr1     12721     13220      +   \n",
       "1       chr1     30039     30563      +   \n",
       "2       chr1     30667     30975      +   \n",
       "3       chr1     35174     35276      -   \n",
       "4       chr1     35481     35720      -   \n",
       "...      ...       ...       ...    ...   \n",
       "389733  chrY  57209733  57209821      +   \n",
       "389734  chrY  57209980  57210639      +   \n",
       "389735  chrY  57210792  57211551      +   \n",
       "389736  chrY  57211569  57211760      +   \n",
       "389737  chrY  57211620  57211760      +   \n",
       "\n",
       "                                                     name  \\\n",
       "0           intron;chr1;12721;13220;+;ENST00000456328.2-2   \n",
       "1           intron;chr1;30039;30563;+;ENST00000473358.1-1   \n",
       "2       intron;chr1;30667;30975;+;ENST00000469289.1-1|...   \n",
       "3           intron;chr1;35174;35276;-;ENST00000417324.1-2   \n",
       "4       intron;chr1;35481;35720;-;ENST00000417324.1-1|...   \n",
       "...                                                   ...   \n",
       "389733  intron;chrY;57209733;57209821;+;ENST0000071127...   \n",
       "389734  intron;chrY;57209980;57210639;+;ENST0000071127...   \n",
       "389735  intron;chrY;57210792;57211551;+;ENST0000071127...   \n",
       "389736  intron;chrY;57211569;57211760;+;ENST0000071128...   \n",
       "389737  intron;chrY;57211620;57211760;+;ENST0000071127...   \n",
       "\n",
       "                               coords  fivep_pos  \\\n",
       "0             (chr1, 12721, 13220, +)      12721   \n",
       "1             (chr1, 30039, 30563, +)      30039   \n",
       "2             (chr1, 30667, 30975, +)      30667   \n",
       "3             (chr1, 35174, 35276, -)      35275   \n",
       "4             (chr1, 35481, 35720, -)      35719   \n",
       "...                               ...        ...   \n",
       "389733  (chrY, 57209733, 57209821, +)   57209733   \n",
       "389734  (chrY, 57209980, 57210639, +)   57209980   \n",
       "389735  (chrY, 57210792, 57211551, +)   57210792   \n",
       "389736  (chrY, 57211569, 57211760, +)   57211569   \n",
       "389737  (chrY, 57211620, 57211760, +)   57211620   \n",
       "\n",
       "                                              annotations  \\\n",
       "0                               ((ENST00000456328.2, 2),)   \n",
       "1                               ((ENST00000473358.1, 1),)   \n",
       "2        ((ENST00000469289.1, 1), (ENST00000473358.1, 2))   \n",
       "3                               ((ENST00000417324.1, 2),)   \n",
       "4        ((ENST00000417324.1, 1), (ENST00000461467.1, 1))   \n",
       "...                                                   ...   \n",
       "389733  ((ENST00000711272.1, 6), (ENST00000711275.1, 5...   \n",
       "389734  ((ENST00000711272.1, 7), (ENST00000711276.1, 5...   \n",
       "389735  ((ENST00000711272.1, 8), (ENST00000711276.1, 6...   \n",
       "389736                          ((ENST00000711285.1, 2),)   \n",
       "389737  ((ENST00000711272.1, 9), (ENST00000711276.1, 7...   \n",
       "\n",
       "                                                adj_exons  \n",
       "0       (((ENST00000456328.2, 2), (ENST00000456328.2, ...  \n",
       "1       (((ENST00000473358.1, 1), (ENST00000473358.1, ...  \n",
       "2       (((ENST00000469289.1, 1), (ENST00000469289.1, ...  \n",
       "3       (((ENST00000417324.1, 2), (ENST00000417324.1, ...  \n",
       "4       (((ENST00000417324.1, 1), (ENST00000417324.1, ...  \n",
       "...                                                   ...  \n",
       "389733  (((ENST00000711272.1, 6), (ENST00000711272.1, ...  \n",
       "389734  (((ENST00000711272.1, 7), (ENST00000711272.1, ...  \n",
       "389735  (((ENST00000711272.1, 8), (ENST00000711272.1, ...  \n",
       "389736  (((ENST00000711285.1, 2), (ENST00000711285.1, ...  \n",
       "389737  (((ENST00000711272.1, 9), (ENST00000711272.1, ...  \n",
       "\n",
       "[389738 rows x 9 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "introns = functions.load_bed(INTRONS_BED)\n",
    "introns = introns.rename(columns={'feat': 'chrom'})\n",
    "\n",
    "introns = introns.groupby(['chrom', 'start', 'end', 'strand'], as_index=False).name.agg(','.join)\n",
    "introns['coords'] = introns[['chrom', 'start', 'end', 'strand']].apply(lambda row: tuple(row.to_list()), axis=1)\n",
    "introns['fivep_pos'] = introns.apply(lambda row: row['start'] if row['strand']=='+' else row['end']-1, axis=1)\n",
    "introns['annotations'] = introns.name.transform(lambda name: parse_feature_name(name)[-1])\n",
    "introns['adj_exons'] = introns.annotations.transform(lambda annos: tuple([ ((tid, num), (tid, num+1)) for tid, num in annos]))\n",
    "# introns['interval'] = introns.apply(lambda row: Interval(row['start'], row['end']), axis=1)\n",
    "print(introns.info())\n",
    "introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>name</th>\n",
       "      <th>coords</th>\n",
       "      <th>fivep_pos</th>\n",
       "      <th>annotations</th>\n",
       "      <th>adj_exons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12721</td>\n",
       "      <td>13220</td>\n",
       "      <td>+</td>\n",
       "      <td>intron;chr1;12721;13220;+;ENST00000456328.2-2</td>\n",
       "      <td>(chr1, 12721, 13220, +)</td>\n",
       "      <td>12721</td>\n",
       "      <td>((ENST00000456328.2, 2),)</td>\n",
       "      <td>((ENST00000456328.2, 2), (ENST00000456328.2, 3))</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30039</td>\n",
       "      <td>30563</td>\n",
       "      <td>+</td>\n",
       "      <td>intron;chr1;30039;30563;+;ENST00000473358.1-1</td>\n",
       "      <td>(chr1, 30039, 30563, +)</td>\n",
       "      <td>30039</td>\n",
       "      <td>((ENST00000473358.1, 1),)</td>\n",
       "      <td>((ENST00000473358.1, 1), (ENST00000473358.1, 2))</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30667</td>\n",
       "      <td>30975</td>\n",
       "      <td>+</td>\n",
       "      <td>intron;chr1;30667;30975;+;ENST00000469289.1-1|...</td>\n",
       "      <td>(chr1, 30667, 30975, +)</td>\n",
       "      <td>30667</td>\n",
       "      <td>((ENST00000469289.1, 1), (ENST00000473358.1, 2))</td>\n",
       "      <td>((ENST00000469289.1, 1), (ENST00000469289.1, 2))</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30667</td>\n",
       "      <td>30975</td>\n",
       "      <td>+</td>\n",
       "      <td>intron;chr1;30667;30975;+;ENST00000469289.1-1|...</td>\n",
       "      <td>(chr1, 30667, 30975, +)</td>\n",
       "      <td>30667</td>\n",
       "      <td>((ENST00000469289.1, 1), (ENST00000473358.1, 2))</td>\n",
       "      <td>((ENST00000473358.1, 2), (ENST00000473358.1, 3))</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35174</td>\n",
       "      <td>35276</td>\n",
       "      <td>-</td>\n",
       "      <td>intron;chr1;35174;35276;-;ENST00000417324.1-2</td>\n",
       "      <td>(chr1, 35174, 35276, -)</td>\n",
       "      <td>35275</td>\n",
       "      <td>((ENST00000417324.1, 2),)</td>\n",
       "      <td>((ENST00000417324.1, 2), (ENST00000417324.1, 3))</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378153</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57211620</td>\n",
       "      <td>57211760</td>\n",
       "      <td>+</td>\n",
       "      <td>intron;chrY;57211620;57211760;+;ENST0000071127...</td>\n",
       "      <td>(chrY, 57211620, 57211760, +)</td>\n",
       "      <td>57211620</td>\n",
       "      <td>((ENST00000711272.1, 9), (ENST00000711276.1, 7...</td>\n",
       "      <td>((ENST00000711272.1, 9), (ENST00000711272.1, 10))</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378154</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57211620</td>\n",
       "      <td>57211760</td>\n",
       "      <td>+</td>\n",
       "      <td>intron;chrY;57211620;57211760;+;ENST0000071127...</td>\n",
       "      <td>(chrY, 57211620, 57211760, +)</td>\n",
       "      <td>57211620</td>\n",
       "      <td>((ENST00000711272.1, 9), (ENST00000711276.1, 7...</td>\n",
       "      <td>((ENST00000711276.1, 7), (ENST00000711276.1, 8))</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378155</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57211620</td>\n",
       "      <td>57211760</td>\n",
       "      <td>+</td>\n",
       "      <td>intron;chrY;57211620;57211760;+;ENST0000071127...</td>\n",
       "      <td>(chrY, 57211620, 57211760, +)</td>\n",
       "      <td>57211620</td>\n",
       "      <td>((ENST00000711272.1, 9), (ENST00000711276.1, 7...</td>\n",
       "      <td>((ENST00000711277.1, 5), (ENST00000711277.1, 6))</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378156</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57211620</td>\n",
       "      <td>57211760</td>\n",
       "      <td>+</td>\n",
       "      <td>intron;chrY;57211620;57211760;+;ENST0000071127...</td>\n",
       "      <td>(chrY, 57211620, 57211760, +)</td>\n",
       "      <td>57211620</td>\n",
       "      <td>((ENST00000711272.1, 9), (ENST00000711276.1, 7...</td>\n",
       "      <td>((ENST00000711278.1, 7), (ENST00000711278.1, 8))</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378157</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57211620</td>\n",
       "      <td>57211760</td>\n",
       "      <td>+</td>\n",
       "      <td>intron;chrY;57211620;57211760;+;ENST0000071127...</td>\n",
       "      <td>(chrY, 57211620, 57211760, +)</td>\n",
       "      <td>57211620</td>\n",
       "      <td>((ENST00000711272.1, 9), (ENST00000711276.1, 7...</td>\n",
       "      <td>((ENST00000711284.1, 2), (ENST00000711284.1, 3))</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1378158 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom     start       end strand  \\\n",
       "0        chr1     12721     13220      +   \n",
       "1        chr1     30039     30563      +   \n",
       "2        chr1     30667     30975      +   \n",
       "3        chr1     30667     30975      +   \n",
       "4        chr1     35174     35276      -   \n",
       "...       ...       ...       ...    ...   \n",
       "1378153  chrY  57211620  57211760      +   \n",
       "1378154  chrY  57211620  57211760      +   \n",
       "1378155  chrY  57211620  57211760      +   \n",
       "1378156  chrY  57211620  57211760      +   \n",
       "1378157  chrY  57211620  57211760      +   \n",
       "\n",
       "                                                      name  \\\n",
       "0            intron;chr1;12721;13220;+;ENST00000456328.2-2   \n",
       "1            intron;chr1;30039;30563;+;ENST00000473358.1-1   \n",
       "2        intron;chr1;30667;30975;+;ENST00000469289.1-1|...   \n",
       "3        intron;chr1;30667;30975;+;ENST00000469289.1-1|...   \n",
       "4            intron;chr1;35174;35276;-;ENST00000417324.1-2   \n",
       "...                                                    ...   \n",
       "1378153  intron;chrY;57211620;57211760;+;ENST0000071127...   \n",
       "1378154  intron;chrY;57211620;57211760;+;ENST0000071127...   \n",
       "1378155  intron;chrY;57211620;57211760;+;ENST0000071127...   \n",
       "1378156  intron;chrY;57211620;57211760;+;ENST0000071127...   \n",
       "1378157  intron;chrY;57211620;57211760;+;ENST0000071127...   \n",
       "\n",
       "                                coords  fivep_pos  \\\n",
       "0              (chr1, 12721, 13220, +)      12721   \n",
       "1              (chr1, 30039, 30563, +)      30039   \n",
       "2              (chr1, 30667, 30975, +)      30667   \n",
       "3              (chr1, 30667, 30975, +)      30667   \n",
       "4              (chr1, 35174, 35276, -)      35275   \n",
       "...                                ...        ...   \n",
       "1378153  (chrY, 57211620, 57211760, +)   57211620   \n",
       "1378154  (chrY, 57211620, 57211760, +)   57211620   \n",
       "1378155  (chrY, 57211620, 57211760, +)   57211620   \n",
       "1378156  (chrY, 57211620, 57211760, +)   57211620   \n",
       "1378157  (chrY, 57211620, 57211760, +)   57211620   \n",
       "\n",
       "                                               annotations  \\\n",
       "0                                ((ENST00000456328.2, 2),)   \n",
       "1                                ((ENST00000473358.1, 1),)   \n",
       "2         ((ENST00000469289.1, 1), (ENST00000473358.1, 2))   \n",
       "3         ((ENST00000469289.1, 1), (ENST00000473358.1, 2))   \n",
       "4                                ((ENST00000417324.1, 2),)   \n",
       "...                                                    ...   \n",
       "1378153  ((ENST00000711272.1, 9), (ENST00000711276.1, 7...   \n",
       "1378154  ((ENST00000711272.1, 9), (ENST00000711276.1, 7...   \n",
       "1378155  ((ENST00000711272.1, 9), (ENST00000711276.1, 7...   \n",
       "1378156  ((ENST00000711272.1, 9), (ENST00000711276.1, 7...   \n",
       "1378157  ((ENST00000711272.1, 9), (ENST00000711276.1, 7...   \n",
       "\n",
       "                                                 adj_exons  \n",
       "0         ((ENST00000456328.2, 2), (ENST00000456328.2, 3))  \n",
       "1         ((ENST00000473358.1, 1), (ENST00000473358.1, 2))  \n",
       "2         ((ENST00000469289.1, 1), (ENST00000469289.1, 2))  \n",
       "3         ((ENST00000473358.1, 2), (ENST00000473358.1, 3))  \n",
       "4         ((ENST00000417324.1, 2), (ENST00000417324.1, 3))  \n",
       "...                                                    ...  \n",
       "1378153  ((ENST00000711272.1, 9), (ENST00000711272.1, 10))  \n",
       "1378154   ((ENST00000711276.1, 7), (ENST00000711276.1, 8))  \n",
       "1378155   ((ENST00000711277.1, 5), (ENST00000711277.1, 6))  \n",
       "1378156   ((ENST00000711278.1, 7), (ENST00000711278.1, 8))  \n",
       "1378157   ((ENST00000711284.1, 2), (ENST00000711284.1, 3))  \n",
       "\n",
       "[1378158 rows x 9 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exploded_introns = introns.explode('adj_exons', ignore_index=True)\n",
    "exploded_introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>align_start</th>\n",
       "      <th>align_end</th>\n",
       "      <th>read_id</th>\n",
       "      <th>quality</th>\n",
       "      <th>align_orient</th>\n",
       "      <th>n_segments</th>\n",
       "      <th>block_lengths</th>\n",
       "      <th>block_starts</th>\n",
       "      <th>features</th>\n",
       "      <th>align_loc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>73249031</td>\n",
       "      <td>73249180</td>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:16459:24298</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>149,</td>\n",
       "      <td>0,</td>\n",
       "      <td>intron;chr2;73249179;73249818;+;ENST0000025809...</td>\n",
       "      <td>chr2_73249031_73249180_+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>73249031</td>\n",
       "      <td>73249180</td>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:16459:24298</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>149,</td>\n",
       "      <td>0,</td>\n",
       "      <td>exon;chr2;73248990;73249061;+;ENST00000488856.1-4</td>\n",
       "      <td>chr2_73249031_73249180_+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>73249031</td>\n",
       "      <td>73249180</td>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:16459:24298</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>149,</td>\n",
       "      <td>0,</td>\n",
       "      <td>exon;chr2;73248990;73249179;+;ENST00000540468....</td>\n",
       "      <td>chr2_73249031_73249180_+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr19</td>\n",
       "      <td>40284680</td>\n",
       "      <td>40284817</td>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:16984:24298</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>137,</td>\n",
       "      <td>0,</td>\n",
       "      <td>intron;chr19;40282304;40284863;-;ENST000004564...</td>\n",
       "      <td>chr19_40284680_40284817_+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr19</td>\n",
       "      <td>40284680</td>\n",
       "      <td>40284817</td>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:16984:24298</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>137,</td>\n",
       "      <td>0,</td>\n",
       "      <td>intron;chr19;40282776;40284764;-;ENST000004231...</td>\n",
       "      <td>chr19_40284680_40284817_+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340111</th>\n",
       "      <td>chr5</td>\n",
       "      <td>133959933</td>\n",
       "      <td>133968508</td>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:14977:24298</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>87,63,</td>\n",
       "      <td>0,8512,</td>\n",
       "      <td>intron;chr5;133960020;133968525;-;ENST00000507...</td>\n",
       "      <td>chr5_133959933_133968508_+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340112</th>\n",
       "      <td>chr5</td>\n",
       "      <td>133959933</td>\n",
       "      <td>133968508</td>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:14977:24298</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>87,63,</td>\n",
       "      <td>0,8512,</td>\n",
       "      <td>exon;chr5;133959493;133960020;-;ENST0000023151...</td>\n",
       "      <td>chr5_133959933_133968508_+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340113</th>\n",
       "      <td>chr5</td>\n",
       "      <td>133959933</td>\n",
       "      <td>133968508</td>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:14977:24298</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>87,63,</td>\n",
       "      <td>0,8512,</td>\n",
       "      <td>exon;chr5;133959534;133960020;-;ENST0000050719...</td>\n",
       "      <td>chr5_133959933_133968508_+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340114</th>\n",
       "      <td>chr5</td>\n",
       "      <td>133959933</td>\n",
       "      <td>133968508</td>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:14977:24298</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>87,63,</td>\n",
       "      <td>0,8512,</td>\n",
       "      <td>exon;chr5;133959777;133960020;-;ENST0000050991...</td>\n",
       "      <td>chr5_133959933_133968508_+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340115</th>\n",
       "      <td>chr5</td>\n",
       "      <td>133959933</td>\n",
       "      <td>133968508</td>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:14977:24298</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>87,63,</td>\n",
       "      <td>0,8512,</td>\n",
       "      <td>exon;chr5;133968445;133968625;-;ENST0000023151...</td>\n",
       "      <td>chr5_133959933_133968508_+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>340116 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom  align_start  align_end  \\\n",
       "0        chr2     73249031   73249180   \n",
       "1        chr2     73249031   73249180   \n",
       "2        chr2     73249031   73249180   \n",
       "3       chr19     40284680   40284817   \n",
       "4       chr19     40284680   40284817   \n",
       "...       ...          ...        ...   \n",
       "340111   chr5    133959933  133968508   \n",
       "340112   chr5    133959933  133968508   \n",
       "340113   chr5    133959933  133968508   \n",
       "340114   chr5    133959933  133968508   \n",
       "340115   chr5    133959933  133968508   \n",
       "\n",
       "                                               read_id  quality align_orient  \\\n",
       "0       NGSNJ-086:229:GW200110425th:1:1101:16459:24298      255            +   \n",
       "1       NGSNJ-086:229:GW200110425th:1:1101:16459:24298      255            +   \n",
       "2       NGSNJ-086:229:GW200110425th:1:1101:16459:24298      255            +   \n",
       "3       NGSNJ-086:229:GW200110425th:1:1101:16984:24298      255            +   \n",
       "4       NGSNJ-086:229:GW200110425th:1:1101:16984:24298      255            +   \n",
       "...                                                ...      ...          ...   \n",
       "340111  NGSNJ-086:229:GW200110425th:1:1101:14977:24298      255            +   \n",
       "340112  NGSNJ-086:229:GW200110425th:1:1101:14977:24298      255            +   \n",
       "340113  NGSNJ-086:229:GW200110425th:1:1101:14977:24298      255            +   \n",
       "340114  NGSNJ-086:229:GW200110425th:1:1101:14977:24298      255            +   \n",
       "340115  NGSNJ-086:229:GW200110425th:1:1101:14977:24298      255            +   \n",
       "\n",
       "        n_segments block_lengths block_starts  \\\n",
       "0                1          149,           0,   \n",
       "1                1          149,           0,   \n",
       "2                1          149,           0,   \n",
       "3                1          137,           0,   \n",
       "4                1          137,           0,   \n",
       "...            ...           ...          ...   \n",
       "340111           2        87,63,      0,8512,   \n",
       "340112           2        87,63,      0,8512,   \n",
       "340113           2        87,63,      0,8512,   \n",
       "340114           2        87,63,      0,8512,   \n",
       "340115           2        87,63,      0,8512,   \n",
       "\n",
       "                                                 features  \\\n",
       "0       intron;chr2;73249179;73249818;+;ENST0000025809...   \n",
       "1       exon;chr2;73248990;73249061;+;ENST00000488856.1-4   \n",
       "2       exon;chr2;73248990;73249179;+;ENST00000540468....   \n",
       "3       intron;chr19;40282304;40284863;-;ENST000004564...   \n",
       "4       intron;chr19;40282776;40284764;-;ENST000004231...   \n",
       "...                                                   ...   \n",
       "340111  intron;chr5;133960020;133968525;-;ENST00000507...   \n",
       "340112  exon;chr5;133959493;133960020;-;ENST0000023151...   \n",
       "340113  exon;chr5;133959534;133960020;-;ENST0000050719...   \n",
       "340114  exon;chr5;133959777;133960020;-;ENST0000050991...   \n",
       "340115  exon;chr5;133968445;133968625;-;ENST0000023151...   \n",
       "\n",
       "                         align_loc  \n",
       "0         chr2_73249031_73249180_+  \n",
       "1         chr2_73249031_73249180_+  \n",
       "2         chr2_73249031_73249180_+  \n",
       "3        chr19_40284680_40284817_+  \n",
       "4        chr19_40284680_40284817_+  \n",
       "...                            ...  \n",
       "340111  chr5_133959933_133968508_+  \n",
       "340112  chr5_133959933_133968508_+  \n",
       "340113  chr5_133959933_133968508_+  \n",
       "340114  chr5_133959933_133968508_+  \n",
       "340115  chr5_133959933_133968508_+  \n",
       "\n",
       "[340116 rows x 11 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "star_aligns = pd.read_csv(f'{out_dir}/STAR/overlaps.bed', sep='\\t', header=None)\n",
    "star_aligns.columns = ['chrom', 'align_start', 'align_end', 'read_id', 'quality', 'align_orient', '_', '_', '_', 'n_segments', 'block_lengths', 'block_starts', '_', '_', '_', '_', 'features', '_', '_']\n",
    "star_aligns = star_aligns.drop(columns='_')\n",
    "# No multimapping\n",
    "star_aligns['align_loc'] = star_aligns.apply(lambda row: f\"{row['chrom']}_{row['align_start']}_{row['align_end']}_{row['align_orient']}\", axis=1)\n",
    "assert star_aligns.groupby('read_id').align_loc.apply(lambda id_: id_.nunique()==1).all()\n",
    "star_aligns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(('exon', 'chr22', 43161051, 43161190, '+', (('ENST00000583777.5', 2), ('ENST00000337554.8', 3), ('ENST00000428336.5', 3), ('ENST00000329563.8', 3), ('ENST00000396265.4', 3)), ('protein_coding',)), ('exon', 'chr22', 43162802, 43163210, '+', (('ENST00000329563.8', 4),), ('protein_coding',)), ('exon', 'chr22', 43162802, 43163241, '+', (('ENST00000583777.5', 3), ('ENST00000396265.4', 4)), ('protein_coding',)), ('exon', 'chr22', 43162802, 43163242, '+', (('ENST00000337554.8', 4),), ('protein_coding',)))\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>align_start</th>\n",
       "      <th>align_end</th>\n",
       "      <th>align_orient</th>\n",
       "      <th>quality</th>\n",
       "      <th>n_segments</th>\n",
       "      <th>block_lengths</th>\n",
       "      <th>block_starts</th>\n",
       "      <th>features</th>\n",
       "      <th>segment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>43161160</td>\n",
       "      <td>43162922</td>\n",
       "      <td>-</td>\n",
       "      <td>255</td>\n",
       "      <td>2</td>\n",
       "      <td>30,120,</td>\n",
       "      <td>0,1642,</td>\n",
       "      <td>((exon, chr22, 43161051, 43161190, +, (('ENST0...</td>\n",
       "      <td>((43161160, 43161190, None), (43162802, 431629...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11992</td>\n",
       "      <td>chr9</td>\n",
       "      <td>9442096</td>\n",
       "      <td>9442246</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>150,</td>\n",
       "      <td>0,</td>\n",
       "      <td>((exon, chr9, 9442059, 9442380, +, (('ENST0000...</td>\n",
       "      <td>((9442096, 9442246, None),)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:12085</td>\n",
       "      <td>chr14</td>\n",
       "      <td>49586699</td>\n",
       "      <td>49586849</td>\n",
       "      <td>-</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>150,</td>\n",
       "      <td>0,</td>\n",
       "      <td>((exon, chr14, 49586579, 49586878, +, (('ENST0...</td>\n",
       "      <td>((49586699, 49586849, None),)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:12117</td>\n",
       "      <td>chr14</td>\n",
       "      <td>49862679</td>\n",
       "      <td>49862792</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>113,</td>\n",
       "      <td>0,</td>\n",
       "      <td>((intron, chr14, 49862316, 49863828, -, (('ENS...</td>\n",
       "      <td>((49862679, 49862792, None),)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:13244</td>\n",
       "      <td>chr14</td>\n",
       "      <td>49862600</td>\n",
       "      <td>49862750</td>\n",
       "      <td>-</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>150,</td>\n",
       "      <td>0,</td>\n",
       "      <td>((intron, chr14, 49862316, 49863828, -, (('ENS...</td>\n",
       "      <td>((49862600, 49862750, None),)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83001</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:27352</td>\n",
       "      <td>chr14</td>\n",
       "      <td>32163959</td>\n",
       "      <td>32164109</td>\n",
       "      <td>-</td>\n",
       "      <td>255</td>\n",
       "      <td>1</td>\n",
       "      <td>150,</td>\n",
       "      <td>0,</td>\n",
       "      <td>(.,)</td>\n",
       "      <td>((32163959, 32164109, None),)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83002</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:3928</td>\n",
       "      <td>chr11</td>\n",
       "      <td>65546991</td>\n",
       "      <td>65547095</td>\n",
       "      <td>+</td>\n",
       "      <td>255</td>\n",
       "      <td>1</td>\n",
       "      <td>104,</td>\n",
       "      <td>0,</td>\n",
       "      <td>((intron, chr11, 65546920, 65547438, -, (('ENS...</td>\n",
       "      <td>((65546991, 65547095, None),)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83003</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>48233331</td>\n",
       "      <td>48253811</td>\n",
       "      <td>-</td>\n",
       "      <td>255</td>\n",
       "      <td>2</td>\n",
       "      <td>146,4,</td>\n",
       "      <td>0,20476,</td>\n",
       "      <td>((exon, chr13, 48233202, 48233477, +, (('ENST0...</td>\n",
       "      <td>((48233331, 48233477, None), (48253807, 482538...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83004</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:5807</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114750152</td>\n",
       "      <td>114757991</td>\n",
       "      <td>-</td>\n",
       "      <td>255</td>\n",
       "      <td>2</td>\n",
       "      <td>55,67,</td>\n",
       "      <td>0,7772,</td>\n",
       "      <td>((exon, chr1, 114749816, 114750207, -, (('ENST...</td>\n",
       "      <td>((114750152, 114750207, None), (114757924, 114...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83005</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>49586704</td>\n",
       "      <td>49586854</td>\n",
       "      <td>-</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>150,</td>\n",
       "      <td>0,</td>\n",
       "      <td>((exon, chr14, 49586579, 49586878, +, (('ENST0...</td>\n",
       "      <td>((49586704, 49586854, None),)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>83006 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              read_id  chrom  align_start  \\\n",
       "0      NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22     43161160   \n",
       "1      NGSNJ-086:229:GW200110425th:1:1101:10004:11992   chr9      9442096   \n",
       "2      NGSNJ-086:229:GW200110425th:1:1101:10004:12085  chr14     49586699   \n",
       "3      NGSNJ-086:229:GW200110425th:1:1101:10004:12117  chr14     49862679   \n",
       "4      NGSNJ-086:229:GW200110425th:1:1101:10004:13244  chr14     49862600   \n",
       "...                                               ...    ...          ...   \n",
       "83001   NGSNJ-086:229:GW200110425th:1:1101:9995:27352  chr14     32163959   \n",
       "83002    NGSNJ-086:229:GW200110425th:1:1101:9995:3928  chr11     65546991   \n",
       "83003    NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13     48233331   \n",
       "83004    NGSNJ-086:229:GW200110425th:1:1101:9995:5807   chr1    114750152   \n",
       "83005    NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14     49586704   \n",
       "\n",
       "       align_end align_orient  quality  n_segments block_lengths block_starts  \\\n",
       "0       43162922            -      255           2       30,120,      0,1642,   \n",
       "1        9442246            +        3           1          150,           0,   \n",
       "2       49586849            -        3           1          150,           0,   \n",
       "3       49862792            +        3           1          113,           0,   \n",
       "4       49862750            -        3           1          150,           0,   \n",
       "...          ...          ...      ...         ...           ...          ...   \n",
       "83001   32164109            -      255           1          150,           0,   \n",
       "83002   65547095            +      255           1          104,           0,   \n",
       "83003   48253811            -      255           2        146,4,     0,20476,   \n",
       "83004  114757991            -      255           2        55,67,      0,7772,   \n",
       "83005   49586854            -        3           1          150,           0,   \n",
       "\n",
       "                                                features  \\\n",
       "0      ((exon, chr22, 43161051, 43161190, +, (('ENST0...   \n",
       "1      ((exon, chr9, 9442059, 9442380, +, (('ENST0000...   \n",
       "2      ((exon, chr14, 49586579, 49586878, +, (('ENST0...   \n",
       "3      ((intron, chr14, 49862316, 49863828, -, (('ENS...   \n",
       "4      ((intron, chr14, 49862316, 49863828, -, (('ENS...   \n",
       "...                                                  ...   \n",
       "83001                                               (.,)   \n",
       "83002  ((intron, chr11, 65546920, 65547438, -, (('ENS...   \n",
       "83003  ((exon, chr13, 48233202, 48233477, +, (('ENST0...   \n",
       "83004  ((exon, chr1, 114749816, 114750207, -, (('ENST...   \n",
       "83005  ((exon, chr14, 49586579, 49586878, +, (('ENST0...   \n",
       "\n",
       "                                                 segment  \n",
       "0      ((43161160, 43161190, None), (43162802, 431629...  \n",
       "1                            ((9442096, 9442246, None),)  \n",
       "2                          ((49586699, 49586849, None),)  \n",
       "3                          ((49862679, 49862792, None),)  \n",
       "4                          ((49862600, 49862750, None),)  \n",
       "...                                                  ...  \n",
       "83001                      ((32163959, 32164109, None),)  \n",
       "83002                      ((65546991, 65547095, None),)  \n",
       "83003  ((48233331, 48233477, None), (48253807, 482538...  \n",
       "83004  ((114750152, 114750207, None), (114757924, 114...  \n",
       "83005                      ((49586704, 49586854, None),)  \n",
       "\n",
       "[83006 rows x 11 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def infer_segments(row):\n",
    "\tstarts = row['block_starts'].split(',')[:-1]\n",
    "\tstarts = [row['align_start']+int(start) for start in starts]\n",
    "\tlengths = row['block_lengths'].split(',')[:-1]\n",
    "\tlengths = [int(length) for length in lengths]\n",
    "\n",
    "\tsegments = []\n",
    "\tfor i in range(len(starts)):\n",
    "\t\tsegment_start = starts[i]\n",
    "\t\tsegment_end = starts[i] + lengths[i] \n",
    "\t\tsegments.append(Interval(segment_start, segment_end))\n",
    "\n",
    "\treturn tuple(segments)\n",
    "\n",
    "\n",
    "\n",
    "star_aligns.features = star_aligns.features.transform(lambda features: parse_feature_name(features, transcripts) if features!='.' else '.')\n",
    "star_aligns = star_aligns.groupby(['read_id', 'chrom', 'align_start', 'align_end', 'align_orient', 'quality', 'n_segments', 'block_lengths', 'block_starts']).agg({'features':tuple}).reset_index()\n",
    "print(star_aligns.at[0, 'features'])\n",
    "star_aligns['segment'] = star_aligns.apply(infer_segments, axis=1)\n",
    "star_aligns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>align_start</th>\n",
       "      <th>align_end</th>\n",
       "      <th>align_orient</th>\n",
       "      <th>quality</th>\n",
       "      <th>n_segments</th>\n",
       "      <th>block_lengths</th>\n",
       "      <th>block_starts</th>\n",
       "      <th>features</th>\n",
       "      <th>segment</th>\n",
       "      <th>seg_features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>43161160</td>\n",
       "      <td>43162922</td>\n",
       "      <td>-</td>\n",
       "      <td>255</td>\n",
       "      <td>2</td>\n",
       "      <td>30,120,</td>\n",
       "      <td>0,1642,</td>\n",
       "      <td>((exon, chr22, 43161051, 43161190, +, (('ENST0...</td>\n",
       "      <td>(43161160, 43161190, None)</td>\n",
       "      <td>((exon, chr22, 43161051, 43161190, +, (('ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>43161160</td>\n",
       "      <td>43162922</td>\n",
       "      <td>-</td>\n",
       "      <td>255</td>\n",
       "      <td>2</td>\n",
       "      <td>30,120,</td>\n",
       "      <td>0,1642,</td>\n",
       "      <td>((exon, chr22, 43161051, 43161190, +, (('ENST0...</td>\n",
       "      <td>(43162802, 43162922, None)</td>\n",
       "      <td>((exon, chr22, 43161051, 43161190, +, (('ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11992</td>\n",
       "      <td>chr9</td>\n",
       "      <td>9442096</td>\n",
       "      <td>9442246</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>150,</td>\n",
       "      <td>0,</td>\n",
       "      <td>((exon, chr9, 9442059, 9442380, +, (('ENST0000...</td>\n",
       "      <td>(9442096, 9442246, None)</td>\n",
       "      <td>((intron, chr9, 9397482, 9574731, -, (('ENST00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:12085</td>\n",
       "      <td>chr14</td>\n",
       "      <td>49586699</td>\n",
       "      <td>49586849</td>\n",
       "      <td>-</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>150,</td>\n",
       "      <td>0,</td>\n",
       "      <td>((exon, chr14, 49586579, 49586878, +, (('ENST0...</td>\n",
       "      <td>(49586699, 49586849, None)</td>\n",
       "      <td>((intron, chr14, 49586049, 49598399, -, (('ENS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:12117</td>\n",
       "      <td>chr14</td>\n",
       "      <td>49862679</td>\n",
       "      <td>49862792</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>113,</td>\n",
       "      <td>0,</td>\n",
       "      <td>((intron, chr14, 49862316, 49863828, -, (('ENS...</td>\n",
       "      <td>(49862679, 49862792, None)</td>\n",
       "      <td>((intron, chr14, 49862316, 49863828, -, (('ENS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83003</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>48233331</td>\n",
       "      <td>48253811</td>\n",
       "      <td>-</td>\n",
       "      <td>255</td>\n",
       "      <td>2</td>\n",
       "      <td>146,4,</td>\n",
       "      <td>0,20476,</td>\n",
       "      <td>((exon, chr13, 48233202, 48233477, +, (('ENST0...</td>\n",
       "      <td>(48233331, 48233477, None)</td>\n",
       "      <td>((exon, chr13, 48233202, 48233477, +, (('ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83003</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>48233331</td>\n",
       "      <td>48253811</td>\n",
       "      <td>-</td>\n",
       "      <td>255</td>\n",
       "      <td>2</td>\n",
       "      <td>146,4,</td>\n",
       "      <td>0,20476,</td>\n",
       "      <td>((exon, chr13, 48233202, 48233477, +, (('ENST0...</td>\n",
       "      <td>(48253807, 48253811, None)</td>\n",
       "      <td>((exon, chr13, 48233202, 48233477, +, (('ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83004</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:5807</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114750152</td>\n",
       "      <td>114757991</td>\n",
       "      <td>-</td>\n",
       "      <td>255</td>\n",
       "      <td>2</td>\n",
       "      <td>55,67,</td>\n",
       "      <td>0,7772,</td>\n",
       "      <td>((exon, chr1, 114749816, 114750207, -, (('ENST...</td>\n",
       "      <td>(114750152, 114750207, None)</td>\n",
       "      <td>((exon, chr1, 114749816, 114750207, -, (('ENST...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83004</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:5807</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114750152</td>\n",
       "      <td>114757991</td>\n",
       "      <td>-</td>\n",
       "      <td>255</td>\n",
       "      <td>2</td>\n",
       "      <td>55,67,</td>\n",
       "      <td>0,7772,</td>\n",
       "      <td>((exon, chr1, 114749816, 114750207, -, (('ENST...</td>\n",
       "      <td>(114757924, 114757991, None)</td>\n",
       "      <td>((exon, chr1, 114749816, 114750207, -, (('ENST...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83005</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>49586704</td>\n",
       "      <td>49586854</td>\n",
       "      <td>-</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>150,</td>\n",
       "      <td>0,</td>\n",
       "      <td>((exon, chr14, 49586579, 49586878, +, (('ENST0...</td>\n",
       "      <td>(49586704, 49586854, None)</td>\n",
       "      <td>((intron, chr14, 49586049, 49598399, -, (('ENS...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104583 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              read_id  chrom  align_start  \\\n",
       "0      NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22     43161160   \n",
       "0      NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22     43161160   \n",
       "1      NGSNJ-086:229:GW200110425th:1:1101:10004:11992   chr9      9442096   \n",
       "2      NGSNJ-086:229:GW200110425th:1:1101:10004:12085  chr14     49586699   \n",
       "3      NGSNJ-086:229:GW200110425th:1:1101:10004:12117  chr14     49862679   \n",
       "...                                               ...    ...          ...   \n",
       "83003    NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13     48233331   \n",
       "83003    NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13     48233331   \n",
       "83004    NGSNJ-086:229:GW200110425th:1:1101:9995:5807   chr1    114750152   \n",
       "83004    NGSNJ-086:229:GW200110425th:1:1101:9995:5807   chr1    114750152   \n",
       "83005    NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14     49586704   \n",
       "\n",
       "       align_end align_orient  quality  n_segments block_lengths block_starts  \\\n",
       "0       43162922            -      255           2       30,120,      0,1642,   \n",
       "0       43162922            -      255           2       30,120,      0,1642,   \n",
       "1        9442246            +        3           1          150,           0,   \n",
       "2       49586849            -        3           1          150,           0,   \n",
       "3       49862792            +        3           1          113,           0,   \n",
       "...          ...          ...      ...         ...           ...          ...   \n",
       "83003   48253811            -      255           2        146,4,     0,20476,   \n",
       "83003   48253811            -      255           2        146,4,     0,20476,   \n",
       "83004  114757991            -      255           2        55,67,      0,7772,   \n",
       "83004  114757991            -      255           2        55,67,      0,7772,   \n",
       "83005   49586854            -        3           1          150,           0,   \n",
       "\n",
       "                                                features  \\\n",
       "0      ((exon, chr22, 43161051, 43161190, +, (('ENST0...   \n",
       "0      ((exon, chr22, 43161051, 43161190, +, (('ENST0...   \n",
       "1      ((exon, chr9, 9442059, 9442380, +, (('ENST0000...   \n",
       "2      ((exon, chr14, 49586579, 49586878, +, (('ENST0...   \n",
       "3      ((intron, chr14, 49862316, 49863828, -, (('ENS...   \n",
       "...                                                  ...   \n",
       "83003  ((exon, chr13, 48233202, 48233477, +, (('ENST0...   \n",
       "83003  ((exon, chr13, 48233202, 48233477, +, (('ENST0...   \n",
       "83004  ((exon, chr1, 114749816, 114750207, -, (('ENST...   \n",
       "83004  ((exon, chr1, 114749816, 114750207, -, (('ENST...   \n",
       "83005  ((exon, chr14, 49586579, 49586878, +, (('ENST0...   \n",
       "\n",
       "                            segment  \\\n",
       "0        (43161160, 43161190, None)   \n",
       "0        (43162802, 43162922, None)   \n",
       "1          (9442096, 9442246, None)   \n",
       "2        (49586699, 49586849, None)   \n",
       "3        (49862679, 49862792, None)   \n",
       "...                             ...   \n",
       "83003    (48233331, 48233477, None)   \n",
       "83003    (48253807, 48253811, None)   \n",
       "83004  (114750152, 114750207, None)   \n",
       "83004  (114757924, 114757991, None)   \n",
       "83005    (49586704, 49586854, None)   \n",
       "\n",
       "                                            seg_features  \n",
       "0      ((exon, chr22, 43161051, 43161190, +, (('ENST0...  \n",
       "0      ((exon, chr22, 43161051, 43161190, +, (('ENST0...  \n",
       "1      ((intron, chr9, 9397482, 9574731, -, (('ENST00...  \n",
       "2      ((intron, chr14, 49586049, 49598399, -, (('ENS...  \n",
       "3      ((intron, chr14, 49862316, 49863828, -, (('ENS...  \n",
       "...                                                  ...  \n",
       "83003  ((exon, chr13, 48233202, 48233477, +, (('ENST0...  \n",
       "83003  ((exon, chr13, 48233202, 48233477, +, (('ENST0...  \n",
       "83004  ((exon, chr1, 114749816, 114750207, -, (('ENST...  \n",
       "83004  ((exon, chr1, 114749816, 114750207, -, (('ENST...  \n",
       "83005  ((intron, chr14, 49586049, 49598399, -, (('ENS...  \n",
       "\n",
       "[104583 rows x 12 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def infer_seg_features(row):\n",
    "\tif row['features'] == ('.',):\n",
    "\t\treturn ('.',)\n",
    "\t\n",
    "\tout = []\n",
    "\tfor feat in row['features']:\n",
    "\t\tif row['segment'].overlaps(feat[2], feat[3]):\n",
    "\t\t\tout.append(feat)\n",
    "\n",
    "\t# If any protein-coding gene features, remove non-protein-coding gene features\n",
    "\tprotein_coding = [feat for feat in row['features'] if 'protein_coding' in feat[6]]\n",
    "\tif len(protein_coding) > 0:\n",
    "\t\tout = tuple(protein_coding)\n",
    "\n",
    "\treturn tuple(out)\n",
    "\n",
    "\n",
    "star_aligns = star_aligns.explode('segment')\n",
    "star_aligns['seg_features'] = star_aligns.apply(infer_seg_features, axis=1)\n",
    "star_aligns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>align_start</th>\n",
       "      <th>align_end</th>\n",
       "      <th>align_orient</th>\n",
       "      <th>quality</th>\n",
       "      <th>n_segments</th>\n",
       "      <th>block_lengths</th>\n",
       "      <th>block_starts</th>\n",
       "      <th>features</th>\n",
       "      <th>segment</th>\n",
       "      <th>seg_features</th>\n",
       "      <th>seg_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>43161160</td>\n",
       "      <td>43162922</td>\n",
       "      <td>-</td>\n",
       "      <td>255</td>\n",
       "      <td>2</td>\n",
       "      <td>30,120,</td>\n",
       "      <td>0,1642,</td>\n",
       "      <td>((exon, chr22, 43161051, 43161190, +, (('ENST0...</td>\n",
       "      <td>(43161160, 43161190, None)</td>\n",
       "      <td>((exon, chr22, 43161051, 43161190, +, (('ENST0...</td>\n",
       "      <td>Exonic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>43161160</td>\n",
       "      <td>43162922</td>\n",
       "      <td>-</td>\n",
       "      <td>255</td>\n",
       "      <td>2</td>\n",
       "      <td>30,120,</td>\n",
       "      <td>0,1642,</td>\n",
       "      <td>((exon, chr22, 43161051, 43161190, +, (('ENST0...</td>\n",
       "      <td>(43162802, 43162922, None)</td>\n",
       "      <td>((exon, chr22, 43161051, 43161190, +, (('ENST0...</td>\n",
       "      <td>Exonic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11992</td>\n",
       "      <td>chr9</td>\n",
       "      <td>9442096</td>\n",
       "      <td>9442246</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>150,</td>\n",
       "      <td>0,</td>\n",
       "      <td>((exon, chr9, 9442059, 9442380, +, (('ENST0000...</td>\n",
       "      <td>(9442096, 9442246, None)</td>\n",
       "      <td>((intron, chr9, 9397482, 9574731, -, (('ENST00...</td>\n",
       "      <td>Intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:12085</td>\n",
       "      <td>chr14</td>\n",
       "      <td>49586699</td>\n",
       "      <td>49586849</td>\n",
       "      <td>-</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>150,</td>\n",
       "      <td>0,</td>\n",
       "      <td>((exon, chr14, 49586579, 49586878, +, (('ENST0...</td>\n",
       "      <td>(49586699, 49586849, None)</td>\n",
       "      <td>((intron, chr14, 49586049, 49598399, -, (('ENS...</td>\n",
       "      <td>Intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:12117</td>\n",
       "      <td>chr14</td>\n",
       "      <td>49862679</td>\n",
       "      <td>49862792</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>113,</td>\n",
       "      <td>0,</td>\n",
       "      <td>((intron, chr14, 49862316, 49863828, -, (('ENS...</td>\n",
       "      <td>(49862679, 49862792, None)</td>\n",
       "      <td>((intron, chr14, 49862316, 49863828, -, (('ENS...</td>\n",
       "      <td>Genic, ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83003</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>48233331</td>\n",
       "      <td>48253811</td>\n",
       "      <td>-</td>\n",
       "      <td>255</td>\n",
       "      <td>2</td>\n",
       "      <td>146,4,</td>\n",
       "      <td>0,20476,</td>\n",
       "      <td>((exon, chr13, 48233202, 48233477, +, (('ENST0...</td>\n",
       "      <td>(48233331, 48233477, None)</td>\n",
       "      <td>((exon, chr13, 48233202, 48233477, +, (('ENST0...</td>\n",
       "      <td>Genic, ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83003</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>48233331</td>\n",
       "      <td>48253811</td>\n",
       "      <td>-</td>\n",
       "      <td>255</td>\n",
       "      <td>2</td>\n",
       "      <td>146,4,</td>\n",
       "      <td>0,20476,</td>\n",
       "      <td>((exon, chr13, 48233202, 48233477, +, (('ENST0...</td>\n",
       "      <td>(48253807, 48253811, None)</td>\n",
       "      <td>((exon, chr13, 48233202, 48233477, +, (('ENST0...</td>\n",
       "      <td>Genic, ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83004</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:5807</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114750152</td>\n",
       "      <td>114757991</td>\n",
       "      <td>-</td>\n",
       "      <td>255</td>\n",
       "      <td>2</td>\n",
       "      <td>55,67,</td>\n",
       "      <td>0,7772,</td>\n",
       "      <td>((exon, chr1, 114749816, 114750207, -, (('ENST...</td>\n",
       "      <td>(114750152, 114750207, None)</td>\n",
       "      <td>((exon, chr1, 114749816, 114750207, -, (('ENST...</td>\n",
       "      <td>Genic, ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83004</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:5807</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114750152</td>\n",
       "      <td>114757991</td>\n",
       "      <td>-</td>\n",
       "      <td>255</td>\n",
       "      <td>2</td>\n",
       "      <td>55,67,</td>\n",
       "      <td>0,7772,</td>\n",
       "      <td>((exon, chr1, 114749816, 114750207, -, (('ENST...</td>\n",
       "      <td>(114757924, 114757991, None)</td>\n",
       "      <td>((exon, chr1, 114749816, 114750207, -, (('ENST...</td>\n",
       "      <td>Genic, ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83005</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>49586704</td>\n",
       "      <td>49586854</td>\n",
       "      <td>-</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>150,</td>\n",
       "      <td>0,</td>\n",
       "      <td>((exon, chr14, 49586579, 49586878, +, (('ENST0...</td>\n",
       "      <td>(49586704, 49586854, None)</td>\n",
       "      <td>((intron, chr14, 49586049, 49598399, -, (('ENS...</td>\n",
       "      <td>Intronic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104583 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              read_id  chrom  align_start  \\\n",
       "0      NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22     43161160   \n",
       "0      NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22     43161160   \n",
       "1      NGSNJ-086:229:GW200110425th:1:1101:10004:11992   chr9      9442096   \n",
       "2      NGSNJ-086:229:GW200110425th:1:1101:10004:12085  chr14     49586699   \n",
       "3      NGSNJ-086:229:GW200110425th:1:1101:10004:12117  chr14     49862679   \n",
       "...                                               ...    ...          ...   \n",
       "83003    NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13     48233331   \n",
       "83003    NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13     48233331   \n",
       "83004    NGSNJ-086:229:GW200110425th:1:1101:9995:5807   chr1    114750152   \n",
       "83004    NGSNJ-086:229:GW200110425th:1:1101:9995:5807   chr1    114750152   \n",
       "83005    NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14     49586704   \n",
       "\n",
       "       align_end align_orient  quality  n_segments block_lengths block_starts  \\\n",
       "0       43162922            -      255           2       30,120,      0,1642,   \n",
       "0       43162922            -      255           2       30,120,      0,1642,   \n",
       "1        9442246            +        3           1          150,           0,   \n",
       "2       49586849            -        3           1          150,           0,   \n",
       "3       49862792            +        3           1          113,           0,   \n",
       "...          ...          ...      ...         ...           ...          ...   \n",
       "83003   48253811            -      255           2        146,4,     0,20476,   \n",
       "83003   48253811            -      255           2        146,4,     0,20476,   \n",
       "83004  114757991            -      255           2        55,67,      0,7772,   \n",
       "83004  114757991            -      255           2        55,67,      0,7772,   \n",
       "83005   49586854            -        3           1          150,           0,   \n",
       "\n",
       "                                                features  \\\n",
       "0      ((exon, chr22, 43161051, 43161190, +, (('ENST0...   \n",
       "0      ((exon, chr22, 43161051, 43161190, +, (('ENST0...   \n",
       "1      ((exon, chr9, 9442059, 9442380, +, (('ENST0000...   \n",
       "2      ((exon, chr14, 49586579, 49586878, +, (('ENST0...   \n",
       "3      ((intron, chr14, 49862316, 49863828, -, (('ENS...   \n",
       "...                                                  ...   \n",
       "83003  ((exon, chr13, 48233202, 48233477, +, (('ENST0...   \n",
       "83003  ((exon, chr13, 48233202, 48233477, +, (('ENST0...   \n",
       "83004  ((exon, chr1, 114749816, 114750207, -, (('ENST...   \n",
       "83004  ((exon, chr1, 114749816, 114750207, -, (('ENST...   \n",
       "83005  ((exon, chr14, 49586579, 49586878, +, (('ENST0...   \n",
       "\n",
       "                            segment  \\\n",
       "0        (43161160, 43161190, None)   \n",
       "0        (43162802, 43162922, None)   \n",
       "1          (9442096, 9442246, None)   \n",
       "2        (49586699, 49586849, None)   \n",
       "3        (49862679, 49862792, None)   \n",
       "...                             ...   \n",
       "83003    (48233331, 48233477, None)   \n",
       "83003    (48253807, 48253811, None)   \n",
       "83004  (114750152, 114750207, None)   \n",
       "83004  (114757924, 114757991, None)   \n",
       "83005    (49586704, 49586854, None)   \n",
       "\n",
       "                                            seg_features       seg_species  \n",
       "0      ((exon, chr22, 43161051, 43161190, +, (('ENST0...            Exonic  \n",
       "0      ((exon, chr22, 43161051, 43161190, +, (('ENST0...            Exonic  \n",
       "1      ((intron, chr9, 9397482, 9574731, -, (('ENST00...          Intronic  \n",
       "2      ((intron, chr14, 49586049, 49598399, -, (('ENS...          Intronic  \n",
       "3      ((intron, chr14, 49862316, 49863828, -, (('ENS...  Genic, ambiguous  \n",
       "...                                                  ...               ...  \n",
       "83003  ((exon, chr13, 48233202, 48233477, +, (('ENST0...  Genic, ambiguous  \n",
       "83003  ((exon, chr13, 48233202, 48233477, +, (('ENST0...  Genic, ambiguous  \n",
       "83004  ((exon, chr1, 114749816, 114750207, -, (('ENST...  Genic, ambiguous  \n",
       "83004  ((exon, chr1, 114749816, 114750207, -, (('ENST...  Genic, ambiguous  \n",
       "83005  ((intron, chr14, 49586049, 49598399, -, (('ENS...          Intronic  \n",
       "\n",
       "[104583 rows x 13 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def identify_segment(row):\n",
    "\tif row['seg_features']==('.',):\n",
    "\t\treturn 'Inter-genic'\n",
    "\t\n",
    "\texons = [feat for feat in row['seg_features'] if feat[0]=='exon']\n",
    "\tintrons = [feat for feat in row['seg_features'] if feat[0]=='intron']\n",
    "\n",
    "\tif len(exons) > 0 and len(introns) == 0:\n",
    "\t\treturn 'Exonic'\n",
    "\n",
    "\tif len(exons) == 0 and len(introns) > 0:\n",
    "\t\t# Check for start at 5'ss\n",
    "\t\tfor intron in introns:\n",
    "\t\t\tif intron[5]=='+' and intron[2]==row['segment'].begin:\n",
    "\t\t\t\treturn \"Ends at a 5'ss\"\n",
    "\t\t\telif intron[5]=='-' and intron[3]==row['segment'].end:\n",
    "\t\t\t\treturn \"Ends at a 5'ss\"\n",
    "\t\treturn 'Intronic'\n",
    "\n",
    "\t# We can deduce that there's at least 1 exon and 1 intron at this point\n",
    "\texons_interval = IntervalTree.from_tuples([(exon[2], exon[3]) for exon in exons])\n",
    "\tif tree_covers_interval(exons_interval, row['segment']) is True:\n",
    "\t\treturn 'Genic, ambiguous'\n",
    "\tintrons_interval = IntervalTree.from_tuples([(intron[2], intron[3]) for intron in introns])\n",
    "\tif tree_covers_interval(introns_interval, row['segment']) is True:\n",
    "\t\treturn 'Genic, ambiguous'\n",
    "\n",
    "\treturn 'pre-mRNA'\n",
    "\n",
    "star_aligns['seg_species'] = star_aligns.apply(identify_segment, axis=1)\n",
    "star_aligns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "species\n",
      "Intronic            29983\n",
      "Genic, ambiguous    20755\n",
      "Exonic              20182\n",
      "mRNA                 8960\n",
      "Inter-genic          2372\n",
      "pre-mRNA              747\n",
      "Weird spliced           7\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_40365/801318579.py:25: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11992</td>\n",
       "      <td>Intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:12085</td>\n",
       "      <td>Intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:12117</td>\n",
       "      <td>Genic, ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:13244</td>\n",
       "      <td>Genic, ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83001</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:27352</td>\n",
       "      <td>Inter-genic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83002</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:3928</td>\n",
       "      <td>Intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83003</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>Genic, ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83004</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:5807</td>\n",
       "      <td>Genic, ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83005</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>Intronic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>83006 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              read_id           species\n",
       "0      NGSNJ-086:229:GW200110425th:1:1101:10004:11303              mRNA\n",
       "1      NGSNJ-086:229:GW200110425th:1:1101:10004:11992          Intronic\n",
       "2      NGSNJ-086:229:GW200110425th:1:1101:10004:12085          Intronic\n",
       "3      NGSNJ-086:229:GW200110425th:1:1101:10004:12117  Genic, ambiguous\n",
       "4      NGSNJ-086:229:GW200110425th:1:1101:10004:13244  Genic, ambiguous\n",
       "...                                               ...               ...\n",
       "83001   NGSNJ-086:229:GW200110425th:1:1101:9995:27352       Inter-genic\n",
       "83002    NGSNJ-086:229:GW200110425th:1:1101:9995:3928          Intronic\n",
       "83003    NGSNJ-086:229:GW200110425th:1:1101:9995:4961  Genic, ambiguous\n",
       "83004    NGSNJ-086:229:GW200110425th:1:1101:9995:5807  Genic, ambiguous\n",
       "83005    NGSNJ-086:229:GW200110425th:1:1101:9995:9502          Intronic\n",
       "\n",
       "[83006 rows x 2 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def identify_read_species(df):\n",
    "\tassigned_species = set(df['seg_species'].values)\n",
    "\tif len(assigned_species) == 1:\n",
    "\t\tassigned_species = assigned_species.pop()\n",
    "\t\tif len(df)>1 and assigned_species == 'Exonic':\n",
    "\t\t\treturn 'mRNA'\n",
    "\t\telse:\n",
    "\t\t\treturn assigned_species\n",
    "\t\n",
    "\tif 'pre-mRNA' in assigned_species:\n",
    "\t\treturn 'pre-mRNA'\n",
    "\tif 'Genic, ambiguous' in assigned_species:\n",
    "\t\treturn 'Genic, ambiguous'\n",
    "\n",
    "\tif assigned_species in ({'Intronic', 'Exonic'},\n",
    "\t\t\t\t\t\t\t{'Inter-genic', 'Exonic'},\n",
    "\t\t\t\t\t\t\t{'Inter-genic', 'Intronic'}):\n",
    "\t\treturn 'Weird spliced'\n",
    "\n",
    "\tprint(assigned_species)\n",
    "\treturn 'Dunno'\n",
    "\n",
    "# spliced_aligns['species'] = spliced_aligns.read_id.map(spliced_aligns.groupby('read_id').apply(identify_spliced_species))\n",
    "\n",
    "star_species = star_aligns.groupby('read_id').apply(identify_read_species).reset_index(name='species')\n",
    "print(star_species.species.value_counts())\n",
    "star_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# unspliced_aligns = star_aligns.loc[star_aligns.n_segments==1].copy().reset_index()\n",
    "# spliced_aligns = star_aligns.loc[star_aligns.n_segments>1].copy().reset_index()\n",
    "# print(len(unspliced_aligns), len(spliced_aligns))\n",
    "# assert unspliced_aligns.read_id.is_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def unspliced_filter_features(row):\n",
    "# \tout = row['features']\n",
    "\n",
    "# \t# If any protein-coding gene features, remove non-protein-coding gene features\n",
    "# \tprotein_coding = [feat for feat in row['features'] if feat.data[6]=='protein_coding']\n",
    "# \tif len(protein_coding) > 0:\n",
    "# \t\tout = IntervalTree(protein_coding)\n",
    "\t\n",
    "# \treturn out\n",
    "\n",
    "def unspliced_filter_features(row):\n",
    "\tif row['features'] == ('.',):\n",
    "\t\treturn IntervalTree([])\n",
    "\t\n",
    "\tout = [Interval(feature[2], feature[3], feature) for feature in row['features']]\n",
    "\treturn IntervalTree(out)\n",
    "\n",
    "\n",
    "def identify_unspliced_species(row):\n",
    "\tif len(row['filtered_features']) == 0:\n",
    "\t\treturn 'Inter-genic'\n",
    "\n",
    "\texons = IntervalTree([feat for feat in row['filtered_features'] if feat.data[0]=='exon'])\n",
    "\tintrons = IntervalTree([feat for feat in row['filtered_features'] if feat.data[0]=='intron'])\n",
    "\n",
    "\tif len(exons) > 0 and len(introns) == 0:\n",
    "\t\treturn 'Exonic'\n",
    "\n",
    "\tif len(exons) == 0 and len(introns) > 0:\n",
    "\t\t# Check for start at 5'ss\n",
    "\t\tfor intron in introns:\n",
    "\t\t\tif intron.data[4]=='+' and intron.begin==row['align_start']:\n",
    "\t\t\t\treturn \"Ends at a 5'ss\"\n",
    "\t\t\telif intron.data[4]=='-' and intron.end==row['align_end']:\n",
    "\t\t\t\treturn \"Ends at a 5'ss\"\n",
    "\t\treturn 'Intronic'\n",
    "\n",
    "\t# We can deduce that there's at least 1 exon and 1 intron at this point\n",
    "\tif tree_covers_interval(exons, Interval(row['align_start'], row['align_end'])) is True:\n",
    "\t\treturn 'Genic, ambiguous'\n",
    "\tif tree_covers_interval(introns, Interval(row['align_start'], row['align_end'])) is True:\n",
    "\t\treturn 'Genic, ambiguous'\n",
    "\n",
    "\treturn 'pre-mRNA'\n",
    "\n",
    "\n",
    "\n",
    "unspliced_aligns = unspliced_aligns.set_index('read_id', drop=False)\n",
    "unspliced_aligns['filtered_features'] = unspliced_aligns.apply(unspliced_filter_features, axis=1, result_type='reduce')\n",
    "# unspliced_aligns['filtered_gene_types'] = unspliced_aligns.filtered_features.transform(lambda features: tuple(set([feat.data[6] for feat in features])))\n",
    "unspliced_aligns['species'] = unspliced_aligns.apply(identify_unspliced_species, axis=1)\n",
    "print(unspliced_aligns.species.value_counts())\n",
    "unspliced_aligns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unspliced_aligns['full_coverage'] = unspliced_aligns.apply(lambda row: tree_covers_interval(row['filtered_features'], Interval(row['align_start'], row['align_end'])), axis=1)\n",
    "print(unspliced_aligns.full_coverage.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def spliced_filter_features(group:pd.DataFrame):\n",
    "\t# out = group.features.iloc[0]\n",
    "\t# all_gids = set([feat.data[4] for feat in out])\n",
    "\t# n_segments = len(group.segments.iloc[0])\n",
    "\t\n",
    "\t# # Remove features from genes that don't cover all segments\n",
    "\t# for gid in all_gids:\n",
    "\t# \tsegments_in_gene = group.loc[group.seg_gids.transform(str).str.contains(gid)]\n",
    "\t# \tif len(segments_in_gene) < n_segments:\n",
    "\t# \t\tout = IntervalTree([feat for feat in out if feat.data[4] != gid])\n",
    "\t\n",
    "\t# If any protein-coding gene features, remove non-protein-coding gene features\n",
    "\tout = group.features.iloc[0]\n",
    "\tprotein_coding = [feat for feat in out if feat.data[6]=='protein_coding']\n",
    "\tif len(protein_coding) > 0:\n",
    "\t\tout = IntervalTree(protein_coding)\n",
    "\t\n",
    "\treturn out\n",
    "\n",
    "\n",
    "def infer_seg_features(row, filtered:bool):\n",
    "\tseg_interval = Interval(row['seg_start'], row['seg_end'])\n",
    "\tfeatures = row['filtered_features'] if filtered is True else row['features']\n",
    "\n",
    "\tout = IntervalTree([feat for feat in features if feat.overlaps(seg_interval)])\n",
    "\treturn out\n",
    "\n",
    "\n",
    "\n",
    "spliced_aligns = spliced_aligns.explode('segments').reset_index(drop=True)\n",
    "spliced_aligns['seg_start'] = spliced_aligns.segments.transform(lambda s: s.begin)\n",
    "spliced_aligns['seg_end'] = spliced_aligns.segments.transform(lambda s: s.end)\n",
    "spliced_aligns['seg_features'] = spliced_aligns.apply(infer_seg_features, filtered=False, axis=1)\n",
    "spliced_aligns['seg_gids'] = spliced_aligns.seg_features.transform(lambda features: tuple(set([feat.data[4] for feat in features])))\n",
    "spliced_aligns['filtered_features'] = spliced_aligns.read_id.map(spliced_aligns.groupby('read_id').apply(spliced_filter_features))\n",
    "spliced_aligns['filtered_seg_features'] = spliced_aligns.apply(infer_seg_features, filtered=True, axis=1)\n",
    "spliced_aligns['filtered_seg_gene_types'] = spliced_aligns.filtered_seg_features.transform(lambda features: tuple(set([feat.data[6] for feat in features])))\n",
    "spliced_aligns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert set(unspliced_aligns.filtered_gene_types.unique()).issubset({(), ('ncRNA',), ('protein_coding',), ('Ig or TcR',)})\n",
    "assert set(spliced_aligns.filtered_seg_gene_types.unique()).issubset({(), ('ncRNA',), ('protein_coding',), ('Ig or TcR',)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def identify_segment(row):\n",
    "\tif len(row['filtered_seg_features'])==0:\n",
    "\t\treturn 'Inter-genic'\n",
    "\t\n",
    "\tseg_interval = Interval(row['seg_start'], row['seg_end'])\n",
    "\texons = IntervalTree([feat for feat in row['filtered_seg_features'] if feat.data[0]=='exon'])\n",
    "\tintrons = IntervalTree([feat for feat in row['filtered_seg_features'] if feat.data[0]=='intron'])\n",
    "\n",
    "\tif len(exons) > 0 and len(introns) == 0:\n",
    "\t\treturn 'Exonic'\n",
    "\n",
    "\tif len(exons) == 0 and len(introns) > 0:\n",
    "\t\t# Check for start at 5'ss\n",
    "\t\tfor intron in introns:\n",
    "\t\t\tif intron.data[3]=='+' and intron.begin==seg_interval.begin:\n",
    "\t\t\t\treturn \"Ends at a 5'ss\"\n",
    "\t\t\telif intron.data[3]=='-' and intron.end==seg_interval.end:\n",
    "\t\t\t\treturn \"Ends at a 5'ss\"\n",
    "\t\treturn 'Intronic'\n",
    "\n",
    "\t# We can deduce that there's at least 1 exon and 1 intron at this point\n",
    "\tif tree_covers_interval(exons, seg_interval) is True:\n",
    "\t\treturn 'Genic, ambiguous'\n",
    "\tif tree_covers_interval(introns, seg_interval) is True:\n",
    "\t\treturn 'Genic, ambiguous'\n",
    "\n",
    "\treturn 'pre-mRNA'\n",
    "\n",
    "spliced_aligns['seg_species'] = spliced_aligns.apply(identify_segment, axis=1)\n",
    "print(spliced_aligns.seg_species.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def identify_spliced_species(df):\t\n",
    "\tassigned_species = set(df['seg_species'].values)\n",
    "\tif len(assigned_species) == 1:\n",
    "\t\tassigned_species = assigned_species.pop()\n",
    "\t\tif assigned_species == 'Exonic':\n",
    "\t\t\treturn 'mRNA'\n",
    "\t\telse:\n",
    "\t\t\treturn assigned_species\n",
    "\t\n",
    "\tif 'pre-mRNA' in assigned_species:\n",
    "\t\treturn 'pre-mRNA'\n",
    "\tif 'Genic, ambiguous' in assigned_species:\n",
    "\t\treturn 'Genic, ambiguous'\n",
    "\tif assigned_species in ({'Intronic', 'Exonic'},\n",
    "\t\t\t\t\t\t\t{'Inter-genic', 'Exonic'},\n",
    "\t\t\t\t\t\t\t{'Inter-genic', 'Intronic'}):\n",
    "\t\treturn 'Weird spliced'\n",
    "\n",
    "\tprint(assigned_species)\n",
    "\treturn 'Dunno'\n",
    "\n",
    "spliced_aligns['species'] = spliced_aligns.read_id.map(spliced_aligns.groupby('read_id').apply(identify_spliced_species))\n",
    "print(spliced_aligns.drop_duplicates('read_id').species.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# If weird spliced is less than 1% prevalent, lump it with Genic, ambiguous\n",
    "prop = spliced_aligns.species.eq('Weird spliced').value_counts(normalize=True)[True]*100\n",
    "assert prop < 1, prop\n",
    "spliced_aligns.species = spliced_aligns.species.replace('Weird spliced', 'Genic, ambiguous')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(spliced_aligns.drop_duplicates('read_id').species.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert spliced_aligns.species.eq('Dunno').sum() == 0\n",
    "# spliced_aligns.loc[spliced_aligns.species.eq('Dunno')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_base='/Users/trumanmooney/Downloads/alpha_main_merge/testing/output/C22_test_read1_single_end_hisat2/C22_test_'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'NGSNJ-086:229:GW200110425th:1:1101:2166:13448',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24551:5713',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30291:23797',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5394:15374',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8314:28354'}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lariat_rids = pd.read_csv(f'{output_base}lariat_reads.tsv', sep='\\t', usecols=[0]).read_id\n",
    "lariat_rids = set(lariat_rids)\n",
    "lariat_rids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "circular_rids = pd.read_csv(f'{output_base}circularized_introns.tsv', sep='\\t').read_id\n",
    "circular_rids = set(circular_rids)\n",
    "circular_rids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/Users/trumanmooney/Downloads/alpha_main_merge/testing/output/C22_test_read1_single_end_hisat2/C22_test_template_switching_alignments.tsv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[28], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m temp_switch_rids \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43moutput_base\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43mtemplate_switching_alignments.tsv\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mread_id\n\u001b[1;32m      2\u001b[0m temp_switch_rids \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mset\u001b[39m(temp_switch_rids)\n\u001b[1;32m      3\u001b[0m temp_switch_rids\n",
      "File \u001b[0;32m~/miniforge3/envs/truth/lib/python3.10/site-packages/pandas/io/parsers/readers.py:1026\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m   1013\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m   1014\u001b[0m     dialect,\n\u001b[1;32m   1015\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1022\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m   1023\u001b[0m )\n\u001b[1;32m   1024\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m-> 1026\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/truth/lib/python3.10/site-packages/pandas/io/parsers/readers.py:620\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    617\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    619\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 620\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    622\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    623\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m~/miniforge3/envs/truth/lib/python3.10/site-packages/pandas/io/parsers/readers.py:1620\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1617\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1619\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1620\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/truth/lib/python3.10/site-packages/pandas/io/parsers/readers.py:1880\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1878\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[1;32m   1879\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1880\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1881\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1882\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1883\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1884\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcompression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1885\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmemory_map\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1886\u001b[0m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1887\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding_errors\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrict\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1888\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstorage_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1889\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1890\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1891\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n",
      "File \u001b[0;32m~/miniforge3/envs/truth/lib/python3.10/site-packages/pandas/io/common.py:873\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    868\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    869\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    870\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    871\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    872\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 873\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m    874\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    875\u001b[0m \u001b[43m            \u001b[49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    876\u001b[0m \u001b[43m            \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    877\u001b[0m \u001b[43m            \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    878\u001b[0m \u001b[43m            \u001b[49m\u001b[43mnewline\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    879\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    880\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    881\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[1;32m    882\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/Users/trumanmooney/Downloads/alpha_main_merge/testing/output/C22_test_read1_single_end_hisat2/C22_test_template_switching_alignments.tsv'"
     ]
    }
   ],
   "source": [
    "temp_switch_rids = pd.read_csv(f'{output_base}template_switching_alignments.tsv', sep='\\t').read_id\n",
    "temp_switch_rids = set(temp_switch_rids)\n",
    "temp_switch_rids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>read_is_reverse</th>\n",
       "      <th>read_seq</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>head_start</th>\n",
       "      <th>head_end</th>\n",
       "      <th>fivep_pos</th>\n",
       "      <th>bp_pos</th>\n",
       "      <th>read_bp_nt</th>\n",
       "      <th>...</th>\n",
       "      <th>genomic_bp_nt</th>\n",
       "      <th>threep_pos</th>\n",
       "      <th>bp_dist_to_threep</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>read_num</th>\n",
       "      <th>align_mismatch</th>\n",
       "      <th>total_mapped_reads</th>\n",
       "      <th>filter_failed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:8314:28354</td>\n",
       "      <td>False</td>\n",
       "      <td>GGGGACCCAGTGGAGCACACGGGTGAGCCCAGTGCAGCCGGGTGGC...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>1634741</td>\n",
       "      <td>1634763</td>\n",
       "      <td>1634913</td>\n",
       "      <td>1634741</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>1634708</td>\n",
       "      <td>-33</td>\n",
       "      <td>ENSG00000008128.23_12</td>\n",
       "      <td>CDK11A</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>55100</td>\n",
       "      <td>not_chosen</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         read_id  read_is_reverse  \\\n",
       "0  NGSNJ-086:229:GW200110425th:1:1101:8314:28354            False   \n",
       "\n",
       "                                            read_seq chrom strand  head_start  \\\n",
       "0  GGGGACCCAGTGGAGCACACGGGTGAGCCCAGTGCAGCCGGGTGGC...  chr1      -     1634741   \n",
       "\n",
       "   head_end  fivep_pos   bp_pos read_bp_nt  ... genomic_bp_nt threep_pos  \\\n",
       "0   1634763    1634913  1634741          G  ...             A    1634708   \n",
       "\n",
       "   bp_dist_to_threep                gene_id gene_name       gene_type  \\\n",
       "0                -33  ENSG00000008128.23_12    CDK11A  protein_coding   \n",
       "\n",
       "  read_num  align_mismatch  total_mapped_reads  filter_failed  \n",
       "0        1            True               55100     not_chosen  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lariat_failed = pd.read_csv(f'{output_base}failed_lariat_alignments.tsv', sep='\\t')\n",
    "lariat_failed = lariat_failed[['read_id', 'filter_failed']]\n",
    "lariat_failed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>filter_failed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:21630:7106</td>\n",
       "      <td>within_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:6072:17143</td>\n",
       "      <td>within_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:13648:19335</td>\n",
       "      <td>within_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:28818:13416</td>\n",
       "      <td>within_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:2700:7952</td>\n",
       "      <td>within_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:14172:15546</td>\n",
       "      <td>within_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:23439:13902</td>\n",
       "      <td>within_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:20618:16470</td>\n",
       "      <td>within_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:13295:19069</td>\n",
       "      <td>within_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:7075:11365</td>\n",
       "      <td>within_intron</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>218 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            read_id  filter_failed\n",
       "0     NGSNJ-086:229:GW200110425th:1:1101:21630:7106  within_intron\n",
       "1     NGSNJ-086:229:GW200110425th:1:1101:6072:17143  within_intron\n",
       "2    NGSNJ-086:229:GW200110425th:1:1101:13648:19335  within_intron\n",
       "3    NGSNJ-086:229:GW200110425th:1:1101:28818:13416  within_intron\n",
       "4      NGSNJ-086:229:GW200110425th:1:1101:2700:7952  within_intron\n",
       "..                                              ...            ...\n",
       "213  NGSNJ-086:229:GW200110425th:1:1101:14172:15546  within_intron\n",
       "214  NGSNJ-086:229:GW200110425th:1:1101:23439:13902  within_intron\n",
       "215  NGSNJ-086:229:GW200110425th:1:1101:20618:16470  within_intron\n",
       "216  NGSNJ-086:229:GW200110425th:1:1101:13295:19069  within_intron\n",
       "217   NGSNJ-086:229:GW200110425th:1:1101:7075:11365  within_intron\n",
       "\n",
       "[218 rows x 2 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trim_failed = pd.read_csv(f'{output_base}failed_trimmed_alignments.tsv', sep='\\t')\n",
    "trim_failed.read_id = trim_failed.read_id.transform(lambda rid: rid[:-4].split('/')[0])\n",
    "trim_failed = trim_failed[['read_id', 'filter_failed']].drop_duplicates(ignore_index=True)\n",
    "trim_failed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'NGSNJ-086:229:GW200110425th:1:1101:30219:27336',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12698:9111',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13856:9643',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4788:11506',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12292:17331',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24361:16031',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24939:25770',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26024:5478',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22010:22733',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5168:22623',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23818:10238',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11460:20149',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28411:17848',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9200:26412',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30942:23171',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6054:9471',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5538:9298',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31295:8688',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2917:17628',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3052:12602',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4634:7075',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3920:5556',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23285:21120',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19144:23156',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18331:11475',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32985:16626',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13214:18740',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32217:6089',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4481:17112',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18114:25942',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8892:18646',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14579:15154',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1904:5321',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4734:27978',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25482:12179',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30201:6104',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10628:17707',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13566:19539',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20681:4742',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32985:28275',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17065:11475',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32054:13260',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2139:26177',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14100:18364',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12870:2487',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16215:19867',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5719:15875',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17644:15765',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5077:23688',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8721:21261',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17074:3662',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24433:11021',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26765:6699',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30346:1658',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28311:22091',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26395:14982',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7310:9486',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19443:14685',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2446:10488',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3323:10034',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11532:24784',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29559:6026',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20229:7936',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31675:11036',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11695:3208',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26820:19633',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31620:14638',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28782:16986',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2510:7654',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2501:14403',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17490:19539',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27100:17832',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3857:16501',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2880:11209',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9824:2848',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24505:10645',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3260:13025',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27968:8093',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13792:16078',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11017:7607',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4282:8500',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2103:9017',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6741:14732',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8766:11350',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15863:6324',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12689:19586',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28682:24518',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16604:2754',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31042:19930',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17400:17660',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10239:4100',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29640:20635',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5231:23829',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16062:9236',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15465:5822',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21034:7670',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9525:1391',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10836:17159',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6940:23844',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20057:14935',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23845:3333',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17155:27633',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26729:25801',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8142:21418',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22471:8594',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29116:17910',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2889:8907',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23185:19194',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17237:28651',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26051:24377',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12337:7701',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25916:3944',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2709:7247',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19605:22889',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16694:26584',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3658:28526',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32244:8672',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17680:11663',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1099:3020',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16929:4038',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19542:14606',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22869:2926',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28203:27352',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3794:14888',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15049:5290',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1054:27023',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4300:24001',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4182:10645',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7618:26553',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24343:24079',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18005:28792',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1732:4335',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3821:7326',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2700:7952',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7536:21559',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32036:9157',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11397:5415',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18168:20431',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19298:17534',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14398:23202',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31729:28228',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26133:7357',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30870:20509',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12400:1172',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5547:24126',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10502:1391',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6931:18255',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22218:20306',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27706:23484',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7509:23453',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13783:24706',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12491:11788',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28908:7247',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18783:8093',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10239:7044',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22182:10285',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14913:28354',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31250:8077',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23918:1799',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15619:28604',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1741:6355',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23484:28541',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17390:27132',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8160:27273',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27507:17127',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24071:7106',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28185:4053',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6361:8312',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32117:13401',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17933:9095',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5954:5760',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29930:6198',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3929:3474',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28691:1642',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2365:22811',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9191:25019',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8314:28354',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28438:11005',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28339:8077',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17544:7889',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3341:14356',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20907:2409',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26693:6042',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21160:23265',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26684:20650',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2609:8234',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31277:5619',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31919:3881',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25075:24565',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18683:23234',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22634:1517',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21820:4930',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17291:3881',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15112:26381',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18493:18020',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18466:7607',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14172:17519',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1398:3349',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4381:21261',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11180:7733',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15049:2065',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6677:10520',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12906:10316',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10330:13088',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9760:10285',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5122:7294',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17499:15170',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24017:21731',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11514:28385',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7536:24001',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1805:10755',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25111:28009',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8034:10802',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23827:11287',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14787:4366',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17589:27164',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9444:15248',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12798:25942',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15980:2738',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2040:25410',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4652:11021',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2031:13495',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14525:19507',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3179:15421',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18566:2362',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22715:20196',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14606:5306',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14841:1329',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7808:27070',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24514:22748',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8250:5353',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3152:2848',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4137:18427',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29903:13886',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32579:4492',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15004:18302',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24180:2284',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6497:20885',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10357:16172',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17201:21136',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18864:14653',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5593:4163',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3540:25817',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25057:9972',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27859:14669',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29451:22874',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24822:21778',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30138:8813',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7518:20369',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10899:22748',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5394:15374',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1777:16658',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29677:23923',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27118:19147',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12906:27915',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11107:12962',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20401:17597',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8151:26412',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29559:23782',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26395:13448',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26377:15546',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14371:23813',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21522:1908',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3450:24533',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20916:4304',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25852:13636',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27923:10551',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25644:18129',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3721:16235',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28212:27211',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9968:23703',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19687:13542',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15329:23688',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7075:8703',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2139:13119',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1832:14779',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14389:9878',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25310:16360',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23312:11365',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6876:1125',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3188:11459',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10438:15624',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22236:6433',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30581:23453',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23882:21527',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28031:13088',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13214:2394',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10791:14857',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8305:5666',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9281:18411',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2917:14653',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15022:9565',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9679:8954',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17318:9189',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25482:8735',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14940:16282',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8657:11913',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24740:16470',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3586:15405',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10411:29105',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16089:10160',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20618:16470',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5132:17519',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23619:21355',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14714:5494',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32452:11663',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4987:17644',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1976:20259',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14217:13119',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28926:23030',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12337:27023',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17436:1094',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17336:16297',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16016:27853',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20763:22921',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29523:1579',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8486:1125',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21585:12696',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22453:8844',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19153:2284',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12264:24674',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15194:4695',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18195:15154',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4652:28213',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21052:1971',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26657:5384',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2627:8578',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25726:10379',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29794:17394',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3911:26115',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20907:22827',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8630:10770',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32479:20196',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8314:8594',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17047:18991',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1443:9596',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1181:8735',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9362:26130',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5023:15295',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26992:18364',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5999:11537',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29649:19899',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5276:3333',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12762:17456',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26549:23829',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30436:9799',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5584:8343',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2193:2033',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32226:7701',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24695:24659',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15971:17973',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17734:26976',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7934:26162',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19831:26036',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28004:7529',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26756:7936',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16902:20525',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9724:28354',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11189:23531',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32985:2816',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28610:22608',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21414:12242',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7672:24017',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23954:27038',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32877:8453',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23601:5979',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13557:25535',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3748:23703',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28221:20462',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21811:7075',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26467:28917',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14561:13964',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26051:12540',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18105:4914',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8739:28181',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6542:7435',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10601:19100',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10086:24878',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14407:17613',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25934:24862',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19226:23296',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19479:15217',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11605:22842',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14769:16673',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14416:1564',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9453:12414',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30789:11193',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20419:27336',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24180:19132',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3803:20259',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31720:9424',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19153:19069',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19325:12477',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26178:24314',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32651:18928',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32425:16720',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11017:9330',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19199:13103',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28619:5462',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21332:16016',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9227:1438',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13006:7012',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25229:9173',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9968:28494',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29098:26052',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28881:6543',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20636:8891',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25238:27195',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3143:14732',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8151:19586',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1072:5791',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19180:14043',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6777:23156',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4761:25238',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19153:25332',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25527:7059',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20012:25410',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22354:6480',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19578:9815',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20076:9956',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6370:15374',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16532:18443',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5132:19272',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2085:2722',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27814:12743',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3947:22013',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12635:6778',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7618:15812',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7600:27273',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24451:9361',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30120:5776',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24316:1297',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20880:28166',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1588:17644',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6027:4038',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22119:18881',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2989:2942',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4634:3693',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16622:13244',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20835:23140',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22860:8453',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24551:5713',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15709:23844',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27299:11036',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19578:23218',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27227:3615',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11668:25895',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32786:13651',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32615:3302',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7057:10019',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4020:27868',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13810:3740',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11984:23750',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26413:4398',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22752:7294',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9462:26553',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20130:1846',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30355:2801',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21938:5791',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30129:22921',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15004:11788',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20048:19554',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28827:25801',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22191:2409',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27218:15812',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13358:9345',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4101:3020',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23285:8719',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1615:1125',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9136:18756',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18204:3740',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26603:27398',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16749:9111',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3061:28087',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25102:8484',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19352:22075',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6027:7482',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2880:25864',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29875:3662',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14660:10316',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12038:5181',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14127:19695',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3595:19648',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30924:20290',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31159:3881',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3043:21887',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11559:14058',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1723:5541',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14714:6809',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32009:13338',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19669:20149',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32506:1078',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19416:22748',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3079:14184',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27462:12821',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20772:14732',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28709:22936',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13259:20384',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28935:14810',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16170:11741',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16812:27539',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2835:6026',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28745:28072',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5909:22435',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5801:9721',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18032:28808',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24623:15327',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2718:9079',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13006:19413',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31747:5181',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12463:10551',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31358:28776',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9218:12195',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9444:2253',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30653:23829',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31349:27884',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5900:4100',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26539:9408',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23430:1986',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19759:16297',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2899:10833',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7545:1407',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31033:4194',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18457:8531',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22968:7451',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31476:13761',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13829:26663',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8693:5932',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22399:26694',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31033:6699',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31250:25238',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19108:13949',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19027:16658',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11360:1720',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24117:12446',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1832:14622',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19425:9737',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17815:23359',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8088:23265',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22073:8469',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18439:18584',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8359:27524',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4363:22827',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11568:1767',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14154:6214',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3640:25238',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13810:7091',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2058:4993',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29107:3677',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23800:1752',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14886:22858',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3658:2910',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21088:2973',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12545:20243',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4933:28635',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15492:24158',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15637:24565',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16667:10817',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28809:10426',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14552:25003',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3549:19977',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18150:8249',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13548:9111',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30743:1564',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20654:3912',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11116:8970',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26042:23641',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17662:10379',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22471:24032',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27218:25520',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31675:10786',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8169:19179',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20582:12117',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19099:24549',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22218:16642',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32524:22153',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5267:27899',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19325:12132',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20916:10661',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1859:3521',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17210:27445',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21866:21574',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22309:18114',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25391:1031',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32832:26819',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23764:9862',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14362:22827',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1497:14325',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28203:22999',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11677:13792',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6045:8641',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29704:20682',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7012:17644',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15194:14747',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31846:24956',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23267:12007',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6198:17550',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11252:20040',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13295:19069',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29225:12806',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30074:11647',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27236:8015',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15031:4288',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23628:13823',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24786:23156',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21251:11929',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28339:24048',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20283:16579',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30879:11913',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5819:3458',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31268:28839',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27326:26522',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13602:4758',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24614:29027',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28700:19727',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2465:8359',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13431:18740',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10149:9236',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3323:22686',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27208:4679',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3540:18239',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26612:24502',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9254:13479',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32660:13463',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20645:22717',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15365:18834',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32117:25895',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16288:5807',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3857:5541',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4869:4977',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26820:24737',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1108:1188',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9697:9737',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13955:18176',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7075:11365',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17689:23860',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6605:3223',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19605:28025',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16450:22404',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27489:5541',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21947:28635',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29441:2597',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11966:27320',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10438:11490',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22598:4993',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24388:26569',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15609:9424',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26811:17801',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19569:4539',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5990:9392',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15917:6417',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24804:12727',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17725:19883',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2022:4147',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6090:23249',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27615:11397',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30544:14622',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6180:26631',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8169:10974',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25861:14465',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22824:24956',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2211:15436',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20383:16063',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21630:7106',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3296:19100',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18466:21104',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25229:10300',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28818:13416',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4734:28447',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12952:18662',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1127:23296',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17183:24612',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12400:12038',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3504:17174',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14371:23124',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32714:7357',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24117:23062',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30083:12383',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24867:27837',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3622:27305',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11098:17206',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28284:10050',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16514:26710',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12961:12070',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18186:25942',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16062:12837',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26006:22200',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29631:9314',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13702:25347',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7699:9784',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16559:24627',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18349:6214',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26549:2722',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32922:19022',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15356:19883',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2899:28401',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2392:17503',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23565:2315',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10529:15186',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16034:22654',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6741:17456',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1949:4961',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11840:1736',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12111:26099',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30933:13166',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7961:6887',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11369:7936',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2374:20854',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13648:24846',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15790:15593',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5086:4445',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10782:14779',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21947:2143',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31159:3912',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19840:9580',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24153:10347',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11053:18693',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2139:22263',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28556:15311',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13774:4679',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15989:12085',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10529:26365',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17020:18756',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5213:4601',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1651:12179',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26549:19476',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3649:18114',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28248:27336',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11252:2910',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27398:3380',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24406:8594',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7337:12320',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6858:2033',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2456:26694',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9010:1939',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24713:2080',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10972:25473',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2844:6887',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26693:22075',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7880:12289',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2166:13448',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30273:14841',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3929:23296',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29270:15515',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21332:19272',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9815:24784',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1714:9314',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26133:4163',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5493:4272',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26259:12743',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4689:12336',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6985:17253',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17499:19398',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27823:23563',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24099:27790',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8956:26083',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28926:20776',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19877:4319',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23529:26553',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21540:24956',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7328:1564',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31105:23171',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14064:4304',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7446:6809',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31693:28917',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11234:19852',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4318:20212',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10890:10019',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12472:10379',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14579:8829',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15103:6198',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23466:4679',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23194:12320',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30825:17550',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30291:23797',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10954:26663',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29134:10582',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14172:8343',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4164:25332',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17173:25629',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24975:6480',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19795:8970',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31937:12179',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31087:4382',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4182:10614',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15582:1329',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22372:19006',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17635:16658',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32570:26522',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12255:22404',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9941:28103',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16676:20008',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17454:10864',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18376:5134',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23836:8046',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18774:15374',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4020:10238',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32190:7795',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1732:1736',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14669:14215',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15067:22952',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26160:4648',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31313:11443',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11062:2613',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7211:28667',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19958:8218',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24813:12555',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22770:4100',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21368:9752',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6072:17143',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11921:7138',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16667:5368',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22661:11616',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27019:6793',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16685:4178',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19461:26678',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32841:26741',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21947:18364',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14823:22654',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22706:6558',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23439:13902',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6533:3223',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31277:19867',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10628:5212',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25111:18803',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23710:2691',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:29116:28902',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7726:11209',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30074:9330',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16278:17159',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2618:9032',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6099:26772',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18692:4993',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10881:13698',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6171:28588',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5529:4554',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28917:6637',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26964:13088',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25907:4335',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23104:19398',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15763:5682',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11288:24893',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31051:10269',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4417:13463',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14172:15546',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28908:20087',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25409:16094',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22489:26412',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18511:9846',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23041:18662',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16170:24925',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30843:26068',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11939:2253',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19949:10457',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25635:10786',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26151:24486',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7771:23688',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12680:11678',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17047:23844',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22598:21433',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18322:1532',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:32434:18364',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27344:3192',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18918:7232',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17580:13025',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31195:8108',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31783:13667',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:6696:2221',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25861:8547',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12707:11036',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21838:3270',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1289:28526',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22905:10128',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16577:24158',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19407:18380',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2230:9048',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20066:15076',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31756:9831',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21938:5822',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18367:15734',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15917:2628',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11849:24706',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21576:9455',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2935:8140',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10511:18818',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20672:15969',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17119:6621',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17951:9095',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21178:13777',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15374:28432',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16848:21559',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15275:12352',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28320:4914',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16776:20181',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15817:19492',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31367:3615',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23176:22216',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25482:3944',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4354:24502',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7762:10081',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4372:27352',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30273:14653',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8250:3067',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22245:13964',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17616:8766',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30915:17707',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25382:11663',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24478:11631',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4526:2409',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7057:22514',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26576:5400',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10402:22075',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21404:6496',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:30617:20572',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13774:28541',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:16712:12587',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5385:5588',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13386:16282',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9588:11992',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24496:5525',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4453:13244',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3595:9972',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11180:16094',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20473:13119',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27145:22983',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18747:10974',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18123:26177',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20708:23547',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9191:18599',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13042:15311',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3441:9831',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13648:19335',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23954:26036',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:31259:27508',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10646:11506',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22923:28322',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15148:28635',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19560:23563',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17842:24189',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14045:15233',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5683:18443',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11840:16047',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:20817:6042',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25536:20791',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19804:7263',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13114:19163',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:14217:14622',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:25934:9956',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28339:29058',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9290:11005',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1687:29089',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23041:2785',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9245:1564',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2700:15593',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24560:5165',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:21395:16689',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:19126:10410',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27850:9236',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23393:13980',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:9607:22420',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11776:10394',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13575:10848',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:23900:11036',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17571:1642',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17155:3740',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28366:11412',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27480:4805',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17155:4241',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13901:9972',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:22290:15796',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27661:1391',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28881:27492',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8467:13933',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5963:26475',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:3956:1767',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:10691:4758',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11984:10567',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:12472:23469',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4237:5447',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:1262:15483',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:8431:1376',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27805:19492',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:2908:21903',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:26693:28244',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:5385:19460',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7979:9455',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:11668:27492',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27742:5979',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24885:8578',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:24080:8218',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:17716:2738',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:15058:14137',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:13630:15859',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:4209:7498',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18846:22326',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:28971:4288',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:27670:26710',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7581:17159',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:7961:21167',\n",
       " 'NGSNJ-086:229:GW200110425th:1:1101:18331:14137',\n",
       " ...}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fivep_passed_rids = pd.read_csv(f'{output_base}fivep_info_table.tsv', sep='\\t').read_id\n",
    "fivep_passed_rids = set(fivep_passed_rids.transform(lambda rid: rid[:-4].split('/')[0]))\n",
    "fivep_passed_rids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'temp_switch_rids' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[27], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(lariat_rids\u001b[38;5;241m.\u001b[39mintersection(circular_rids)) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m, lariat_rids\u001b[38;5;241m.\u001b[39mintersection(circular_rids)\n\u001b[0;32m----> 2\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(lariat_rids\u001b[38;5;241m.\u001b[39mintersection(\u001b[43mtemp_switch_rids\u001b[49m)) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m, lariat_rids\u001b[38;5;241m.\u001b[39mintersection(temp_switch_rids)\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(temp_switch_rids\u001b[38;5;241m.\u001b[39mintersection(circular_rids)) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m, temp_switch_rids\u001b[38;5;241m.\u001b[39mintersection(circular_rids)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'temp_switch_rids' is not defined"
     ]
    }
   ],
   "source": [
    "assert len(lariat_rids.intersection(circular_rids)) == 0, lariat_rids.intersection(circular_rids)\n",
    "assert len(lariat_rids.intersection(temp_switch_rids)) == 0, lariat_rids.intersection(temp_switch_rids)\n",
    "assert len(temp_switch_rids.intersection(circular_rids)) == 0, temp_switch_rids.intersection(circular_rids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "read_class = []\n",
    "for read_id in fivep_passed_rids:\n",
    "\tif read_id in lariat_rids:\n",
    "\t\tread_class.append((read_id, 'Lariat'))\n",
    "\telif read_id in lariat_failed.loc[lariat_failed.filter_failed.isin(('in_repeat', 'ubiquitin_gene'))]:\n",
    "\t\tread_class.append((read_id, 'Repeat region'))\n",
    "\telif read_id in circular_rids:\n",
    "\t\tread_class.append((read_id, 'Circularized intron'))\n",
    "\telif read_id in temp_switch_rids:\n",
    "\t\tread_class.append((read_id, 'Template-switching'))\n",
    "\telse:\n",
    "\t\tread_class.append((read_id, \"Has 5'ss alignment\"))\n",
    "\n",
    "read_class = pd.DataFrame(read_class, columns='read_id')\n",
    "read_class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "read_class_counts = read_class.value_counts().to_dict()\n",
    "read_class_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lariat_passed = pd.read_csv(f'{out_dir}/lariat_reads.tsv', sep='\\t')\n",
    "lariat_passed = lariat_passed.set_index('read_id', verify_integrity=True)\n",
    "\n",
    "matched_introns = []\n",
    "for rid, row in lariat_passed.iterrows():\n",
    "\tmatches = introns.loc[(introns.chrom==row['chrom']) & \n",
    "\t\t\t\t\t\t(introns.strand==row['strand']) & \n",
    "\t\t\t\t\t\t(introns.fivep_pos==row['fivep_pos']) & \n",
    "\t\t\t\t\t\t(introns.start<row['bp_pos']) & \n",
    "\t\t\t\t\t\t(introns.end>row['bp_pos'])]\n",
    "\tmatched_introns.append(tuple(matches.coords))\n",
    "lariat_passed['introns'] = matched_introns\n",
    "\n",
    "lariat_passed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lariat_failed = pd.read_csv(f'{out_dir}/failed_lariat_mappings.tsv', sep='\\t')\n",
    "\n",
    "matched_introns = []\n",
    "for rid, row in lariat_failed.iterrows():\n",
    "\tmatches = introns.loc[(introns.chrom==row['chrom']) & \n",
    "\t\t\t\t\t\t(introns.strand==row['strand']) & \n",
    "\t\t\t\t\t\t(introns.fivep_pos==row['fivep_pos']) & \n",
    "\t\t\t\t\t\t(introns.start<row['bp_pos']) & \n",
    "\t\t\t\t\t\t(introns.end>row['bp_pos'])]\n",
    "\tmatched_introns.append(tuple(matches.coords))\n",
    "lariat_failed['introns'] = matched_introns\n",
    "\n",
    "lariat_failed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_fivep_sites(sites:str):\n",
    "\tout = []\n",
    "\tfor site in sites.split(','):\n",
    "\t\tchrom, start, end, strand = site.split(';')\n",
    "\t\tstart = int(start)\n",
    "\t\tend = int(end)\n",
    "\t\tout.append((chrom, start, end, strand))\n",
    "\n",
    "\treturn tuple(out)\n",
    "\n",
    "fivep_passed = pd.read_csv(f'{out_dir}/fivep_info_table.tsv', sep='\\t')\n",
    "fivep_passed.read_id = fivep_passed.read_id.str.slice(0, -4)\n",
    "fivep_passed.fivep_sites = fivep_passed.fivep_sites.transform(process_fivep_sites)\n",
    "fivep_passed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fivep_failed = pd.read_csv(f'{out_dir}/failed_fivep_alignments.tsv', sep='\\t')\n",
    "fivep_failed.fivep_site = fivep_failed.fivep_site.transform(process_fivep_sites)\n",
    "assert fivep_failed.fivep_site.transform(len).eq(1).all()\n",
    "fivep_failed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trim_failed = pd.read_csv(f'{out_dir}/failed_trimmed_alignments.tsv', sep='\\t')\n",
    "trim_failed.read_id = trim_failed.read_id.str.slice(0, -4)\n",
    "trim_failed = pd.merge(trim_failed, fivep_passed[['read_id', 'read_is_reverse', 'fivep_sites', 'fivep_seq']], 'left', on=['read_id', 'read_is_reverse'], validate='many_to_one')\n",
    "print(trim_failed.filter_failed.value_counts())\n",
    "trim_failed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n=2\n",
    "print(trim_failed.at[0, 'trim_seq'][-n:])\n",
    "print(trim_failed.at[0, 'fivep_seq'][:n])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len('GTAAGTAGTGCTTGTGCTCA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def seq_match(row, n:int):\n",
    "\tif not row['align_is_reverse']:\n",
    "\t\treturn row['trim_seq'][-n:]==row['fivep_seq'][:n]\n",
    "\telse:\t\n",
    "\t\treturn Seq.reverse_complement(row['trim_seq'][:n])==row['fivep_seq'][:n]\n",
    "\t\n",
    "\n",
    "\t\n",
    "trim_failed['match_1'] = trim_failed.apply(seq_match, n=1, axis=1)\n",
    "print(trim_failed.match_1.value_counts())\n",
    "trim_failed['match_2'] = trim_failed.apply(seq_match, n=2, axis=1)\n",
    "print(trim_failed.match_2.value_counts())\n",
    "trim_failed['match_3'] = trim_failed.apply(seq_match, n=3, axis=1)\n",
    "print(trim_failed.match_3.value_counts())\n",
    "trim_failed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unspliced_dict = unspliced_aligns.species.to_dict()\n",
    "spliced_dict = spliced_aligns[['read_id', 'species']].drop_duplicates().set_index('read_id')['species'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_rids = []\n",
    "with gzip.open(input_fastq) as w:\n",
    "\tfor read_id_line in w:\n",
    "\t\trid = read_id_line.decode()[1:].split(' ')[0]\n",
    "\t\tinput_rids.append(rid)\n",
    "\t\tw.readline()\n",
    "\t\tw.readline()\n",
    "\t\tw.readline()\n",
    "print(len(input_rids))\n",
    "input_rids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "BAD_TRIM_FILTERS = ('mismatch_number', 'mismatch_percent', 'indel')\n",
    "\n",
    "read_species = []\n",
    "for rid in input_rids:\n",
    "\t# Passed lariat filtering\n",
    "\tif rid in lariat_passed.index:\n",
    "\t\tread_species.append((rid, 'Lariat'))\n",
    "\n",
    "\t# Passed trim filtering, failed final lariat filtering\n",
    "\telif rid in lariat_failed.read_id.values:\n",
    "\t\tfails = set(lariat_failed.loc[lariat_failed.read_id==rid, 'filter_failed'])\n",
    "\t\tassert len(fails)==1, f'{source},{rid},{fails}'\n",
    "\t\tthe_filter = fails.pop()\n",
    "\n",
    "\t\tif the_filter == 'near_ss':\n",
    "\t\t\tread_species.append((rid, 'Circular intron'))\n",
    "\t\telif the_filter in ('ubiquitin_gene', 'in_repeat'):\n",
    "\t\t\tread_species.append((rid, 'Repeat region'))\n",
    "\t\telse:\n",
    "\t\t\traise RuntimeError(rid)\n",
    "\n",
    "\t# Trim seq(s) mapped (not counting output alignments that were filtered out for bad quality), failed trim filtering\n",
    "\t# elif rid in trim_failed.loc[~trim_failed.filter_failed.isin(BAD_TRIM_FILTERS), 'read_id'].values:\n",
    "\telif rid in trim_failed.read_id.values:\n",
    "\t\tread_species.append((rid, 'Template switching'))\n",
    "\t\n",
    "\t# Mapped linearly, no trim seq(s) mapped\n",
    "\telif rid in unspliced_dict:\n",
    "\t\tread_species.append((rid, unspliced_dict[rid]))\n",
    "\telif rid in spliced_dict:\n",
    "\t\tread_species.append((rid, spliced_dict[rid]))\n",
    "\t\n",
    "\t# 5'ss mapped to it, no linear or trim seq(s) mapped\n",
    "\telif rid in fivep_passed.read_id.values or rid in fivep_failed.read_id.values:\n",
    "\t\tread_species.append((rid, \"Unmapped, has a 5'ss alignment\"))\n",
    "\n",
    "\t# No 5'ss mapped to it and no linear alignments\n",
    "\telse:\n",
    "\t\tread_species.append((rid, \"Unmapped\"))\n",
    "\n",
    "\n",
    "read_species = pd.DataFrame(read_species, columns=['read_id', 'species'])\n",
    "assert read_species.read_id.is_unique\n",
    "read_species = read_species.set_index('read_id')\n",
    "print(read_species.species.value_counts().sort_index())\n",
    "read_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unspliced_aligns['introns'] = unspliced_aligns.apply(lambda row: tuple(set([(row['chrom'], feat.begin, feat.end, feat.data[3]) for feat in row['filtered_features'] if feat.data[0]=='intron'])), axis=1)\n",
    "\n",
    "unspliced_aligns['exon_annos'] = unspliced_aligns.filtered_features.transform(lambda features: set([(feat.data[1], feat.data[2]) for feat in features if feat.data[0]=='exon']))\n",
    "merged = pd.merge(unspliced_aligns.loc[unspliced_aligns.species=='Exonic'].explode('exon_annos'), exploded_introns.explode('adj_exons'), left_on='exon_annos', right_on='adj_exons')\n",
    "merged = merged.groupby('read_id').coords.agg(lambda coords: tuple(set(coords)))\n",
    "unspliced_aligns.loc[unspliced_aligns.species=='Exonic', 'introns'] = unspliced_aligns.loc[unspliced_aligns.species=='Exonic', 'read_id'].map(merged)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spliced_aligns['introns'] = spliced_aligns.apply(lambda row: tuple(set([(row['chrom'], feat.begin, feat.end, feat.data[3]) for feat in row['filtered_seg_features'] if feat.data[0]=='intron'])), axis=1)\n",
    "\n",
    "spliced_aligns['exon_annos'] = spliced_aligns.filtered_seg_features.transform(lambda features: set([(feat.data[1], feat.data[2]) for feat in features if feat.data[0]=='exon']))\n",
    "merged = pd.merge(spliced_aligns.loc[spliced_aligns.seg_species=='Exonic'].explode('exon_annos'), exploded_introns.explode('adj_exons'), left_on='exon_annos', right_on='adj_exons')\n",
    "merged = merged.groupby('read_id').coords.agg(lambda coords: tuple(set(coords)))\n",
    "spliced_aligns.loc[spliced_aligns.seg_species=='Exonic', 'introns'] = spliced_aligns.loc[spliced_aligns.seg_species=='Exonic', 'read_id'].map(merged)\n",
    "spliced_aligns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_ends_at_fivep_introns(row, df:str):\n",
    "\tmatches = set()\n",
    "\tstart = row['seg_start'] if df=='spliced' else row['align_start']\n",
    "\tend = row['seg_end'] if df=='spliced' else row['align_end']\n",
    "\tfor intron in row['introns']:\n",
    "\t\tif intron[-1]=='+' and intron[1]==start:\n",
    "\t\t\tmatches.add(intron)\n",
    "\t\telif intron[-1]=='-' and intron[2]==end:\n",
    "\t\t\tmatches.add(intron)\n",
    "\n",
    "\treturn tuple(matches)\n",
    "\n",
    "\n",
    "\n",
    "spliced_aligns.loc[spliced_aligns.species==\"Ends at a 5'ss\", 'introns'] = spliced_aligns.loc[spliced_aligns.species==\"Ends at a 5'ss\"].apply(filter_ends_at_fivep_introns,  df='spliced', axis=1)\n",
    "unspliced_aligns.loc[unspliced_aligns.species==\"Ends at a 5'ss\", 'introns'] = unspliced_aligns.loc[unspliced_aligns.species==\"Ends at a 5'ss\"].apply(filter_ends_at_fivep_introns, df='unspliced', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_introns = []\n",
    "# def get_introns(, lariat_passed, lariat_failed, unspliced_aligns, spliced_aligns, )\n",
    "for rid in read_species.index:\n",
    "\tspecies = read_species.at[rid, 'species']\n",
    "\t\n",
    "\tif species in ('Inter-genic', 'Unmapped'):\n",
    "\t\tmatched_introns.append(())\n",
    "\n",
    "\telif species == 'Lariat':\n",
    "\t\tmatched_introns.append(lariat_passed.at[rid, 'introns'])\n",
    "\n",
    "\telif species == 'Circular intron':\n",
    "\t\tmatches = lariat_failed.loc[(lariat_failed.read_id==rid) & (lariat_failed.filter_failed=='near_ss'), 'introns']\n",
    "\t\tmatches = set([feat for feats in matches for feat in feats])\n",
    "\t\tmatched_introns.append(tuple(matches))\n",
    "\n",
    "\telif species == 'Repeat region':\n",
    "\t\tmatches = lariat_failed.loc[(lariat_failed.read_id==rid) & (lariat_failed.filter_failed.isin(('in_repeat', 'ubiquitin_gene'))), 'introns']\n",
    "\t\tmatches = set([feat for feats in matches for feat in feats])\n",
    "\t\tmatched_introns.append(tuple(matches))\n",
    "\n",
    "\telif species == 'Template switching':\n",
    "\t\t# rows = trim_failed.loc[(trim_failed.read_id==rid) & (~trim_failed.filter_failed.isin(BAD_TRIM_FILTERS))]\n",
    "\t\t# introns_from_rows = set()\n",
    "\t\t# for i, row in rows.iterrows():\n",
    "\t\t# \tfor fivep_chrom, fivep_start, fivep_end, fivep_strand in row['fivep_sites']:\n",
    "\t\t# \t\tmatches = introns.loc[(introns.chrom==fivep_chrom) & (introns.strand==fivep_strand)]\n",
    "\t\t# \t\tmatches = matches.loc[((matches.strand=='+') & (matches.start==fivep_start)) | ((matches.strand=='-') & (matches.end-1==fivep_end))]\n",
    "\t\t# \t\tintrons_from_rows.update(matches.coords)\n",
    "\t\t# matched_introns.append(tuple(introns_from_rows))\n",
    "\t\tfivep_sites = trim_failed.loc[trim_failed.read_id==rid, 'fivep_sites']\n",
    "\t\tfivep_sites = set([site for sites in fivep_sites for site in sites])\n",
    "\t\tall_matches = set()\n",
    "\t\tfor fivep_chrom, fivep_start, fivep_end, fivep_strand in fivep_sites:\n",
    "\t\t\tmatches = introns.loc[(introns.chrom==fivep_chrom) & (introns.strand==fivep_strand)]\n",
    "\t\t\tmatches = matches.loc[((matches.strand=='+') & (matches.fivep_pos==fivep_start)) | ((matches.strand=='-') & (matches.fivep_pos==fivep_end-1))]\n",
    "\t\t\tall_matches.update(matches.coords)\n",
    "\t\tmatched_introns.append(tuple(all_matches))\n",
    "\t\n",
    "\telif species ==  \"Unmapped, has a 5'ss alignment\":\n",
    "\t\tfivep_sites = fivep_failed.loc[(fivep_failed.read_id==rid) & (fivep_failed.fail_reason!='furthest_upstream'), 'fivep_site']\n",
    "\t\tfivep_sites = pd.concat([fivep_sites, fivep_passed.loc[fivep_passed.read_id==rid, 'fivep_sites']])\n",
    "\t\tfivep_sites = set([site for sites in fivep_sites for site in sites])\n",
    "\t\tall_matches = set()\n",
    "\t\tfor fivep_chrom, fivep_start, fivep_end, fivep_strand in fivep_sites:\n",
    "\t\t\tmatches = introns.loc[(introns.chrom==fivep_chrom) & (introns.strand==fivep_strand)]\n",
    "\t\t\tmatches = matches.loc[((matches.strand=='+') & (matches.fivep_pos==fivep_start)) | ((matches.strand=='-') & (matches.fivep_pos==fivep_end-1))]\n",
    "\t\t\tall_matches.update(matches.coords)\n",
    "\t\tmatched_introns.append(tuple(all_matches))\n",
    "\n",
    "\telif species == 'Genic, ambiguous':\n",
    "\t\tif rid in unspliced_aligns.index:\n",
    "\t\t\tmatched_introns.append(unspliced_aligns.at[rid, 'introns'])\n",
    "\t\telse:\n",
    "\t\t\tmatches = spliced_aligns.loc[(spliced_aligns.read_id==rid) & (spliced_aligns.introns.notna()), 'introns']\n",
    "\t\t\tmatches = set([feat for feats in matches for feat in feats])\n",
    "\t\t\tmatched_introns.append(tuple(matches))\n",
    "\n",
    "\t# For exonic, match introns adjacent to the exon\n",
    "\telif species == 'Exonic':\n",
    "\t\tif pd.isna(unspliced_aligns.at[rid, 'introns']):\n",
    "\t\t\tmatched_introns.append(())\n",
    "\t\telse:\n",
    "\t\t\tmatched_introns.append(unspliced_aligns.at[rid, 'introns'])\n",
    "\n",
    "\telif species == 'Intronic':\n",
    "\t\tif rid in unspliced_aligns.index:\n",
    "\t\t\tmatched_introns.append(unspliced_aligns.at[rid, 'introns'])\n",
    "\t\telse:\n",
    "\t\t\tmatches = spliced_aligns.loc[spliced_aligns.read_id==rid, 'introns']\n",
    "\t\t\tmatches = set([feat for feats in matches for feat in feats])\n",
    "\t\t\tmatched_introns.append(tuple(matches))\n",
    "\n",
    "\t# For mRNA, match introns for which both adjacent exons are covered\n",
    "\telif species == 'mRNA':\n",
    "\t\t# Gather adjacent exons\n",
    "\t\texons = spliced_aligns.loc[spliced_aligns.read_id==rid, 'filtered_seg_features']\n",
    "\t\texons = [feat for feats in exons for feat in feats]\n",
    "\t\texons = [feat for feat in exons if feat.data[0]=='exon']\n",
    "\t\texon_pairs = set()\n",
    "\t\tfor exon_1, exon_2 in it.product(exons, exons):\n",
    "\t\t\tif exon_1.data[1] == exon_2.data[1] and exon_1.data[2] == exon_2.data[2]-1:\n",
    "\t\t\t\texon_pairs.add( ((exon_1.data[1], exon_1.data[2]), (exon_2.data[1], exon_2.data[2])) )\n",
    "\t\t\t\t# exon_pairs = introns.loc[introns.adj_exons.transform(lambda adj_exons: exon_1_anno in adj_exons and exon_2_anno in adj_exons)]\n",
    "\t\t\t\t# all_matches.update(matches.coords)\n",
    "\t\t\t# exon_1_anno = (exon_1.data[1], exon_1.data[2]) \n",
    "\t\t\t# exon_2_anno = (exon_2.data[1], exon_2.data[2]) \n",
    "\t\tmatches = exploded_introns.loc[exploded_introns.adj_exons.isin(exon_pairs), 'coords']\n",
    "\t\tmatches = set(matches)\n",
    "\t\tmatched_introns.append(tuple(matches))\n",
    "\n",
    "\telif species == 'pre-mRNA':\n",
    "\t\tif rid in unspliced_aligns.index:\n",
    "\t\t\t# row = unspliced_aligns.loc[rid]\n",
    "\t\t\t# matches = [feat for feat in row['filtered_features'] if feat.data[0]=='intron']\n",
    "\t\t\t# matched_introns.append(tuple(matches))\n",
    "\t\t\tmatched_introns.append(unspliced_aligns.at[rid, 'introns'])\n",
    "\t\telse:\n",
    "\t\t\t# rows = spliced_aligns.loc[(spliced_aligns.read_id==rid) & (spliced_aligns.seg_species=='pre-mRNA')]\n",
    "\t\t\t# matches = [feat for feats in rows['filtered_seg_features'] for feat in feats]\n",
    "\t\t\t# matches = [feat for feat in matches if feat.data[0]=='intron']\n",
    "\t\t\tmatches = spliced_aligns.loc[spliced_aligns.read_id==rid, 'introns']\n",
    "\t\t\tmatches = set([feat for feats in matches for feat in feats])\n",
    "\t\t\tmatched_introns.append(tuple(matches))\n",
    "\n",
    "\telif species == \"Ends at a 5'ss\":\n",
    "\t\tif rid in unspliced_aligns.index:\n",
    "\t\t\tmatched_introns.append(unspliced_aligns.at[rid, 'introns'])\n",
    "\t\telse:\n",
    "\t\t\tmatches = spliced_aligns.loc[spliced_aligns.read_id==rid, 'introns']\n",
    "\t\t\tmatches = set([feat for feats in matches for feat in feats])\n",
    "\t\t\tmatched_introns.append(tuple(matches))\n",
    "\n",
    "\telse:\n",
    "\t\traise RuntimeError(rid, species)\n",
    "\n",
    "read_species['introns'] = matched_introns\n",
    "read_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read_species['intron_count'] = read_species.introns.transform(len)\n",
    "for species in READ_SPECIES:\n",
    "\tprint(species)\n",
    "\tprint(read_species.loc[read_species.species==species, ['species', 'intron_count']].value_counts().sort_index(), '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "read_species = read_species.explode('introns').rename(columns={'introns': 'intron'}).reset_index()\n",
    "read_species.loc[read_species.intron.notna(), ['chrom', 'intron_start', 'intron_end', 'strand']] = read_species[read_species.intron.notna()].intron.to_list()\n",
    "read_species.intron_start = read_species.intron_start.astype(int, errors='ignore')\n",
    "read_species.intron_end = read_species.intron_end.astype(int, errors='ignore')\n",
    "read_species.to_csv(f'/home/tmooney/Lariat_mapping/testing/output/100k_truncs/{samp}_{r_dir}k_lariat_mapping/read_species.tsv', sep='\\t')\n",
    "read_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "read_species[read_species.duplicated(keep=False)].sort_values(['read_id', 'species', 'intron'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exit()\n",
    "d = done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read_species = pd.read_csv('/home/tmooney/Lariat_mapping/testing/output/100k_truncs/C22-1_R1_100k_lariat_mapping/read_species.tsv', sep='\\t')\n",
    "read_species = read_species.drop_duplicates()\n",
    "print(read_species.info())\n",
    "read_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "read_total = 100_000\n",
    "per_intron_counts = read_species.groupby(['intron', 'species'], as_index=False).agg('size')\n",
    "per_intron_counts = per_intron_counts.rename(columns={'size': 'reads'})\n",
    "per_intron_counts['read_prop'] = per_intron_counts.reads / read_total\n",
    "per_intron_counts['reads_total'] = per_intron_counts.intron.map(per_intron_counts.groupby('intron').reads.agg('sum'))\n",
    "per_intron_counts['species_rep'] = per_intron_counts.reads / per_intron_counts.reads_total\n",
    "per_intron_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "per_intron_counts.sort_values('reads_total', ascending=False)[:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = read_species.species.value_counts().reset_index()\n",
    "df = df.rename(columns={'count': 'reads'})\n",
    "df.species = pd.Categorical(df.species, READ_SPECIES + ('Dunno',), ordered=True)\n",
    "df['source'] = source\n",
    "print([row.to_list() for i, row in df.iterrows()])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = [['Intronic', 29109, 'C22-1|R1_100'], ['Exonic', 20108, 'C22-1|R1_100'], ['Inter-genic', 16477, 'C22-1|R1_100'], ['Unmapped', 15668, 'C22-1|R1_100'], ['Genic, ambiguous', 13954, 'C22-1|R1_100'], ['mRNA', 1619, 'C22-1|R1_100'], ['Template switching', 1490, 'C22-1|R1_100'], [\"Unmapped, has a 5'ss alignment\", 968, 'C22-1|R1_100'], ['pre-mRNA', 590, 'C22-1|R1_100'], [\"Ends at a 5'ss\", 10, 'C22-1|R1_100'], ['Lariat', 5, 'C22-1|R1_100'], ['Repeat region', 2, 'C22-1|R1_100']] + \\\n",
    "\t[['Intronic', 28879, 'C22-1|R2_100'], ['Exonic', 20098, 'C22-1|R2_100'], ['Inter-genic', 16433, 'C22-1|R2_100'], ['Unmapped', 15848, 'C22-1|R2_100'], ['Genic, ambiguous', 14093, 'C22-1|R2_100'], ['mRNA', 1578, 'C22-1|R2_100'], ['Template switching', 1563, 'C22-1|R2_100'], [\"Unmapped, has a 5'ss alignment\", 918, 'C22-1|R2_100'], ['pre-mRNA', 577, 'C22-1|R2_100'], [\"Ends at a 5'ss\", 11, 'C22-1|R2_100'], ['Lariat', 2, 'C22-1|R2_100']] + \\\n",
    "\t[['Intronic', 29161, 'C22-3|R1_100'], ['Exonic', 20028, 'C22-3|R1_100'], ['Inter-genic', 16456, 'C22-3|R1_100'], ['Unmapped', 15978, 'C22-3|R1_100'], ['Genic, ambiguous', 13873, 'C22-3|R1_100'], ['mRNA', 1603, 'C22-3|R1_100'], ['Template switching', 1391, 'C22-3|R1_100'], [\"Unmapped, has a 5'ss alignment\", 919, 'C22-3|R1_100'], ['pre-mRNA', 587, 'C22-3|R1_100'], [\"Ends at a 5'ss\", 3, 'C22-3|R1_100'], ['Lariat', 1, 'C22-3|R1_100']] + \\\n",
    "\t[['Intronic', 28977, 'C22-3|R2_100'], ['Exonic', 20002, 'C22-3|R2_100'], ['Inter-genic', 16585, 'C22-3|R2_100'], ['Unmapped', 16052, 'C22-3|R2_100'], ['Genic, ambiguous', 13943, 'C22-3|R2_100'], ['mRNA', 1505, 'C22-3|R2_100'], ['Template switching', 1416, 'C22-3|R2_100'], [\"Unmapped, has a 5'ss alignment\", 933, 'C22-3|R2_100'], ['pre-mRNA', 582, 'C22-3|R2_100'], ['Repeat region', 3, 'C22-3|R2_100'], [\"Ends at a 5'ss\", 2, 'C22-3|R2_100']] + \\\n",
    "\t[['Intronic', 28434, 'C22-5|R1_100'], ['Exonic', 20399, 'C22-5|R1_100'], ['Inter-genic', 16801, 'C22-5|R1_100'], ['Unmapped', 15365, 'C22-5|R1_100'], ['Genic, ambiguous', 14349, 'C22-5|R1_100'], ['mRNA', 1716, 'C22-5|R1_100'], ['Template switching', 1349, 'C22-5|R1_100'], [\"Unmapped, has a 5'ss alignment\", 1034, 'C22-5|R1_100'], ['pre-mRNA', 541, 'C22-5|R1_100'], [\"Ends at a 5'ss\", 6, 'C22-5|R1_100'], ['Lariat', 3, 'C22-5|R1_100'], ['Repeat region', 2, 'C22-5|R1_100'], ['Circular intron', 1, 'C22-5|R1_100']] + \\\n",
    "\t[['Intronic', 28383, 'C22-5|R2_100'], ['Exonic', 20246, 'C22-5|R2_100'], ['Inter-genic', 16776, 'C22-5|R2_100'], ['Unmapped', 15470, 'C22-5|R2_100'], ['Genic, ambiguous', 14378, 'C22-5|R2_100'], ['mRNA', 1738, 'C22-5|R2_100'], ['Template switching', 1409, 'C22-5|R2_100'], [\"Unmapped, has a 5'ss alignment\", 1045, 'C22-5|R2_100'], ['pre-mRNA', 539, 'C22-5|R2_100'], [\"Ends at a 5'ss\", 8, 'C22-5|R2_100'], ['Repeat region', 4, 'C22-5|R2_100'], ['Lariat', 4, 'C22-5|R2_100']] + \\\n",
    "\t[['Exonic', 22887, 'HEK293T-1|R1_100'], ['Intronic', 22738, 'HEK293T-1|R1_100'], ['Inter-genic', 19406, 'HEK293T-1|R1_100'], ['Unmapped', 16508, 'HEK293T-1|R1_100'], ['Genic, ambiguous', 14170, 'HEK293T-1|R1_100'], ['mRNA', 1816, 'HEK293T-1|R1_100'], ['Template switching', 1102, 'HEK293T-1|R1_100'], [\"Unmapped, has a 5'ss alignment\", 776, 'HEK293T-1|R1_100'], ['pre-mRNA', 593, 'HEK293T-1|R1_100'], ['Repeat region', 2, 'HEK293T-1|R1_100'], ['Lariat', 1, 'HEK293T-1|R1_100'], [\"Ends at a 5'ss\", 1, 'HEK293T-1|R1_100']] + \\\n",
    "\t[['Exonic', 22616, 'HEK293T-1|R2_100'], ['Intronic', 21850, 'HEK293T-1|R2_100'], ['Inter-genic', 19052, 'HEK293T-1|R2_100'], ['Unmapped', 17342, 'HEK293T-1|R2_100'], ['Genic, ambiguous', 15088, 'HEK293T-1|R2_100'], ['mRNA', 1764, 'HEK293T-1|R2_100'], [\"Unmapped, has a 5'ss alignment\", 800, 'HEK293T-1|R2_100'], ['Template switching', 797, 'HEK293T-1|R2_100'], ['pre-mRNA', 684, 'HEK293T-1|R2_100'], [\"Ends at a 5'ss\", 4, 'HEK293T-1|R2_100'], ['Repeat region', 3, 'HEK293T-1|R2_100']] + \\\n",
    "\t[['Intronic', 26355, 'HEK293T-4|R1_100'], ['Exonic', 22348, 'HEK293T-4|R1_100'], ['Inter-genic', 17736, 'HEK293T-4|R1_100'], ['Unmapped', 14907, 'HEK293T-4|R1_100'], ['Genic, ambiguous', 14136, 'HEK293T-4|R1_100'], ['mRNA', 1741, 'HEK293T-4|R1_100'], ['Template switching', 1411, 'HEK293T-4|R1_100'], [\"Unmapped, has a 5'ss alignment\", 787, 'HEK293T-4|R1_100'], ['pre-mRNA', 576, 'HEK293T-4|R1_100'], [\"Ends at a 5'ss\", 2, 'HEK293T-4|R1_100'], ['Lariat', 1, 'HEK293T-4|R1_100']] + \\\n",
    "\t[['Intronic', 26151, 'HEK293T-4|R2_100'], ['Exonic', 22234, 'HEK293T-4|R2_100'], ['Inter-genic', 17563, 'HEK293T-4|R2_100'], ['Unmapped', 15186, 'HEK293T-4|R2_100'], ['Genic, ambiguous', 14297, 'HEK293T-4|R2_100'], ['mRNA', 1769, 'HEK293T-4|R2_100'], ['Template switching', 1406, 'HEK293T-4|R2_100'], [\"Unmapped, has a 5'ss alignment\", 822, 'HEK293T-4|R2_100'], ['pre-mRNA', 567, 'HEK293T-4|R2_100'], [\"Ends at a 5'ss\", 5, 'HEK293T-4|R2_100']] + \\\n",
    "\t[['Intronic', 26701, 'HEK293T-5|R1_100'], ['Exonic', 21825, 'HEK293T-5|R1_100'], ['Inter-genic', 17413, 'HEK293T-5|R1_100'], ['Unmapped', 15274, 'HEK293T-5|R1_100'], ['Genic, ambiguous', 14210, 'HEK293T-5|R1_100'], ['mRNA', 1780, 'HEK293T-5|R1_100'], ['Template switching', 1371, 'HEK293T-5|R1_100'], [\"Unmapped, has a 5'ss alignment\", 873, 'HEK293T-5|R1_100'], ['pre-mRNA', 550, 'HEK293T-5|R1_100'], ['Repeat region', 2, 'HEK293T-5|R1_100'], [\"Ends at a 5'ss\", 1, 'HEK293T-5|R1_100']] + \\\n",
    "\t[['Intronic', 26484, 'HEK293T-5|R2_100'], ['Exonic', 21783, 'HEK293T-5|R2_100'], ['Inter-genic', 17425, 'HEK293T-5|R2_100'], ['Unmapped', 15507, 'HEK293T-5|R2_100'], ['Genic, ambiguous', 14293, 'HEK293T-5|R2_100'], ['mRNA', 1744, 'HEK293T-5|R2_100'], ['Template switching', 1373, 'HEK293T-5|R2_100'], [\"Unmapped, has a 5'ss alignment\", 860, 'HEK293T-5|R2_100'], ['pre-mRNA', 528, 'HEK293T-5|R2_100'], [\"Ends at a 5'ss\", 2, 'HEK293T-5|R2_100'], ['Repeat region', 1, 'HEK293T-5|R2_100']]\n",
    "\n",
    "df = pd.DataFrame(df, columns=['species', 'reads', 'source'])\n",
    "for sou, species in it.product(df.source.unique(), READ_SPECIES):\n",
    "\tif len(df.loc[(df.source==sou) & (df.species==species)]) == 0:\n",
    "\t\tdf.loc[len(df)] = [species, 0, sou]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.species = pd.Categorical(df.species, READ_SPECIES, ordered=True)\n",
    "df[['samp', 'r_dir']] = df.source.str.split('|', expand=True)\n",
    "df['cell'] = df.samp.transform(lambda samp: samp.split('-')[0]).map({'HEK293T': 'WT', 'C22': 'DBR1-KO'})\n",
    "df = df[['cell', 'samp', 'r_dir', 'source', 'species', 'reads']]\n",
    "df['prop'] = df.reads / 100_000\n",
    "df = df.sort_values(['cell', 'samp', 'r_dir', 'species'], ignore_index=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mRNA_ratios = []\n",
    "for cell in df.cell.unique():\n",
    "\tfor s in df.loc[df.cell==cell, 'source'].unique():\n",
    "\t\tpre_count = df.loc[(df.source==s) & (df.species=='mRNA'), 'reads'].iloc[0]\n",
    "\t\tm_count = df.loc[(df.source==s) & (df.species=='pre-mRNA'), 'reads'].iloc[0]\n",
    "\t\tmRNA_ratios.append([cell, pre_count/m_count])\n",
    "\n",
    "mRNA_ratios = pd.DataFrame(mRNA_ratios, columns=['cell', 'ratio'])\n",
    "mRNA_ratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mann_whitney_u = stats.levene(mRNA_ratios.loc[mRNA_ratios.cell=='WT', 'ratio'], mRNA_ratios.loc[mRNA_ratios.cell=='DBR1-KO', 'ratio'])\n",
    "print(mann_whitney_u)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot = (p9.ggplot(mRNA_ratios, p9.aes(x=\"cell\", y=\"ratio\", fill='cell')) +\n",
    "\t\tp9.geom_boxplot() +\n",
    "\t\t# p9.geom_() +\n",
    "\t\t# p9.scale_x_() +\n",
    "\t\tp9.scale_y_continuous(limits=(2,4)) +\n",
    "\t\t# p9.scale_color_brewer(type='', palette=1, direction=1) +\n",
    "\t\t# p9.scale_fill_brewer(type='', palette=1, direction=1) +\n",
    "\t\tp9.scale_fill_brewer(type='qual', palette=2, direction=-1, guide=None) +\n",
    "\t\tp9.labs(title=f\"p={mann_whitney_u.pvalue:.3}\", x=\"\", y=\"mRNA reads / pre-mRNA reads\") +\n",
    "\t\tp9.theme(axis_title=p9.element_text(size=16), axis_text_x=p9.element_text(size=12), axis_text_y=p9.element_text(size=12))\n",
    ")\n",
    "plot.show()\n",
    "plot.save(PLOTS_DIR + '/premRNA_ratio.png', dpi=500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sum = df.groupby(['cell', 'species'])['reads'].agg(['mean', 'sem']).reset_index()\n",
    "df_sum "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_ratios = {}\n",
    "for species in df_sum.species.unique():\n",
    "\tc22_mean = df_sum.loc[(df_sum.cell=='DBR1-KO') & (df_sum.species==species), 'mean'].iloc[0] \n",
    "\tc22_sem = df_sum.loc[(df_sum.cell=='DBR1-KO') & (df_sum.species==species), 'sem'].iloc[0] \n",
    "\thek_mean = df_sum.loc[(df_sum.cell=='WT') & (df_sum.species==species), 'mean'].iloc[0] \n",
    "\thek_sem = df_sum.loc[(df_sum.cell=='WT') & (df_sum.species==species), 'sem'].iloc[0] \n",
    "\n",
    "\tratio = c22_mean/hek_mean if hek_mean!=0 else pd.NA\n",
    "\terror = ratio * np.sqrt((0.05*c22_sem/c22_mean)**2 + (0.05*hek_sem/hek_mean)**2) if pd.notna(ratio) else pd.NA\n",
    "\tspecies_ratios[species] = (ratio, error)\n",
    "species_ratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_ratios = pd.DataFrame(species_ratios)\n",
    "species_ratios = species_ratios.transpose().reset_index()\n",
    "species_ratios.columns = ['species', 'ratio', 'error']\n",
    "species_ratios['label'] = species_ratios.apply(lambda row: f\"{row['ratio']:.1%} ± {row['error']:.1%}\" if pd.notna(row['ratio']) else 'N/A', axis=1)\n",
    "species_ratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot = (p9.ggplot(df_sum, p9.aes(x=\"species\")) +\n",
    "\t\tp9.geom_col(p9.aes(y=\"mean\", fill='cell'), position='dodge') +\n",
    "\t\tp9.geom_errorbar(p9.aes(ymin=\"mean-sem\", ymax=\"mean+sem\", group='cell'), width=0.9, position='dodge') +\n",
    "\t\tp9.geom_text(data=species_ratios, mapping=p9.aes(x='species', y=35000, label='label'), ha='center', size=10) +\n",
    "\t\tp9.scale_y_continuous(limits=(0,38_000), breaks=np.arange(0,40_000,10_000), labels=comma_format()) +\n",
    "\t\t# p9.annotate('text', x=13.5, y=35000, label='Enrichment') +\n",
    "\t\tp9.scale_fill_brewer(type='qual', palette=2, direction=-1) +\n",
    "\t\tp9.labs(title=\"\", x=\"Read species\", y=\"Read count\", fill='Cell line') +\n",
    "\t\tp9.coord_flip() +\n",
    "\t\tp9.guides(fill=p9.guide_legend(reverse=True, position='right'), color=None) +\n",
    "\t\tp9.theme(axis_title=p9.element_text(size=16), axis_text=p9.element_text(size=10), legend_text=p9.element_text(size=10))\n",
    ")\n",
    "plot.save('/home/tmooney/Lariat_mapping/visuals/plots/species_100k.png', dpi=500, height=6, width=10)\n",
    "plot.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ##### Snippets #####\n",
    "# plot = (p9.ggplot(, p9.aes(x=\"\", y=\"\")) +\n",
    "# \t\t# p9.geom_() +\n",
    "# \t\t# p9.geom_() +\n",
    "# \t\t# p9.scale_x_() +\n",
    "# \t\t# p9.scale_y_() +\n",
    "# \t\t# p9.scale_color_brewer(type='', palette=1, direction=1) +\n",
    "# \t\t# p9.scale_fill_brewer(type='', palette=1, direction=1) +\n",
    "# \t\tp9.labs(title=\"\") +\n",
    "# \t\tp9.theme()\n",
    "# )\n",
    "# print(plot)\n",
    "# # plot.save(PLOTS_DIR + '/.png', dpi=500)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Discarded code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script false --no-raise-error\n",
    "# def process_features(features, ):\n",
    "# \tif features == ('.',) or features == ('.', '.'):\n",
    "# \t\treturn IntervalTree([])\n",
    "\t\n",
    "# \tout = []\n",
    "# \tfor feat in features:\n",
    "# \t\t# feat, tid, num, start, end, strand, gid, name, type_ = feat.split(';')\n",
    "# \t\tif type_ in ('artifact', 'TEC'):\n",
    "# \t\t\tcontinue\n",
    "# \t\telif type_ in NCRNA_TYPES:\n",
    "# \t\t\ttype_ = 'ncRNA'\n",
    "# \t\telif type_ in IG_TcR_TYPES:\n",
    "# \t\t\ttype_ = 'Ig or TcR'\n",
    "# \t\tstart = int(start)\n",
    "# \t\tend = int(end)\n",
    "# \t\tnum = int(num)\n",
    "# \t\tout.append(Interval(start, end, (feat, tid, num, strand, gid, name, type_)))\n",
    "\t\n",
    "# \treturn IntervalTree(out)\n",
    "\n",
    "\t\n",
    "\t\t# rows = lariat_failed.loc[lariat_failed.read_id==rid]\n",
    "\t\t# introns_from_rows = set()\n",
    "\t\t# for i, row in rows.iterrows():\n",
    "\t\t# \tmatches = introns.loc[(introns.chrom==row['chrom']) & (introns.strand==row['strand'])]\n",
    "\t\t# \tmatches = matches.loc[((matches.strand=='+') & (matches.start==row['fivep_pos'])) | ((matches.strand=='-') & (matches.end-1==row['fivep_pos']))]\n",
    "\t\t# \tmatches = matches.loc[((matches.strand=='+') & (matches.end-1-row['bp_pos']<=2)) | ((matches.strand=='-') & (row['bp_pos']-matches.start<=2))]\n",
    "\t\t# \tintrons_from_rows.update(matches.coords)\n",
    "\t\t# matched_introns.append(tuple(introns_from_rows))\n",
    "\n",
    "\t\t# rows = lariat_failed.loc[lariat_failed.read_id==rid]\n",
    "\t\t# introns_from_rows = set()\n",
    "\t\t# for i, row in rows.iterrows():\n",
    "\t\t# \tmatches = introns.loc[(introns.chrom==row['chrom']) & (introns.strand==row['strand'])]\n",
    "\t\t# \tmatches = matches.loc[((matches.strand=='+') & (matches.start==row['fivep_pos'])) | ((matches.strand=='-') & (matches.end-1==row['fivep_pos']))]\n",
    "\t\t# \tmatches = matches.loc[(matches.start<row['bp_pos']) & (matches.end>row['bp_pos'])]\n",
    "\t\t# \tintrons_from_rows.update(matches.coords)\n",
    "\t\t# matched_introns.append(tuple(introns_from_rows))\n",
    "\t\t\t# rows = spliced_aligns.loc[spliced_aligns.read_id==rid]\n",
    "\t\t\t# assert len(rows['chrom'].unique())==1\n",
    "\t\t\t# chrom = rows['chrom'].unique()[0]\n",
    "\t\t\t# matches = [feat for feats in rows['filtered_seg_features'] for feat in feats]\n",
    "\t\t\t# matches = [(chrom, feat.begin, feat.end, feat.data[3]) for feat in matches if feat.data[0]=='intron']\n",
    "\n",
    "\t\t# row = lariat_passed.loc[rid]\n",
    "\t\t# matches = introns.loc[(introns.chrom==row['chrom']) & (introns.strand==row['strand'])]\n",
    "\t\t# matches = matches.loc[((matches.strand=='+') & (matches.start==row['fivep_pos'])) | ((matches.strand=='-') & (matches.end-1==row['fivep_pos']))]\n",
    "\t\t# matches = matches.loc[(matches.start<row['bp_pos']) & (matches.end>row['bp_pos'])]\n",
    "\n",
    "\t\t\t\t# row = unspliced_aligns.loc[rid]\n",
    "\t\t# exons = [feat for feat in row['filtered_features'] if feat.data[0]=='exon']\n",
    "\t\t# matches = set()\n",
    "\t\t# for exon in exons:\n",
    "\t\t# \tfeat, tid, num, strand, gid, name, type_ = exon.data\n",
    "\t\t# \tm = introns.loc[(introns.transcript_id==tid) & ((introns.intron_num-num==0) | (introns.intron_num-num==-1))]\n",
    "\t\t# \tmatches.update(m['intron'])\n",
    "\t\t# matched_introns.append(tuple(matches))\n",
    "\t\t\n",
    "\t\t# if rid in unspliced_aligns.index:\n",
    "\t\t# \trow = unspliced_aligns.loc[rid]\n",
    "\t\t# \tmatches = [feat for feat in row['filtered_features'] if feat.data[0]=='intron']\n",
    "\t\t# \tmatched_introns.append(tuple(matches))\n",
    "\t\t# else:\n",
    "\t\t# \trows = spliced_aligns.loc[spliced_aligns.read_id==rid]\n",
    "\t\t# \tmatches = [feat for feats in rows['filtered_seg_features'] for feat in feats]\n",
    "\t\t# \tmatches = [feat for feat in matches if feat.data[0]=='intron']\n",
    "\t\t# \tmatched_introns.append(tuple(matches))\n",
    "\n",
    "\t\t# Get the intron between each adjacent exon pair\n",
    "\t\t# for exon_1, _ in adj_exons:\n",
    "\t\t\t# feat, tid, num, strand, gid, name, type_ = exon_1.data\n",
    "\t\t\t# m = introns.loc[(introns.transcript_id==tid) & (introns.intron_num==num)]\n",
    "\t\t\t# assert len(m) == 1, m\n",
    "\t\t\t# matches.update(m.intron)\n",
    "\n",
    "\t\t\t\n",
    "# BAD_TRIM_FILTERS = set(('mismatch_number', 'mismatch_percent', 'indel'))\n",
    "# trim_mapped_rids = set()\n",
    "# for align in pysam.AlignmentFile(f'{out_dir}/trimmed_reads_to_genome.bam', 'rb'):\n",
    "# \ttrim_mapped_rids.add(align.query_name[:-4])\n",
    "\n",
    "# # Get rid of those that mapped but got filtered out due to low quality alignment\n",
    "# rids_to_remove = set()\n",
    "# for rid in trim_failed.read_id.unique():\n",
    "# \tfail_reasons = set(trim_failed.loc[trim_failed.read_id==rid, 'filter_failed'])\n",
    "# \tif fail_reasons.issubset(BAD_ALIGNMENT_FILTERS):\n",
    "# \t\trids_to_remove.add(rid)\n",
    "\n",
    "# trim_mapped_rids = trim_mapped_rids.difference(rids_to_remove)\n",
    "# trim_mapped_rids\n",
    "\n",
    "plot = (p9.ggplot(df, p9.aes(x=\"species\")) +\n",
    "\t\tp9.geom_col(p9.aes(y=\"reads\")) +\n",
    "\t\tp9.scale_fill_brewer(type='qual', palette=2, direction=-1) +\n",
    "\t\tp9.labs(title=source) +\n",
    "\t\tp9.coord_flip() +\n",
    "\t\tp9.theme()\n",
    ")\n",
    "print(plot)\n",
    "\n",
    "# mapped_species_counts = []\n",
    "# for key, val in unspliced_aligns['species'].value_counts().to_dict().items():\n",
    "# \tmapped_species_counts.append([source, 'unspliced', key, val])\n",
    "# for key, val in spliced_aligns[['read_id', 'align_id', 'species']].drop_duplicates()['species'].value_counts().to_dict().items():\n",
    "# \tmapped_species_counts.append([source, 'spliced', key, val])\n",
    "# mapped_species_counts\n",
    "\n",
    "\n",
    "# run_data = pd.read_csv(f'{out_dir}/run_data.tsv', sep='\\t', header=None)\n",
    "# total_reads = run_data.loc[run_data[0]=='ref_mapped_reads', 1].iloc[0] + run_data.loc[run_data[0]=='ref_unmapped_reads', 1].iloc[0]\n",
    "# run_data.loc[len(run_data)] = ['total_reads', total_reads] \n",
    "\n",
    "# run_data = run_data.rename(columns={0:'cat', 1:'reads'}).reset_index(drop=True)\n",
    "# run_data.cat = run_data.cat.str.replace('_reads', '')\n",
    "# run_data\n",
    "# READ_SPECIES =  ('Lariat', 'Repeat region', 'Circular intron', 'Template switching', \"Unknown, has at least one 5'ss alignment\", \"Unknown, no 5'ss alignments\")\n",
    "# read_species_counts = []\n",
    "# unmapped = run_data.loc[run_data.cat=='ref_unmapped', 'reads'].iloc[0]\n",
    "# # no_fivep_maps = unmapped - run_data.loc[run_data.cat=='fivep_mapped', 'reads'].iloc[0]\n",
    "\n",
    "# for species in READ_SPECIES:\n",
    "# \t# if species == \"Unknown, no 5'ss alignments\":\n",
    "# \t\t# species_counts.append([source, unmapped, species, no_fivep_maps])\n",
    "# \t\t# species_counts.append([source, unmapped, species])\n",
    "# \t\t# continue\n",
    "\t\t\n",
    "# \tcount = len(read_species.loc[read_species_counts.species==species])\n",
    "# \tread_species_counts.append([source, unmapped, species, count])\n",
    "\n",
    "# read_species_counts = pd.DataFrame(read_species_counts, columns=['source', 'unmapped', 'species', 'reads'])\n",
    "# read_species_counts['species'] = pd.Categorical(read_species_counts.species, READ_SPECIES, ordered=True)\n",
    "# read_species_counts['prop'] = read_species_counts.reads / read_species_counts.unmapped\n",
    "# read_species_counts\n",
    "# m = pd.merge(read_species, unspliced_aligns[['read_id', 'species', 'filtered_gene_types']], 'inner', on='read_id')\n",
    "# print(m[['species_x', 'species_y']].value_counts().sort_index())\n",
    "# m\n",
    "# species_counts_avg = species_counts.groupby(['samp', 'species'])[['reads', 'prop']].agg('mean').reset_index()\n",
    "# species_counts_avg['label'] = species_counts_avg.reads.transform(lambda val: f'{val:,.0f}')\n",
    "# species_counts_avg\n",
    "# plot = (p9.ggplot(species_counts, p9.aes(x=\"species\", fill='samp')) +\n",
    "# \t\tp9.geom_col(p9.aes(y=\"prop\"), position='dodge') +\n",
    "# \t\t# p9.geom_text(p9.aes(y=\"prop+0.05\", label=\"label\", color='samp'), ha='left', size=8, position=p9.position_dodge(width=1)) +\n",
    "# \t\t# p9.annotate('text', x=6.5, y=1.05, label='Avg. reads', fontweight='bold') +\n",
    "# \t\tp9.scale_y_continuous(limits=(0,1), breaks=np.arange(0,1.2,0.2)) +\n",
    "# \t\tp9.scale_fill_brewer(type='qual', palette=2, direction=-1) +\n",
    "# \t\t# p9.scale_color_brewer(type='qual', palette=2, direction=-1) +\n",
    "# \t\tp9.labs(title=\"\", x=\"Read species\", y=\"Proportion of unmapped reads\", fill='Sample') +\n",
    "# \t\tp9.coord_flip() +\n",
    "# \t\tp9.guides(fill=p9.guide_legend(reverse=True), color=None) +\n",
    "# \t\tp9.theme()\n",
    "# )\n",
    "# print(plot)\n",
    "# plot.save(PLOTS_DIR + '/unmapped.png', dpi=500)\n",
    "\n",
    "OVERLAPS_COLUMNS = ('chrom', 'align_start', 'align_end', 'read_id', '_', 'align_orient', '_', '_', '_', '_', '_', '_', '_', 'f_start', 'f_end', 'f_name', '_', 'f_strand')\n",
    "def load_feature_overlaps(overlap_bed:str, include_overlap_info:bool=False):\n",
    "\t'''\n",
    "\tReturns a Series \n",
    "\tstart and end positions are 0-based inclusive\n",
    "\tIndices are alignment ids ( [read id];[reference name];[alignment start position];[alignment end position];[alignment orientation] )\n",
    "\tValues are tuples of overlap lists ( ([[transcript id], [feature number], [feature strand], [feature start position], [feature end position]]...) )\n",
    "\t\t\t\t\t\t\t\t\tor ( ([[transcript id], [feature number], [feature strand], [feature start position], [feature end position], [overlap start position], [overlap end position], [overlap length]]...) )\n",
    "\t'''\n",
    "\toverlaps = pd.read_csv(overlap_bed, sep='\\t', header=None)\n",
    "\toverlaps.columns = OVERLAPS_COLUMNS\n",
    "\toverlaps.align_end += -1\n",
    "\toverlaps.f_end += -1\n",
    "\n",
    "\toverlaps['tid'] = overlaps.f_name.transform(lambda name: name.split('_')[0])\n",
    "\toverlaps['f_num'] = overlaps.f_name.transform(lambda name: name.split('_')[2])\n",
    "\toverlaps['f_num'] = overlaps.f_num.transform(lambda num: int(num)+1)\t# exon/intron number is 0-based too, apparently\n",
    "\toverlaps['align_id'] = overlaps.apply(lambda row: f\"{row['read_id']};{row['chrom']};{row['align_start']};{row['align_end']};{row['align_orient']}\", axis=1)\n",
    "\n",
    "\tif include_overlap_info is True:\n",
    "\t\toverlaps['overlap_start'] = overlaps[['align_start', 'f_start']].max(axis=1)\n",
    "\t\toverlaps['overlap_end'] = overlaps[['align_end', 'f_end']].min(axis=1)\n",
    "\t\toverlaps['overlap_len'] = overlaps.overlap_end - overlaps.overlap_start + 1\n",
    "\t\toverlaps['combined'] = overlaps[['tid', 'f_num', 'f_strand', 'f_start', 'f_end', 'overlap_start', 'overlap_end', 'overlap_len']].values.tolist()\n",
    "\telse:\n",
    "\t\toverlaps['combined'] = overlaps[['tid', 'f_num', 'f_strand', 'f_start', 'f_end']].values.tolist()\n",
    "\n",
    "\toverlaps = overlaps.groupby('align_id')['combined'].agg(tuple)\n",
    "\treturn overlaps\n",
    "\n",
    "\t\n",
    "\n",
    "lariat_passed_rids = {}\n",
    "for sample, r, source, out_dir in sources:\n",
    "\trids = pd.read_csv(f'{out_dir}/lariat_reads.tsv', sep='\\t', usecols=[3])\n",
    "\tlariat_passed_rids[source] = set(rids.read_id)\n",
    "\n",
    "lariat_passed_rids\n",
    "failed_lariats = pd.DataFrame()\n",
    "for sample, r, source, out_dir in sources:\n",
    "\tdata = pd.read_csv(f'{out_dir}/failed_lariat_mappings.tsv', sep='\\t')\n",
    "\tdata = data.assign(samp=sample, r=r, source=source)\n",
    "\tfailed_lariats = pd.concat([failed_lariats, data], ignore_index=True)\n",
    "\n",
    "print(len(failed_lariats.read_id.unique()))\n",
    "print(len(failed_lariats[['read_id', 'source']].drop_duplicates()))\n",
    "failed_lariats\n",
    "trim_passed_rids = {}\n",
    "for sample, r, source, out_dir in sources:\n",
    "\trids = pd.read_csv(f'{out_dir}/final_info_table.tsv', sep='\\t', usecols=[0])\n",
    "\ttrim_passed_rids[source] = set(rids.read_id.str.slice(0,-4))\n",
    "trim_passed_rids\n",
    "trim_failed = pd.DataFrame()\n",
    "for sample, r, source, out_dir in sources:\n",
    "\tdata = pd.read_csv(f'{out_dir}/failed_trimmed_alignments.tsv', sep='\\t', nrows=6000)\n",
    "\tdata = data.assign(samp=sample, r=r, source=source)\n",
    "\ttrim_failed = pd.concat([trim_failed, data], ignore_index=True)\n",
    "\n",
    "print(trim_failed.filter_failed.value_counts())\n",
    "trim_failed['read_id_orient'] = trim_failed.read_id\n",
    "trim_failed.read_id = trim_failed.read_id.str.slice(0, -4)\n",
    "trim_failed\n",
    "fivep_passed_rids = {}\n",
    "for sample, r, source, out_dir in sources:\n",
    "\trids = pd.read_csv(f'{out_dir}/fivep_info_table.tsv', sep='\\t', usecols=[0])\n",
    "\tfivep_passed_rids[source] = set(rids.read_id.str.slice(0, -4))\n",
    "fivep_passed_rids\n",
    "run_data = pd.DataFrame()\n",
    "for sample, r, source, out_dir in sources:\n",
    "\tdata = pd.read_csv(f'{test_dir}/{sample}_{r}_lariat_mapping/run_data.tsv', sep='\\t', header=None)\n",
    "\ttotal_reads = data.loc[data[0]=='ref_mapped_reads', 1].iloc[0] + data.loc[data[0]=='ref_unmapped_reads', 1].iloc[0]\n",
    "\tdata.loc[len(data)] = ['total_reads', total_reads] \n",
    "\tdata = data.assign(samp=sample, r=r, source=source)\n",
    "\trun_data = pd.concat([run_data, data])\n",
    "\n",
    "run_data = run_data.rename(columns={0:'cat', 1:'reads'}).reset_index(drop=True)\n",
    "run_data.cat = run_data.cat.str.replace('_reads', '')\n",
    "run_data.samp = pd.Categorical(run_data.samp, categories=['HEK293T', 'C22'], ordered=True)\n",
    "samp_means = run_data.groupby(['samp', 'cat'])['reads'].aggregate('mean')\n",
    "for samp, cat in samp_means.index:\n",
    "\trun_data.loc[(run_data.samp==samp) & (run_data.cat==cat), 'samp_mean'] = samp_means[samp, cat]\n",
    "run_data\n",
    "reads_with_fivep = {}\n",
    "for sample, r, source, out_dir in sources:\n",
    "\trids = set()\n",
    "\twith open(f'{out_dir}/fivep_to_reads.sam') as r:\n",
    "\t\tfor line in r:\n",
    "\t\t\trids.add(line.split('\\t')[2])\n",
    "\treads_with_fivep[source] = rids\t\n",
    "reads_with_fivep\n",
    "BAD_ALIGNMENT_FILTERS = set(('mismatch_num', 'mismatch_percent', 'indel', 'map_quality'))\n",
    "reads_with_trim = {}\n",
    "for sample, r, source, out_dir in sources:\n",
    "\trids = set()\n",
    "\tfor align in pysam.AlignmentFile(f'{test_dir}/{sample}_{r}_lariat_mapping/trimmed_reads_to_genome.bam', 'rb'):\n",
    "\t\trids.add(align.query_name[:-4])\n",
    "\n",
    "\t# Get rid of those that mapped but got filtered out due to low quality alignment\n",
    "\tfor rid in rids:\n",
    "\t\trids_to_remove = set()\n",
    "\t\tif rid in trim_passed_rids:\n",
    "\t\t\tcontinue\n",
    "\n",
    "\t\tfail_reasons = set(trim_failed.loc[(trim_failed.source==source) & (trim_failed.read_id==rid), 'filter_failed'])\n",
    "\t\tif len(fail_reasons)==0:\n",
    "\t\t\tcontinue\n",
    "\t\tif fail_reasons.issubset(BAD_ALIGNMENT_FILTERS):\n",
    "\t\t\trids_to_remove.add(rid)\n",
    "\n",
    "\trids = rids.difference(rids_to_remove)\n",
    "\treads_with_trim[source] = rids\t\n",
    "\n",
    "reads_with_trim\n",
    "fates = []\n",
    "for sample, r, source, out_dir in sources:\n",
    "\trids = reads_with_fivep[source]\n",
    "\tfor rid in rids:\n",
    "\n",
    "\t\tif rid not in fivep_passed_rids[source]:\n",
    "\t\t\t# if rid in start_at_fivep.loc[start_at_fivep.source==source, 'read_id'].values:\n",
    "\t\t\t\t# fates.append((sample, r, source, rid, \"Starts at a 5'ss\"))\n",
    "\t\t\t\t# continue\n",
    "\t\t\tfates.append((sample, r, source, rid, \"Filtered after 5'ss mapping\"))\n",
    "\t\t\tcontinue\n",
    "\t\t\n",
    "\t\tif rid not in reads_with_trim[source]:\n",
    "\t\t\tfates.append((sample, r, source, rid, \"Unknown, has at least one 5'ss alignment\"))\n",
    "\t\t\tcontinue\n",
    "\n",
    "\t\tif rid not in trim_passed_rids[source]:\n",
    "\t\t\tfates.append((sample, r, source, rid, 'Template switching'))\n",
    "\t\t\tcontinue\n",
    "\t\t\n",
    "\t\tif rid not in lariat_passed_rids[source]:\n",
    "\t\t\tfails = set(failed_lariats.loc[(failed_lariats.source==source) & (failed_lariats.read_id==rid), 'filter_failed'])\n",
    "\t\t\tassert len(fails)==1, f'{source},{rid},{fails}'\n",
    "\t\t\t\n",
    "\t\t\tif len(fails)==1 and fails.pop()=='near_ss':\n",
    "\t\t\t\tfates.append((sample, r, source, rid, 'Circular intron'))\n",
    "\t\t\t\tcontinue\n",
    "\t\t\tif len(fails)==1 and fails.pop() in ('ubiquitin_gene', 'in_repeat'):\n",
    "\t\t\t\tfates.append((sample, r, source, rid, 'Repeat region'))\n",
    "\t\t\t\tcontinue\n",
    "\n",
    "\t\t\traise RuntimeError(rid)\n",
    "\t\t\n",
    "\t\tfates.append((sample, r, source, rid, 'Lariat'))\n",
    "\n",
    "fates = pd.DataFrame(fates, columns=['samp', 'r', 'source', 'read_id', 'fate'])\n",
    "assert fates[['source', 'read_id']].duplicated().sum()==0, fates[['source', 'read_id']].duplicated().sum()\n",
    "fates\n",
    "READ_SPECIES =  ('Lariat', 'Repeat region', 'Circular intron', 'Template switching', \"Unknown, has at least one 5'ss alignment\", \"Unknown, no 5'ss alignments\")\n",
    "species_counts = []\n",
    "for sample, r, source, out_dir in sources:\n",
    "\tunmapped = run_data.loc[(run_data.cat=='ref_unmapped') & (run_data.source==source), 'reads'].iloc[0]\n",
    "\tno_fivep_maps = unmapped - run_data.loc[(run_data.cat=='fivep_mapped') & (run_data.source==source), 'reads'].iloc[0]\n",
    "\n",
    "\tfor species in READ_SPECIES:\n",
    "\t\tif species == \"Unknown, no 5'ss alignments\":\n",
    "\t\t\tspecies_counts.append([sample, r, source, unmapped, species, no_fivep_maps])\n",
    "\t\t\tcontinue\n",
    "\t\t\t\n",
    "\t\tcount = len(fates.loc[(fates.source==source) & (fates.fate==species)])\n",
    "\t\tspecies_counts.append([sample, r, source, unmapped, species, count])\n",
    "\n",
    "species_counts = pd.DataFrame(species_counts, columns=['samp', 'r', 'source',  'unmapped', 'species', 'reads'])\n",
    "species_counts['species'] = pd.Categorical(species_counts.species, READ_SPECIES, ordered=True)\n",
    "species_counts['prop'] = species_counts.reads / species_counts.unmapped\n",
    "species_counts\n",
    "species_counts_avg = species_counts.groupby(['samp', 'species'])[['reads', 'prop']].agg('mean').reset_index()\n",
    "species_counts_avg['label'] = species_counts_avg.reads.transform(lambda val: f'{val:,.0f}')\n",
    "species_counts_avg\n",
    "plot = (p9.ggplot(species_counts_avg, p9.aes(x=\"species\", fill='samp')) +\n",
    "\t\tp9.geom_col(p9.aes(y=\"prop\"), position='dodge') +\n",
    "\t\t# p9.geom_text(p9.aes(y=\"prop+0.05\", label=\"label\", color='samp'), ha='left', size=8, position=p9.position_dodge(width=1)) +\n",
    "\t\t# p9.annotate('text', x=6.5, y=1.05, label='Avg. reads', fontweight='bold') +\n",
    "\t\tp9.scale_y_continuous(limits=(0,1), breaks=np.arange(0,1.2,0.2)) +\n",
    "\t\tp9.scale_fill_brewer(type='qual', palette=2, direction=-1) +\n",
    "\t\t# p9.scale_color_brewer(type='qual', palette=2, direction=-1) +\n",
    "\t\tp9.labs(title=\"\", x=\"Read species\", y=\"Proportion of unmapped reads\", fill='Sample') +\n",
    "\t\tp9.coord_flip() +\n",
    "\t\tp9.guides(fill=p9.guide_legend(reverse=True), color=None) +\n",
    "\t\tp9.theme()\n",
    ")\n",
    "print(plot)\n",
    "plot.save(PLOTS_DIR + '/unmapped.png', dpi=500)\n",
    "\n",
    "\n",
    "\n",
    "# mapped_reads = functions.load_bam( f'{out_dir}/mapped_reads.bam')\n",
    "# mapped_reads = mapped_reads.rename(columns={'q_id': 'read_id', 'r_id': 'chrom', 'orient': 'align_orient', 'r_start': 'align_start', 'r_end': 'align_end'})\n",
    "# mapped_reads\n",
    "# mapped_reads = functions.load_bam( f'{out_dir}/mapped_reads.bam')\n",
    "# mapped_reads = mapped_reads.rename(columns={'q_id': 'read_id', 'r_id': 'chrom', 'orient': 'align_orient', 'r_start': 'align_start', 'r_end': 'align_end'})\n",
    "# mapped_reads\n",
    "# mapped_reads = functions.load_bam( f'{out_dir}/mapped_reads.bam')\n",
    "# mapped_reads = mapped_reads.rename(columns={'q_id': 'read_id', 'r_id': 'chrom', 'orient': 'align_orient', 'r_start': 'align_start', 'r_end': 'align_end'})\n",
    "# mapped_reads\n",
    "# mapped_reads = functions.load_bam( f'{out_dir}/mapped_reads.bam')\n",
    "# mapped_reads = mapped_reads.rename(columns={'q_id': 'read_id', 'r_id': 'chrom', 'orient': 'align_orient', 'r_start': 'align_start', 'r_end': 'align_end'})\n",
    "# mapped_reads\n",
    "# mapped_reads = mapped_reads[mapped_reads.mapped].reset_index(drop=True).drop(columns=['mapped', 'q_start', 'q_end', 'quality', 'seq'])\n",
    "# mapped_reads['align_id'] = mapped_reads.apply(lambda row: f\"{row['read_id']};{row['chrom']};{row['align_start']};{row['align_end']};{row['align_orient']}\", axis=1)\n",
    "# assert mapped_reads.align_id.is_unique\n",
    "# mapped_reads\n",
    "# intron_overlaps = load_feature_overlaps(f'{out_dir}/intron_overlaps.bed')\n",
    "# mapped_reads['introns'] = mapped_reads.align_id.map(intron_overlaps.to_dict())\n",
    "# print(mapped_reads.introns.notna().sum())\n",
    "# exon_overlaps = load_feature_overlaps(f'{out_dir}/exon_overlaps.bed')\n",
    "# mapped_reads['exons'] = mapped_reads.align_id.map(exon_overlaps.to_dict())\n",
    "# print(mapped_reads.exons.notna().sum())\n",
    "# tx2gene = pd.read_csv('/Users/trumanmooney/Library/CloudStorage/OneDrive-BrownUniversity/Documents/RNA-seq data mining/reference_data/annotations/tx2gene.gencode.v44.csv')\n",
    "# tx2gene = tx2gene.set_index('TXNAME').GENEID.to_dict()\n",
    "# tx2gene\n",
    "# def append_gene_id(features):\n",
    "# \tif pd.isna(features):\n",
    "# \t\treturn pd.NA\n",
    "\t\n",
    "# \tappended_features = []\n",
    "# \tfor tid, f_num, f_strand, f_start, f_end in features:\n",
    "# \t\tgene_id = tx2gene[tid] \n",
    "# \t\tappended_features.append([tid, f_num, f_strand, f_start, f_end, gene_id])\n",
    "\n",
    "# \treturn tuple(appended_features) \n",
    "\n",
    "# mapped_reads.introns = mapped_reads.introns.transform(append_gene_id)\n",
    "# mapped_reads.exons = mapped_reads.exons.transform(append_gene_id)\n",
    "# print(mapped_reads.at[1,'introns'])\n",
    "# print(mapped_reads.at[1,'exons'])\n",
    "# def pair_features(row):\n",
    "# \tfeatures = []\n",
    "# \tif pd.notna(row['exons']):\n",
    "# \t\tfeatures += [['exon', *exon] for exon in row['exons']]\n",
    "# \tif pd.notna(row['introns']):\n",
    "# \t\tfeatures += [['intron', *intron] for intron in row['introns']]\n",
    "# \tif features == []:\n",
    "# \t\treturn (), ()\n",
    "\t\n",
    "# \tsame_tid = []\n",
    "# \tdiff_tid_same_gid = []\n",
    "# \tfor feature_1, feature_2 in it.permutations(features, 2):\n",
    "# \t\tif feature_1[1] == feature_2[1]:\n",
    "# \t\t\tsame_tid.append((feature_1, feature_2))\n",
    "# \t\t\tcontinue\n",
    "# \t\tif feature_1[-1] == feature_2[-1]:\n",
    "# \t\t\tdiff_tid_same_gid.append((feature_1, feature_2))\n",
    "\t\t\n",
    "# \treturn tuple(same_tid), tuple(diff_tid_same_gid)\n",
    "\n",
    "# mapped_reads[['same_tid', 'same_gid']] = mapped_reads.apply(pair_features, axis=1, result_type='expand')\n",
    "# mapped_reads\n",
    "# def cat_pairs(same_id):\n",
    "# \t# cats = {'pre-mRNA': [], 'mRNA': [], 'in_then_in': [], 'ex_ex_far': [], 'in_in_far': [], 'in_ex_far': []}\n",
    "# \tcats = []\n",
    "# \tfor feature_1, feature_2 in same_id:\n",
    "# \t\tnum_within_1 = abs(feature_1[2] - feature_2[2]) <= 1\n",
    "\n",
    "# \t\tif feature_1[0] == 'exon' and feature_2[0] == 'exon':\n",
    "# \t\t\tif num_within_1 is True:\n",
    "# \t\t\t\t# cats['mRNA'].append(feature_1[1])\n",
    "# \t\t\t\tcats.append('mRNA')\n",
    "# \t\t\telse:\n",
    "# \t\t\t\t# cats['ex_ex_far'].append(feature_1[1])\n",
    "# \t\t\t\tcats.append('ex_ex_far')\n",
    "\n",
    "# \t\tif feature_1[0] == 'intron' and feature_2[0] == 'intron':\n",
    "# \t\t\tif num_within_1 is True:\n",
    "# \t\t\t\t# cats['in_then_in'].append(feature_1[1])\n",
    "# \t\t\t\tcats.append('in_then_in')\n",
    "# \t\t\telse:\n",
    "# \t\t\t\t# cats['in_in_far'].append(feature_1[1])\n",
    "# \t\t\t\tcats.append('in_in_far')\n",
    "\t\t\n",
    "# \t\tif num_within_1 is True:\n",
    "# \t\t\t# cats['pre-mRNA'].append(feature_1[1])\n",
    "# \t\t\tcats.append('pre-mRNA')\n",
    "# \t\telse:\n",
    "# \t\t\t# cats['in_ex_far'].append(feature_1[1])\n",
    "# \t\t\tcats.append('in_ex_far')\n",
    "\n",
    "# \t# return sorted(cats)\n",
    "# \treturn set(cats)\n",
    "\n",
    "\t\t\n",
    "\n",
    "# mapped_reads['same_tid_cat'] = mapped_reads.same_tid.transform(cat_pairs)\n",
    "# mapped_reads['same_gid_cat'] = mapped_reads.same_gid.transform(cat_pairs)\n",
    "# print(mapped_reads.same_tid_cat[63])\n",
    "# print(mapped_reads.same_gid_cat[1])\n",
    "# mapped_reads\n",
    "# print(mapped_reads.loc[mapped_reads.same_tid_cat.transform(len)>0, 'same_tid_cat'].value_counts())\n",
    "# print(mapped_reads.loc[mapped_reads.same_gid_cat.transform(len)>0, 'same_gid_cat'].value_counts())\n",
    "# mapped_reads['type'] = mapped_reads.apply(categorize_mapped_species, axis=1)\n",
    "# print(mapped_reads.type.value_counts())\n",
    "# mapped_reads\n",
    "# out = [(source, sample, r, type_, count) for type_, count in mapped_reads.type.value_counts().to_dict().items()]\n",
    "# print(out)\n",
    "# mapped_reads['gaps'] = mapped_reads.cigar.transform(gaps_from_cigartuples)\n",
    "# print(mapped_reads.gaps.value_counts())\n",
    "# mapped_reads['skips'] = mapped_reads.cigar.transform(skips_from_cigartuples)\n",
    "# print(mapped_reads.skips.value_counts())\n",
    "# mapped_avg = [('C22|R2', 'C22', 'R2', 'Exonic', 15510), ('C22|R2', 'C22', 'R2', 'Genic, ambiguous', 15289), ('C22|R2', 'C22', 'R2', 'Intronic', 12314), ('C22|R2', 'C22', 'R2', 'Inter-genic', 3197), ('C22|R2', 'C22', 'R2', 'pre-mRNA', 1359), ('C22|R2', 'C22', 'R2', \"Ends at a 5'ss\", 9)] + \\\n",
    "# \t[('C22|R1', 'C22', 'R1', 'Exonic', 15507), ('C22|R1', 'C22', 'R1', 'Genic, ambiguous', 15483), ('C22|R1', 'C22', 'R1', 'Intronic', 12332), ('C22|R1', 'C22', 'R1', 'Inter-genic', 3204), ('C22|R1', 'C22', 'R1', 'pre-mRNA', 1384), ('C22|R1', 'C22', 'R1', \"Ends at a 5'ss\", 4), ('C22|R1', 'C22', 'R1', 'mRNA', 1)] + \\\n",
    "# \t[('HEK293T|R1', 'HEK293T', 'R1', 'Exonic', 17228), ('HEK293T|R1', 'HEK293T', 'R1', 'Genic, ambiguous', 14489), ('HEK293T|R1', 'HEK293T', 'R1', 'Intronic', 9101), ('HEK293T|R1', 'HEK293T', 'R1', 'Inter-genic', 3485), ('HEK293T|R1', 'HEK293T', 'R1', 'pre-mRNA', 1339), ('HEK293T|R1', 'HEK293T', 'R1', \"Ends at a 5'ss\", 4)] + \\\n",
    "# \t[('HEK293T|R2', 'HEK293T', 'R2', 'Exonic', 18974), ('HEK293T|R2', 'HEK293T', 'R2', 'Genic, ambiguous', 11244), ('HEK293T|R2', 'HEK293T', 'R2', 'Intronic', 8803), ('HEK293T|R2', 'HEK293T', 'R2', 'Inter-genic', 3461), ('HEK293T|R2', 'HEK293T', 'R2', 'pre-mRNA', 1253), ('HEK293T|R2', 'HEK293T', 'R2', \"Ends at a 5'ss\", 6)]\n",
    "\n",
    "# mapped_avg = pd.DataFrame(mapped_avg, columns = ['source', 'samp', 'r', 'type', 'reads'])\n",
    "# mapped_avg\n",
    "# mapped_avg.type = pd.Categorical(mapped_avg.type, MAPPED_SPECIES, ordered=True)\n",
    "# mapped_avg['mapped'] = mapped_avg.source.map(mapped_avg.groupby('source').reads.agg('sum'))\n",
    "# mapped_avg['prop'] = mapped_avg.reads / mapped_avg.mapped\n",
    "# mapped_avg\n",
    "# mapped_avg = mapped_avg.groupby(['samp', 'type'])[['reads', 'prop']].agg('mean').reset_index().fillna(0)\n",
    "# mapped_avg['label'] = mapped_avg.reads.transform(lambda val: f'{val:,.0f}')\n",
    "# mapped_avg\n",
    "# plot = (p9.ggplot(mapped_avg, p9.aes(x=\"type\", fill='samp')) +\n",
    "# \t\tp9.geom_col(p9.aes(y=\"prop\"), position='dodge') +\n",
    "# \t\tp9.geom_text(p9.aes(y=\"prop+0.05\", label=\"label\", color='samp'), ha='left', size=8, position=p9.position_dodge(width=1)) +\n",
    "# \t\t# p9.annotate('text', x=6.5, y=1.05, label='Avg. reads', fontweight='bold') +\n",
    "# \t\tp9.scale_y_continuous(limits=(0,1), breaks=np.arange(0,1.2,0.2)) +\n",
    "# \t\tp9.scale_fill_brewer(type='qual', palette=2, direction=-1) +\n",
    "# \t\t# p9.scale_color_brewer(type='qual', palette=2, direction=-1) +\n",
    "# \t\tp9.labs(title=\"\", x=\"Read species\", y=\"Proportion of mapped reads\", fill='Sample') +\n",
    "# \t\tp9.coord_flip() +\n",
    "# \t\tp9.guides(fill=p9.guide_legend(reverse=True), color=None) +\n",
    "# \t\tp9.theme()\n",
    "# )\n",
    "# print(plot)\n",
    "# plot.save(PLOTS_DIR + '/mapped.png', dpi=500)\n",
    "\n",
    "MAPPED_SPECIES = (\n",
    "\t\t\t\t  'pre-mRNA', \t\t\t\t# At least 1 exon overlap AND at least 1 intron overlap AND at least 1 shared transcript ID between an exon and an intron\n",
    "\t\t\t\t  'mRNA', \t\t\t\t\t# At least 2 exon overlaps AND at least 1 shared transcript ID between two exons AND not \"pre-mRNA\"\n",
    "\t\t\t\t  \"Ends at a 5'ss\", \t\t# No exon overlaps AND read starts at the same position as a (+)-strand intron's start OR reads ends at the same position as a (-)-strand intron's end\n",
    "\t\t\t\t  'Intronic', \t\t\t\t# No exon overlaps AND not \"Ends at a 5'ss\"\n",
    "\t\t\t\t  'Exonic', \t\t\t\t# At least 1 exon overlap AND no intron overlaps AND not \"mRNA\"\n",
    "\t\t\t\t  'Genic, ambiguous', \t\t# At least 1 exon overlap AND at least 1 intron overlap AND not \"pre-mRNA\"\n",
    "\t\t\t\t  'Inter-genic',\t\t\t# No exon overlaps AND no intron overlaps\n",
    "\t\t\t\t  )\n",
    "\n",
    "# def categorize_mapped_species(row: pd.Series):\n",
    "# \tif pd.isna(row['exons']) and pd.isna(row['introns']):\n",
    "# \t\treturn 'Inter-genic'\n",
    "\t\n",
    "# \tif pd.isna(row['exons']) and pd.notna(row['introns']):\n",
    "# \t\tfor tid, intron_num, intron_strand, intron_start, intron_end, gene_id in row['introns']:\n",
    "# \t\t\tif intron_strand == '+' and row['align_start'] == intron_start:\n",
    "# \t\t\t\treturn \"Ends at a 5'ss\"\n",
    "# \t\t\tif intron_strand == '-' and row['align_end'] == intron_end:\n",
    "# \t\t\t\treturn \"Ends at a 5'ss\"\n",
    "\t\n",
    "# \t\treturn 'Intronic'\n",
    "\t\n",
    "# \t# At this point we know we have at least 1 exon\n",
    "# \tif pd.notna(row['exons']) and pd.isna(row['introns']):\n",
    "\t\t\n",
    "\t\n",
    "# \texon_tids = set([exon[0] for exon in row['exons']])\n",
    "# \tintron_tids = set([intron[0] for intron in row['introns']]) if pd.notna(row['introns']) else set()\n",
    "# \tfor exon, intron in it.product(row['exons'], row['introns']):\n",
    "\t\t\n",
    "# \t# At least 1 exon and intron share a tid\n",
    "# \tif len(exon_tids.intersection(intron_tids)) > 0:\n",
    "# \t\treturn 'pre-mRNA'\n",
    "# \t# At least 2 exons share a tid\n",
    "# \tif len(row['exons']) > len(exon_tids):\n",
    "# \t\treturn 'mRNA'\n",
    "# \t# Only exons that don't share a tid\n",
    "# \tif pd.isna(row['introns']):\n",
    "# \t\treturn 'Exonic'\n",
    "\t\n",
    "# \t# At this point we know we have at least 1 exon\n",
    "# \treturn 'Genic, ambiguous'\n",
    "\n",
    "\n",
    "# fivep_failed = pd.DataFrame()\n",
    "# for sample, r, source, out_dir in sources:\n",
    "# \tdata = pd.read_csv(f'{out_dir}/failed_fivep_alignments.tsv', sep='\\t')\n",
    "# \tdata = data.assign(samp=sample, r=r, source=source)\n",
    "# \tfivep_failed = pd.concat([fivep_failed, data], ignore_index=True)\n",
    "\n",
    "# print(fivep_failed.fail_reason.value_counts())\n",
    "# fivep_failed\n",
    "# fivep_failed_rids = {}\n",
    "# for sample, r, source, out_dir in sources:\n",
    "# \trids = pd.read_csv(f'{out_dir}/failed_fivep_alignments.tsv', sep='\\t', usecols=[0])\n",
    "# \tfivep_failed_rids[source] = set(rids.read_id)\n",
    "# fivep_failed_rids\n",
    "\n",
    "\n",
    "\n",
    "# df = fate_counts.copy()\n",
    "# df.loc[df.fate.isin([\"Filtered after 5'ss mapping\", \"Starts at a 5'ss\", 'No trimmed alignments']), 'fate'] = \"Unknown, has a 5'ss alignment\"\n",
    "# df = df.groupby(['source', 'samp', 'r', 'fate', 'unmapped'], as_index=False, observed=True)['reads'].agg(sum)\n",
    "\n",
    "# zerorows = []\n",
    "# for i, row in df[['source', 'samp', 'r', 'unmapped']].drop_duplicates().iterrows():\n",
    "# \tfor fate in READ_SPECIES:\n",
    "# \t\tif len(df.loc[(df.source==row['source']) & (df.fate==fate)]) == 0:\n",
    "# \t\t\tzerorows.append([row['source'], row['samp'], row['r'], fate, row['unmapped'], 0])\n",
    "# df = pd.concat([df, pd.DataFrame(zerorows, columns=df.columns)])\n",
    "\n",
    "# df.fate = pd.Categorical(df.fate, categories=READ_SPECIES, ordered=True) \n",
    "# df['prop'] = df.reads / df.unmapped\n",
    "# df = df.sort_values(['source', 'fate'])\n",
    "# df\n",
    "\n",
    "# fate_counts = fates[['source', 'fate']].value_counts().sort_index()\n",
    "# fate_counts = fate_counts.reset_index().set_axis(['source', 'fate', 'reads'], axis=1)\n",
    "# fate_counts[['samp', 'r']] = fate_counts.source.str.split('|', expand=True)\n",
    "# for sample, r, source, out_dir in sources:\n",
    "# \tunmapped = run_data.loc[(run_data.cat=='ref_unmapped') & (run_data.source==source), 'reads'].iloc[0]\n",
    "# \tno_fivep_maps = unmapped - run_data.loc[(run_data.cat=='fivep_mapped') & (run_data.source==source), 'reads'].iloc[0]\n",
    "\t\n",
    "# \tfate_counts.loc[fate_counts.source==source, 'unmapped'] = unmapped\n",
    "# \tfate_counts.loc[len(fate_counts)] = [source, \"Unknown, no 5'ss alignments\", no_fivep_maps, sample, r, unmapped]\n",
    "# \tcovered = len(fates.loc[fates.source==source]) + no_fivep_maps\n",
    "\t\n",
    "# \tprint(source, covered, 'of', unmapped)\n",
    "# fate_counts['prop'] = fate_counts.reads / fate_counts.unmapped\n",
    "# fate_counts = fate_counts.sort_values(['source', 'fate'])\n",
    "# fate_counts\n",
    "\n",
    "\n",
    "start_at_fivep = fivep_failed.loc[(~fivep_failed.read_is_reverse) & (fivep_failed.read_fivep_start==0)]\n",
    "start_at_fivep\n",
    "\n",
    "plot = (p9.ggplot(read_outcomes, p9.aes(x='category')) +\n",
    "\t\tp9.geom_boxplot(p9.aes(y='rpm')) +\n",
    "\t\tp9.geom_text(p9.aes(y='mean_rpm', label='label'), size=10, nudge_y=2000, family='arial') +\n",
    "\t\tp9.scale_y_continuous(labels=comma_format()) +\n",
    "\t\tp9.labs(x='Read classification',\n",
    "\t\t  \t\ty=\"Reads per million mapped reads\") +\n",
    "\t\tp9.facet_wrap('samp')\n",
    "\t)\n",
    "print(plot)\n",
    "plot.save('/Users/trumanmooney/Library/CloudStorage/OneDrive-BrownUniversity/Documents/Lariat_mapping/visuals/plots/read_classes.png', dpi=500, height=6, width=10)\n",
    "\n",
    "read_means = read_outcomes.groupby(['samp', 'category'])['rpm'].aggregate('mean')\n",
    "for samp, cat in read_means.index:\n",
    "\tread_outcomes.loc[(read_outcomes.samp==samp) & (read_outcomes.category==cat), 'mean_rpm'] = read_means[samp, cat]\n",
    "\n",
    "read_outcomes['label'] = read_outcomes.mean_rpm.transform(lambda x: comma_format()([int(round(x))])[0])\n",
    "read_outcomes\n",
    "for sample, r, source, out_dir in sources:\n",
    "\tx = read_outcomes.loc[read_outcomes.source=='C22|R1', 'mapped_to_genome'].iloc[0] + read_outcomes.loc[read_outcomes.source=='C22|R1', 'reads'].sum()\n",
    "\tprint(source, x)\n",
    "rows = []\n",
    "for sample, r, source, out_dir in sources:\n",
    "\t\n",
    "\tog_mapped_to_genome = run_data.loc[(run_data.cat=='ref_mapped') & (run_data.source==source), 'reads'].iloc[0]\n",
    "\tpremRNA = run_data.loc[(run_data.cat=='premRNA') & (run_data.source==source), 'reads'].iloc[0]\n",
    "\tunmapped = run_data.loc[(run_data.cat=='ref_unmapped') & (run_data.source==source), 'reads'].iloc[0]\n",
    "\tno_fivep_maps = unmapped - run_data.loc[(run_data.cat=='fivep_mapped') & (run_data.source==source), 'reads'].iloc[0]\n",
    "\t# fivep_filtered_out = \n",
    "\tno_trim_align = run_data.loc[(run_data.cat=='fivep_filtered') & (run_data.source==source), 'reads'].iloc[0] - run_data.loc[(run_data.cat=='trimmed_mapped') & (run_data.source==source), 'reads'].iloc[0]\n",
    "\t\n",
    "\tstart_at_fivep_rids = set(start_at_fivep.loc[start_at_fivep.source==source, 'read_id'])\n",
    "\tlariats_rids = set(lariats.loc[lariats.source==source, 'read_id'])\n",
    "\t# trimmed_mapped_nonlariat_rids = fivep_passed.loc[(fivep_passed.source==source) & (fivep_passed.trimmed_mapped) & (~fivep_passed.passed_trimmed_filters)]\n",
    "\t# trimmed_mapped_nonlariat_rids = set(trimmed_mapped_nonlariat_rids['read_id'])\n",
    "\trepetitive_rids = set(failed_lariats.loc[(failed_lariats.filter_failed.isin(['ubiquitin_gene', 'in_repeat'])) & (~failed_lariats.read_id.isin(lariats.read_id.values)), 'read_id'])\n",
    "\tcircular_rids = set(failed_lariats.loc[(failed_lariats.filter_failed=='near_ss') & (~failed_lariats.read_id.isin(lariats.read_id)), 'read_id'])\n",
    "\n",
    "\tprint('start at fivep and lariat', len(start_at_fivep.intersection(lariats_rids)), start_at_fivep_rids.intersection(lariats_rids))\n",
    "\t# print('end at fivep and trimmed mapped nonlariat', len(end_at_fivep_rids.intersection(trimmed_mapped_nonlariat_rids)), end_at_fivep_rids.intersection(trimmed_mapped_nonlariat_rids))\n",
    "\t# print('lariat and trimmed mapped nonlariat', len(lariats_rids.intersection(trimmed_mapped_nonlariat_rids)), lariats_rids.intersection(trimmed_mapped_nonlariat_rids))\n",
    "\n",
    "\trow = [sample, r, source, og_mapped_to_genome, premRNA, unmapped, no_fivep_maps, len(start_at_fivep), no_trim_align, len(repetitive_rids), len(circular_rids), len(lariats_rids)]\n",
    "\trows.append(row)\n",
    "\n",
    "read_outcomes = pd.DataFrame(rows, columns = ['samp', 'r', 'source', 'mapped_to_genome', 'pre-mRNA', 'unmapped', 'no fivep', \"Ends at a 5'ss\", 'no trim', 'repetitive', 'circular', 'Lariat'])\n",
    "read_outcomes = read_outcomes.melt(id_vars=['samp', 'r', 'source', 'mapped_to_genome', 'unmapped'], var_name='category', value_name='reads')\n",
    "read_outcomes['rpm'] = (read_outcomes.reads * 1_000_000)/ read_outcomes.mapped_to_genome\n",
    "read_outcomes.category = pd.Categorical(read_outcomes.category, categories=read_outcomes.category.unique(), ordered=True)\n",
    "read_outcomes.samp = pd.Categorical(read_outcomes.samp, categories=['HEK293T', 'C22'], ordered=True)\n",
    "read_outcomes = read_outcomes.sort_values('source')\n",
    "read_outcomes\n",
    "\n",
    "\n",
    "\n",
    "df.fate = df.fate.map({'filt fivep': 'Unknown', 'no trim': 'Unknown', 'start fivep': \"Starts at a 5'ss\", 'circular': 'Circular intron', 'repeat': 'Repetitive region', 'lariat':'Lariat', 'temp switch': 'Template switching', 'no fivep': 'Unknown'})\n",
    "\n",
    "\n",
    "for sample, r, source, out_dir in sources:\n",
    "\ttrimmed_mapped_rids = set()\n",
    "\tfor align in pysam.AlignmentFile(f'{out_dir}/trimmed_reads_to_genome.bam', 'rb'):\t\n",
    "\t\ttrimmed_mapped_rids.add(align.query_name)\n",
    "\t\t\n",
    "\tfivep_passed.loc[fivep_passed.source==source, 'trimmed_mapped'] = fivep_passed.loc[fivep_passed.source==source, 'read_id'].isin(trimmed_mapped_rids)\n",
    "\tfivep_passed.loc[fivep_passed.source==source, 'passed_trimmed_filters'] = fivep_passed.loc[fivep_passed.source==source, 'read_id'].isin(threep_passed_rids[source]) \n",
    "\n",
    "\treads_total = len(fivep_passed.loc[(fivep_passed.source==source), 'read_id'].unique())\n",
    "\treads_mapped = len(fivep_passed.loc[(fivep_passed.source==source) & (fivep_passed.passed_trimmed_filters), 'read_id'].unique())\n",
    "\tprint(f'{reads_mapped:,} of {reads_total:,}')\n",
    "\n",
    "fivep_passed\n",
    "\n",
    "end_at_fivep_counts = {}\n",
    "for sample, r, source, out_dir in sources:\n",
    "\tunique_reads = end_at_fivep.loc[end_at_fivep.source==source, 'read_id'].unique()\n",
    "\tend_at_fivep_counts[source] = len(unique_reads)\n",
    "end_at_fivep_counts\n",
    "\n",
    "\n",
    "\n",
    "funnel = copy.deepcopy(run_data)\n",
    "\n",
    "for sample, r, source, out_dir in sources:\n",
    "\tunmapped = funnel.loc[(funnel.source==source) & (funnel.cat=='ref_unmapped'), 'reads'].iloc[0]\n",
    "\tfunnel.loc[funnel.source==source, 'unmapped'] = unmapped\n",
    "\n",
    "funnel['unmap_p'] = funnel.reads / funnel.unmapped\n",
    "funnel\n",
    "funnel.cat = funnel.cat.map({'fivep_mapped': \"At least one 5'ss mapped to the read\", 'trimmed_mapped':\"Trimmed section mapped to the genome\", 'filtered_lariats':'Lariats'})\n",
    "funnel = funnel.dropna()\n",
    "funnel.cat = pd.Categorical(funnel.cat, categories=['Lariats', \"Trimmed section mapped to the genome\", \"Mapped to a 5'ss\"], ordered=True)\n",
    "funnel\n",
    "funnel.loc[funnel.samp=='C22', 'unmap_p']\n",
    "samp_means = funnel.groupby(['samp', 'cat'])['unmap_p'].aggregate('mean')\n",
    "for samp, cat in samp_means.index:\n",
    "\tfunnel.loc[(funnel.samp==samp) & (funnel.cat==cat), 'unmap_p_mean'] = samp_means[samp, cat]\n",
    "\tfunnel.loc[(funnel.samp==samp) & (funnel.cat==cat), 'unmap_p_sem'] = funnel.loc[(funnel.samp==samp) & (funnel.cat==cat), 'unmap_p'].sem()\n",
    "\t\n",
    "funnel['ebar_top'] = funnel.unmap_p_mean + funnel.unmap_p_sem\n",
    "funnel['ebar_bottom'] = funnel.unmap_p_mean - funnel.unmap_p_sem\n",
    "funnel\n",
    "samp_means = funnel.groupby(['samp', 'cat'])['reads'].aggregate('mean')\n",
    "for samp, cat in samp_means.index:\n",
    "\tfunnel.loc[(funnel.samp==samp) & (funnel.cat==cat), 'reads_mean'] = samp_means[samp, cat]\n",
    "\n",
    "funnel.reads_mean = funnel.reads_mean.astype(int)\n",
    "funnel['label_y'] = funnel.unmap_p_mean + 0.01\n",
    "funnel\n",
    "plot = (p9.ggplot(funnel, p9.aes(x='cat', fill='samp', group='samp')) +\n",
    "\t\tp9.geom_col(p9.aes(y='unmap_p_mean'), position='dodge') +\n",
    "\t\tp9.geom_errorbar(p9.aes(ymax='ebar_top', ymin='ebar_bottom'), position='dodge', width=0.9) +\n",
    "\t\t# p9.geom_boxplot(p9.aes(y='unmap_p')) +\n",
    "\t\tp9.geom_text(p9.aes(y='label_y', label='reads_mean'), position=p9.position_dodge(width=0.9), size=10, family='arial') +\n",
    "\t\tp9.scale_y_continuous(labels=percent_format()) +\n",
    "\t\tp9.scale_fill_brewer(type='qual', palette=2) +\n",
    "\t\tp9.labs(x='Lariat-mapping step', \n",
    "\t\t  \t\ty='% of unmapped reads',\n",
    "\t\t\t\tfill='Cell line') +\n",
    "\t\tp9.coord_flip() +\n",
    "\t\tp9.theme()\n",
    "\t\t# p9.facet_wrap('samp')\n",
    "\t)\n",
    "plot\n",
    "plot.save('/Users/trumanmooney/Library/CloudStorage/OneDrive-BrownUniversity/Documents/Lariat_mapping/visuals/plots/funnel.png', dpi=500)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.6 ('base': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "15a590390ab2d07a8b29d5c0d429762a50d82d00365b164ce6f1192d25159d8e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
