{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "\n",
    "import pysam\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from intervaltree import Interval, IntervalTree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "LINEAR_COLUMNS = ('read_id', \n",
    "\t\t\t\t  'chrom', \n",
    "\t\t\t\t#   'start', \n",
    "\t\t\t\t#   'end', \n",
    "\t\t\t\t  'blocks', \n",
    "\t\t\t\t  'cigar',\n",
    "\t\t\t\t  'read',\n",
    "\t\t\t\t  )\n",
    "CIGARTUPLE_CODES = {0: 'M',\n",
    "\t\t\t\t\t1: 'I',\n",
    "\t\t\t\t\t2: 'D',\n",
    "\t\t\t\t\t3: 'N',\n",
    "\t\t\t\t\t4: 'S',\n",
    "\t\t\t\t\t5: 'H',\n",
    "\t\t\t\t\t6: 'P',\n",
    "\t\t\t\t\t7: '=',\n",
    "\t\t\t\t\t8: 'X',\n",
    "\t\t\t\t\t9: 'B',\n",
    "\t\t\t\t\t}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def infer_segments(row):\n",
    "\tif len(row['blocks']) == 1:\n",
    "\t\treturn row['blocks']\n",
    "\t\n",
    "\tsegs = []\n",
    "\tcurrent_block = row['blocks'][0]\n",
    "\tfor blocks_index in range(1, len(row['blocks'])):\n",
    "\t\tnext_block = row['blocks'][blocks_index]\n",
    "\t\tcigar_index = 2*blocks_index - 1\n",
    "\t\tgap_type = row['cigar'][cigar_index][0]\n",
    "\n",
    "\t\tif gap_type == 'N':\n",
    "\t\t\tsegs.append(current_block)\n",
    "\t\t\tcurrent_block = next_block\n",
    "\t\telif gap_type in ('I', 'D'):\n",
    "\t\t\tcurrent_block = (current_block[0], next_block[1])\n",
    "\t\telse:\n",
    "\t\t\traise RuntimeError(f\"{row['blocks']}\\n{row['cigar']}\\n{blocks_index}\")\n",
    "\t\n",
    "\tsegs.append(current_block)\n",
    "\t\t\n",
    "\treturn tuple(segs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_exons = '/Users/trumanmooney/Library/CloudStorage/OneDrive-BrownUniversity/Documents/Projects/Lariat_mapping/reference_data/hg38_ref/exons.tsv.gz'\n",
    "ref_introns = '/Users/trumanmooney/Library/CloudStorage/OneDrive-BrownUniversity/Documents/Projects/Lariat_mapping/reference_data/hg38_ref/introns.tsv.gz'\n",
    "single_end = False\n",
    "output_base = '/Users/trumanmooney/Library/CloudStorage/OneDrive-BrownUniversity/Documents/Projects/Lariat_mapping/output/pipeline/C22-1_100k/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "introns = pd.read_csv(ref_introns, sep='\\t')\n",
    "introns['gene_id'] = introns.gene_id.transform(lambda gid: set(gid.split(',')))\n",
    "introns['interval'] = introns.apply(lambda row: Interval(row['start'], row['end'], {'gene_id': row['gene_id']}), axis=1)\n",
    "introns = introns.groupby('chrom').interval.agg(IntervalTree).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "exons = pd.read_csv(ref_exons, sep='\\t')\n",
    "exons['gene_id'] = exons.gene_id.transform(lambda gid: set(gid.split(',')))\n",
    "exons['interval'] = exons.apply(lambda row: Interval(row['start'], row['end'], {'gene_id': row['gene_id']}), axis=1)\n",
    "exons = exons.groupby('chrom').interval.agg(IntervalTree).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>blocks</th>\n",
       "      <th>cigar</th>\n",
       "      <th>read</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41609</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>[(43161160, 43161190), (43162802, 43162922)]</td>\n",
       "      <td>((M, 30), (N, 1612), (M, 120))</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41608</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>[(43161160, 43161190), (43162802, 43162922)]</td>\n",
       "      <td>((M, 30), (N, 1612), (M, 120))</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44423</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11992</td>\n",
       "      <td>chr14</td>\n",
       "      <td>[(49586616, 49586766)]</td>\n",
       "      <td>((M, 150),)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44422</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11992</td>\n",
       "      <td>chr14</td>\n",
       "      <td>[(49586616, 49586766)]</td>\n",
       "      <td>((M, 150),)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44761</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:12085</td>\n",
       "      <td>chr14</td>\n",
       "      <td>[(49586699, 49586849)]</td>\n",
       "      <td>((M, 150),)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105752</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:27352</td>\n",
       "      <td>chr14</td>\n",
       "      <td>[(32163959, 32164109)]</td>\n",
       "      <td>((M, 150),)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15918</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>[(48233331, 48233481)]</td>\n",
       "      <td>((M, 150),)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15919</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>[(48233331, 48233481)]</td>\n",
       "      <td>((M, 150),)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34304</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>[(49586704, 49586854)]</td>\n",
       "      <td>((M, 150),)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34305</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>[(49586704, 49586854)]</td>\n",
       "      <td>((M, 150),)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112834 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               read_id  chrom  \\\n",
       "41609   NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "41608   NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22   \n",
       "44423   NGSNJ-086:229:GW200110425th:1:1101:10004:11992  chr14   \n",
       "44422   NGSNJ-086:229:GW200110425th:1:1101:10004:11992  chr14   \n",
       "44761   NGSNJ-086:229:GW200110425th:1:1101:10004:12085  chr14   \n",
       "...                                                ...    ...   \n",
       "105752   NGSNJ-086:229:GW200110425th:1:1101:9995:27352  chr14   \n",
       "15918     NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13   \n",
       "15919     NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13   \n",
       "34304     NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14   \n",
       "34305     NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14   \n",
       "\n",
       "                                              blocks  \\\n",
       "41609   [(43161160, 43161190), (43162802, 43162922)]   \n",
       "41608   [(43161160, 43161190), (43162802, 43162922)]   \n",
       "44423                         [(49586616, 49586766)]   \n",
       "44422                         [(49586616, 49586766)]   \n",
       "44761                         [(49586699, 49586849)]   \n",
       "...                                              ...   \n",
       "105752                        [(32163959, 32164109)]   \n",
       "15918                         [(48233331, 48233481)]   \n",
       "15919                         [(48233331, 48233481)]   \n",
       "34304                         [(49586704, 49586854)]   \n",
       "34305                         [(49586704, 49586854)]   \n",
       "\n",
       "                                 cigar read  \n",
       "41609   ((M, 30), (N, 1612), (M, 120))    2  \n",
       "41608   ((M, 30), (N, 1612), (M, 120))    1  \n",
       "44423                      ((M, 150),)    2  \n",
       "44422                      ((M, 150),)    1  \n",
       "44761                      ((M, 150),)    2  \n",
       "...                                ...  ...  \n",
       "105752                     ((M, 150),)    1  \n",
       "15918                      ((M, 150),)    1  \n",
       "15919                      ((M, 150),)    2  \n",
       "34304                      ((M, 150),)    1  \n",
       "34305                      ((M, 150),)    2  \n",
       "\n",
       "[112834 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "linear_reads = []\n",
    "for align in pysam.AlignmentFile(f'{output_base}/mapped_reads.bam', 'rb'):\n",
    "\tlinear_reads.append([\n",
    "\t\t\t\t\talign.query_name, \n",
    "\t\t\t\t\talign.reference_name, \n",
    "\t\t\t\t\talign.get_blocks(),\n",
    "\t\t\t\t\talign.cigartuples,\n",
    "\t\t\t\t\talign.is_read1,\n",
    "\t\t\t\t\t])\n",
    "\n",
    "linear_reads = pd.DataFrame(linear_reads, columns=LINEAR_COLUMNS)\n",
    "\n",
    "linear_reads.cigar = linear_reads.cigar.transform(lambda cigar: tuple((CIGARTUPLE_CODES[op],length) for op, length in cigar))\n",
    "linear_reads.read = linear_reads.read.map({True: '1', False: '2'})\n",
    "linear_reads = linear_reads.sort_values(['read_id'])\n",
    "\n",
    "linear_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>read</th>\n",
       "      <th>seg</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41609</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>2</td>\n",
       "      <td>(43161160, 43161190, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41609</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>2</td>\n",
       "      <td>(43162802, 43162922, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41608</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>1</td>\n",
       "      <td>(43161160, 43161190, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41608</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>1</td>\n",
       "      <td>(43162802, 43162922, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44423</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11992</td>\n",
       "      <td>chr14</td>\n",
       "      <td>2</td>\n",
       "      <td>(49586616, 49586766, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105752</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:27352</td>\n",
       "      <td>chr14</td>\n",
       "      <td>1</td>\n",
       "      <td>(32163959, 32164109, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15918</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>1</td>\n",
       "      <td>(48233331, 48233481, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15919</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>2</td>\n",
       "      <td>(48233331, 48233481, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34304</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>1</td>\n",
       "      <td>(49586704, 49586854, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34305</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>2</td>\n",
       "      <td>(49586704, 49586854, None)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>143056 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               read_id  chrom read  \\\n",
       "41609   NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22    2   \n",
       "41609   NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22    2   \n",
       "41608   NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22    1   \n",
       "41608   NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22    1   \n",
       "44423   NGSNJ-086:229:GW200110425th:1:1101:10004:11992  chr14    2   \n",
       "...                                                ...    ...  ...   \n",
       "105752   NGSNJ-086:229:GW200110425th:1:1101:9995:27352  chr14    1   \n",
       "15918     NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13    1   \n",
       "15919     NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13    2   \n",
       "34304     NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14    1   \n",
       "34305     NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14    2   \n",
       "\n",
       "                               seg  \n",
       "41609   (43161160, 43161190, None)  \n",
       "41609   (43162802, 43162922, None)  \n",
       "41608   (43161160, 43161190, None)  \n",
       "41608   (43162802, 43162922, None)  \n",
       "44423   (49586616, 49586766, None)  \n",
       "...                            ...  \n",
       "105752  (32163959, 32164109, None)  \n",
       "15918   (48233331, 48233481, None)  \n",
       "15919   (48233331, 48233481, None)  \n",
       "34304   (49586704, 49586854, None)  \n",
       "34305   (49586704, 49586854, None)  \n",
       "\n",
       "[143056 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "linear_reads['segs'] = linear_reads.apply(infer_segments, axis=1, result_type='reduce')\n",
    "linear_reads = linear_reads.explode('segs')\n",
    "linear_reads['seg'] = linear_reads.segs.transform(lambda segs: Interval(*segs))\n",
    "linear_reads = linear_reads.drop(columns=['blocks', 'segs', 'cigar'])\n",
    "linear_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No introns in chrM\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>read</th>\n",
       "      <th>seg</th>\n",
       "      <th>exons</th>\n",
       "      <th>introns</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41609</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>2</td>\n",
       "      <td>(43161160, 43161190, None)</td>\n",
       "      <td>{(43161051, 43161190, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41609</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>2</td>\n",
       "      <td>(43162802, 43162922, None)</td>\n",
       "      <td>{(43162802, 43163210, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41608</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>1</td>\n",
       "      <td>(43161160, 43161190, None)</td>\n",
       "      <td>{(43161051, 43161190, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41608</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>1</td>\n",
       "      <td>(43162802, 43162922, None)</td>\n",
       "      <td>{(43162802, 43163210, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44423</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11992</td>\n",
       "      <td>chr14</td>\n",
       "      <td>2</td>\n",
       "      <td>(49586616, 49586766, None)</td>\n",
       "      <td>{(49586579, 49586878, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>{(49586049, 49598399, {'gene_id': {'ENSG000002...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105752</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:27352</td>\n",
       "      <td>chr14</td>\n",
       "      <td>1</td>\n",
       "      <td>(32163959, 32164109, None)</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15918</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>1</td>\n",
       "      <td>(48233331, 48233481, None)</td>\n",
       "      <td>{(48233202, 48233477, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{(48232698, 48253806, {'gene_id': {'ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15919</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>2</td>\n",
       "      <td>(48233331, 48233481, None)</td>\n",
       "      <td>{(48233202, 48233477, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{(48232698, 48253806, {'gene_id': {'ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34304</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>1</td>\n",
       "      <td>(49586704, 49586854, None)</td>\n",
       "      <td>{(49586579, 49586878, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>{(49586049, 49598399, {'gene_id': {'ENSG000002...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34305</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>2</td>\n",
       "      <td>(49586704, 49586854, None)</td>\n",
       "      <td>{(49586579, 49586878, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>{(49586049, 49598399, {'gene_id': {'ENSG000002...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>143056 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               read_id  chrom read  \\\n",
       "41609   NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22    2   \n",
       "41609   NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22    2   \n",
       "41608   NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22    1   \n",
       "41608   NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22    1   \n",
       "44423   NGSNJ-086:229:GW200110425th:1:1101:10004:11992  chr14    2   \n",
       "...                                                ...    ...  ...   \n",
       "105752   NGSNJ-086:229:GW200110425th:1:1101:9995:27352  chr14    1   \n",
       "15918     NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13    1   \n",
       "15919     NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13    2   \n",
       "34304     NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14    1   \n",
       "34305     NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14    2   \n",
       "\n",
       "                               seg  \\\n",
       "41609   (43161160, 43161190, None)   \n",
       "41609   (43162802, 43162922, None)   \n",
       "41608   (43161160, 43161190, None)   \n",
       "41608   (43162802, 43162922, None)   \n",
       "44423   (49586616, 49586766, None)   \n",
       "...                            ...   \n",
       "105752  (32163959, 32164109, None)   \n",
       "15918   (48233331, 48233481, None)   \n",
       "15919   (48233331, 48233481, None)   \n",
       "34304   (49586704, 49586854, None)   \n",
       "34305   (49586704, 49586854, None)   \n",
       "\n",
       "                                                    exons  \\\n",
       "41609   {(43161051, 43161190, {'gene_id': {'ENSG000001...   \n",
       "41609   {(43162802, 43163210, {'gene_id': {'ENSG000001...   \n",
       "41608   {(43161051, 43161190, {'gene_id': {'ENSG000001...   \n",
       "41608   {(43162802, 43163210, {'gene_id': {'ENSG000001...   \n",
       "44423   {(49586579, 49586878, {'gene_id': {'ENSG000002...   \n",
       "...                                                   ...   \n",
       "105752                                                 {}   \n",
       "15918   {(48233202, 48233477, {'gene_id': {'ENSG000001...   \n",
       "15919   {(48233202, 48233477, {'gene_id': {'ENSG000001...   \n",
       "34304   {(49586579, 49586878, {'gene_id': {'ENSG000002...   \n",
       "34305   {(49586579, 49586878, {'gene_id': {'ENSG000002...   \n",
       "\n",
       "                                                  introns  \n",
       "41609                                                  {}  \n",
       "41609                                                  {}  \n",
       "41608                                                  {}  \n",
       "41608                                                  {}  \n",
       "44423   {(49586049, 49598399, {'gene_id': {'ENSG000002...  \n",
       "...                                                   ...  \n",
       "105752                                                 {}  \n",
       "15918   {(48232698, 48253806, {'gene_id': {'ENSG000001...  \n",
       "15919   {(48232698, 48253806, {'gene_id': {'ENSG000001...  \n",
       "34304   {(49586049, 49598399, {'gene_id': {'ENSG000002...  \n",
       "34305   {(49586049, 49598399, {'gene_id': {'ENSG000002...  \n",
       "\n",
       "[143056 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for chrom in linear_reads.chrom.unique():\n",
    "\tif chrom not in exons.keys():\n",
    "\t\tprint(f'No exons in {chrom}')\n",
    "\t\tcontinue\n",
    "\tchrom_exons = exons[chrom]\n",
    "\tlinear_reads.loc[linear_reads.chrom==chrom, 'exons'] = linear_reads.loc[linear_reads.chrom==chrom, 'seg'].transform(chrom_exons.overlap)\n",
    "\t\n",
    "\tif chrom not in introns.keys():\n",
    "\t\tprint(f'No introns in {chrom}')\n",
    "\t\tcontinue\n",
    "\tchrom_introns = introns[chrom]\n",
    "\tlinear_reads.loc[linear_reads.chrom==chrom, 'introns'] = linear_reads.loc[linear_reads.chrom==chrom, 'seg'].transform(chrom_introns.overlap)\n",
    "\n",
    "linear_reads.exons = linear_reads.exons.fillna('').transform(set)\n",
    "linear_reads.introns = linear_reads.introns.fillna('').transform(set)\n",
    "linear_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "common_genes\n",
      "1    89714\n",
      "2    36712\n",
      "0    14798\n",
      "3     1706\n",
      "4      114\n",
      "5       12\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/zp/3k8jr14j3wgcym12kxf4tmfw0000gr/T/ipykernel_93919/1809448561.py:13: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  linear_reads['common_genes'] = linear_reads.read_id.map(linear_reads.groupby('read_id').apply(infer_common_genes))\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>read</th>\n",
       "      <th>seg</th>\n",
       "      <th>exons</th>\n",
       "      <th>introns</th>\n",
       "      <th>common_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41609</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>2</td>\n",
       "      <td>(43161160, 43161190, None)</td>\n",
       "      <td>{(43161051, 43161190, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ENSG00000100300.18}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41609</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>2</td>\n",
       "      <td>(43162802, 43162922, None)</td>\n",
       "      <td>{(43162802, 43163210, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ENSG00000100300.18}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41608</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>1</td>\n",
       "      <td>(43161160, 43161190, None)</td>\n",
       "      <td>{(43161051, 43161190, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ENSG00000100300.18}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41608</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11303</td>\n",
       "      <td>chr22</td>\n",
       "      <td>1</td>\n",
       "      <td>(43162802, 43162922, None)</td>\n",
       "      <td>{(43162802, 43163210, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ENSG00000100300.18}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44423</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:10004:11992</td>\n",
       "      <td>chr14</td>\n",
       "      <td>2</td>\n",
       "      <td>(49586616, 49586766, None)</td>\n",
       "      <td>{(49586579, 49586878, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>{(49586049, 49598399, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>{ENSG00000213741.11, ENSG00000276168.1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105752</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:27352</td>\n",
       "      <td>chr14</td>\n",
       "      <td>1</td>\n",
       "      <td>(32163959, 32164109, None)</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15918</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>1</td>\n",
       "      <td>(48233331, 48233481, None)</td>\n",
       "      <td>{(48233202, 48233477, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{(48232698, 48253806, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{ENSG00000136156.15}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15919</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:4961</td>\n",
       "      <td>chr13</td>\n",
       "      <td>2</td>\n",
       "      <td>(48233331, 48233481, None)</td>\n",
       "      <td>{(48233202, 48233477, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{(48232698, 48253806, {'gene_id': {'ENSG000001...</td>\n",
       "      <td>{ENSG00000136156.15}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34304</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>1</td>\n",
       "      <td>(49586704, 49586854, None)</td>\n",
       "      <td>{(49586579, 49586878, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>{(49586049, 49598399, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>{ENSG00000213741.11, ENSG00000276168.1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34305</th>\n",
       "      <td>NGSNJ-086:229:GW200110425th:1:1101:9995:9502</td>\n",
       "      <td>chr14</td>\n",
       "      <td>2</td>\n",
       "      <td>(49586704, 49586854, None)</td>\n",
       "      <td>{(49586579, 49586878, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>{(49586049, 49598399, {'gene_id': {'ENSG000002...</td>\n",
       "      <td>{ENSG00000213741.11, ENSG00000276168.1}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>143056 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               read_id  chrom read  \\\n",
       "41609   NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22    2   \n",
       "41609   NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22    2   \n",
       "41608   NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22    1   \n",
       "41608   NGSNJ-086:229:GW200110425th:1:1101:10004:11303  chr22    1   \n",
       "44423   NGSNJ-086:229:GW200110425th:1:1101:10004:11992  chr14    2   \n",
       "...                                                ...    ...  ...   \n",
       "105752   NGSNJ-086:229:GW200110425th:1:1101:9995:27352  chr14    1   \n",
       "15918     NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13    1   \n",
       "15919     NGSNJ-086:229:GW200110425th:1:1101:9995:4961  chr13    2   \n",
       "34304     NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14    1   \n",
       "34305     NGSNJ-086:229:GW200110425th:1:1101:9995:9502  chr14    2   \n",
       "\n",
       "                               seg  \\\n",
       "41609   (43161160, 43161190, None)   \n",
       "41609   (43162802, 43162922, None)   \n",
       "41608   (43161160, 43161190, None)   \n",
       "41608   (43162802, 43162922, None)   \n",
       "44423   (49586616, 49586766, None)   \n",
       "...                            ...   \n",
       "105752  (32163959, 32164109, None)   \n",
       "15918   (48233331, 48233481, None)   \n",
       "15919   (48233331, 48233481, None)   \n",
       "34304   (49586704, 49586854, None)   \n",
       "34305   (49586704, 49586854, None)   \n",
       "\n",
       "                                                    exons  \\\n",
       "41609   {(43161051, 43161190, {'gene_id': {'ENSG000001...   \n",
       "41609   {(43162802, 43163210, {'gene_id': {'ENSG000001...   \n",
       "41608   {(43161051, 43161190, {'gene_id': {'ENSG000001...   \n",
       "41608   {(43162802, 43163210, {'gene_id': {'ENSG000001...   \n",
       "44423   {(49586579, 49586878, {'gene_id': {'ENSG000002...   \n",
       "...                                                   ...   \n",
       "105752                                                 {}   \n",
       "15918   {(48233202, 48233477, {'gene_id': {'ENSG000001...   \n",
       "15919   {(48233202, 48233477, {'gene_id': {'ENSG000001...   \n",
       "34304   {(49586579, 49586878, {'gene_id': {'ENSG000002...   \n",
       "34305   {(49586579, 49586878, {'gene_id': {'ENSG000002...   \n",
       "\n",
       "                                                  introns  \\\n",
       "41609                                                  {}   \n",
       "41609                                                  {}   \n",
       "41608                                                  {}   \n",
       "41608                                                  {}   \n",
       "44423   {(49586049, 49598399, {'gene_id': {'ENSG000002...   \n",
       "...                                                   ...   \n",
       "105752                                                 {}   \n",
       "15918   {(48232698, 48253806, {'gene_id': {'ENSG000001...   \n",
       "15919   {(48232698, 48253806, {'gene_id': {'ENSG000001...   \n",
       "34304   {(49586049, 49598399, {'gene_id': {'ENSG000002...   \n",
       "34305   {(49586049, 49598399, {'gene_id': {'ENSG000002...   \n",
       "\n",
       "                                   common_genes  \n",
       "41609                      {ENSG00000100300.18}  \n",
       "41609                      {ENSG00000100300.18}  \n",
       "41608                      {ENSG00000100300.18}  \n",
       "41608                      {ENSG00000100300.18}  \n",
       "44423   {ENSG00000213741.11, ENSG00000276168.1}  \n",
       "...                                         ...  \n",
       "105752                                       {}  \n",
       "15918                      {ENSG00000136156.15}  \n",
       "15919                      {ENSG00000136156.15}  \n",
       "34304   {ENSG00000213741.11, ENSG00000276168.1}  \n",
       "34305   {ENSG00000213741.11, ENSG00000276168.1}  \n",
       "\n",
       "[143056 rows x 7 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def infer_common_genes(df):\n",
    "\tgene_ids = []\n",
    "\tfor i, seg_row in df.iterrows():\n",
    "\t\tfeatures = seg_row['exons'].union(seg_row['introns'])\n",
    "\t\tseg_genes = set()\n",
    "\t\tfor feat in features:\n",
    "\t\t\tseg_genes.update(feat.data['gene_id'])\n",
    "\t\tgene_ids.append(seg_genes)\n",
    "\n",
    "\tcommon_genes = set(gid for gid in gene_ids[0] if all(gid in set_genes for set_genes in gene_ids))\n",
    "\treturn common_genes\n",
    "\n",
    "linear_reads['common_genes'] = linear_reads.read_id.map(linear_reads.groupby('read_id').apply(infer_common_genes))\n",
    "print(linear_reads.common_genes.transform(len).value_counts())\n",
    "linear_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "linear_reads['exons_f'] = linear_reads.apply(lambda row: set(exon for exon in row['exons'] if len(exon.data['gene_id'].intersection(row['common_genes']))>0), axis=1)\n",
    "linear_reads['introns_f'] = linear_reads.apply(lambda row: set(intron for intron in row['introns'] if len(intron.data['gene_id'].intersection(row['common_genes']))>0), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "truth",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
